<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Fishing for Clues in the Depths of Transcriptomic Data Series ‚Ä¢ cluefish</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Fishing for Clues in the Depths of Transcriptomic Data Series">
<meta name="description" content="A specialised bioinformatics workflow designed for improving biological interpretation of transcriptomic data series. Using Cluefish, you may consider a larger portion of the dataset of interest for biological interest and improve functional enrichment sensitivity towards smaller and more specific biological functions.">
<meta property="og:description" content="A specialised bioinformatics workflow designed for improving biological interpretation of transcriptomic data series. Using Cluefish, you may consider a larger portion of the dataset of interest for biological interest and improve functional enrichment sensitivity towards smaller and more specific biological functions.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">cluefish</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/cluefish.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ellfran-7/cluefish/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p><a id="readme-top"></a></p>
<!-- PROJECT LOGO -->
<div class="section level1">
<div class="page-header"><h1 id="cluefish-">cluefish <img src="reference/figures/cluefish-logo-with-text.png" alt="Logo" align="right" height="150" style="float:right; height:150px;"><a class="anchor" aria-label="anchor" href="#cluefish-"></a>
</h1></div>
<!-- PROJECT SHIELDS -->
<!--
*** I'm using markdown "reference style" links for readability.
*** Reference links are enclosed in brackets [ ] instead of parentheses ( ).
*** See the bottom of this document for the declaration of the reference variables
*** for contributors-url, forks-url, etc. This is an optional, concise syntax you may use.
*** https://www.markdownguide.org/basic-syntax/#reference-style-links
-->
<p><a href="https://github.com/ellfran-7/cluefish/graphs/contributors" class="external-link"><img src="https://img.shields.io/github/contributors/ellfran-7/cluefish.svg?style=flat-square" alt="Contributors"></a> <a href="https://github.com/ellfran-7/cluefish/network/members" class="external-link"><img src="https://img.shields.io/github/forks/ellfran-7/cluefish.svg?style=flat-square" alt="Forks"></a> <a href="https://github.com/ellfran-7/cluefish/stargazers" class="external-link"><img src="https://img.shields.io/github/stars/ellfran-7/cluefish.svg?style=flat-square" alt="Stargazers"></a> <a href="https://github.com/ellfran-7/cluefish/issues" class="external-link"><img src="https://img.shields.io/github/issues/ellfran-7/cluefish.svg?style=flat-square" alt="Issues"></a> <a href="https://github.com/ellfran-7/cluefish/blob/main/LICENSE" class="external-link"><img src="https://img.shields.io/badge/licence-CECILL_2.1-blue" alt="License"></a> <a href="https://linkedin.com/in/ellis-franklin-6188831ba" class="external-link"><img src="https://img.shields.io/badge/-LinkedIn-black.svg?style=flat-square&amp;logo=linkedin&amp;colorB=555" alt="LinkedIn"></a></p>
<!-- TABLE OF CONTENTS -->
<details><summary>
Table of Contents
</summary><ol>
<li>
<a href="#overview">Overview</a>
</li>
<li>
<a href="#installation">Installation</a>
</li>
<li>
<a href="#usage">Usage</a>
</li>
<li>
<a href="#contributing">Contributing</a>
</li>
<li>
<a href="#license">License</a>
</li>
<li>
<a href="#contact">Contact</a>
</li>
<li>
<a href="#acknowledgments">Acknowledgments</a>
</li>
</ol></details><!-- ABOUT THE PROJECT --><div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Cluefish is a free and open-source, semi-automated R workflow designed for comprehensive and untargeted exploration of transcriptomic data series. Its name reflects the three key concepts driving the workflow: <strong>Clustering</strong>, <strong>Enrichment</strong>, and <strong>Fishing</strong>‚Äîmetaphorically aligned with ‚Äú<em>fishing for clues</em>‚Äùüé£ in complex biological data.</p>
<p>When used alongside the <a href="https://lbbe-software.github.io/DRomics/" class="external-link">DRomics</a> (Dose-Response for Omics) R package, Cluefish provides a more comprehensive analysis of dose-response transcriptomic data. In toxicology/ecotoxicology, this will support the understanding/highlighting of contaminant‚Äôs mode of action.</p>
<p>This workflow addresses the limitations of the standard Over-Representation Analysis (ORA) by applying ORA to pre-clustered networks. These clusters serve as anchors for ORA, enhancing enrichment detection sensitivity and thus enabling the identification of smaller, more specific biological processes while simultaneously forming exploratory gene groups.</p>
<p>Cluefish is designed to be adaptable to a wide range of organisms, both model and non-model, ensuring broad applicability across various biological contexts.</p>
<p align="right">
</p>
<p>(<a href="#readme-top">back to top</a>)</p>

<!-- INSTALLATION -->
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The Cluefish tool is developed in <strong>R</strong>, so having <strong>R</strong> installed is a prerequisite. You can download it <a href="https://posit.co/download/rstudio-desktop/" class="external-link">here</a>.</p>
<p>For an enhanced experience, we recommend using the <strong>RStudio</strong> integrated development environment (IDE), which is available for download at the same link, <a href="https://posit.co/download/rstudio-desktop/" class="external-link">here</a>.</p>
<p>You can use Cluefish locally in one of two ways:</p>
<ol style="list-style-type: decimal">
<li>
<p>Clone the repository via a terminal:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/ellfran-7/cluefish.git</span></code></pre></div>
</li>
<li>
<p>Install the developmental version of Cluefish from GitHub in R (<code>remotes</code> needed):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ellfran-7/cluefish"</span><span class="op">)</span></span></code></pre></div>
</li>
</ol>
<p align="right">
</p>
<p>(<a href="#readme-top">back to top</a>)</p>

<!-- ADDITIONAL REQUIREMENTS -->
</div>
<div class="section level2">
<h2 id="additional-requirements">Additional Requirements<a class="anchor" aria-label="anchor" href="#additional-requirements"></a>
</h2>
<p>Cluefish relies on external open source software for an intermediate step within its workflow. Please ensure the following tools are installed:</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Cytoscape</strong>:</p>
<p>Cluefish uses Cytoscape in order to visualize PPI networks. Install Cytoscape from their <a href="https://cytoscape.org/download.html" class="external-link">download page</a>.</p>
</li>
<li>
<p><strong>Required Cytoscape Apps</strong>:</p>
<p>Within Cytoscape, install the <strong>StringApp</strong> and <strong>clusterMaker2</strong> apps. To do this:</p>
<ul>
<li>Open Cytoscape</li>
<li>Navigate to <code>Apps</code> &gt; <code>App Store</code> &gt; <code>Show App Store</code>
</li>
<li>Search for and install ‚ÄúStringApp‚Äù (for retrieving STRING protein interactions) and ‚ÄúclusterMaker2‚Äù‚Äù (for clustering network data).</li>
</ul>
</li>
</ol>
<p><em>You can also view more about these apps on the <a href="https://apps.cytoscape.org/" class="external-link">Cytoscape App Store</a>.</em></p>
<p align="right">
</p>
<p>(<a href="#readme-top">back to top</a>)</p>

<!-- USAGE EXAMPLES -->
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>To run the Cluefish workflow, you can use the <code>make.R</code> script, which serves as the ‚Äòmaster‚Äô script for the entire process. We recommend using this script as a template to ensure smooth and sequential execution of the workflow steps.</p>
<div class="section level3">
<h3 id="required-r-packages">Required R packages<a class="anchor" aria-label="anchor" href="#required-r-packages"></a>
</h3>
<p>A key feature of Cluefish is the integration of <code>renv</code> to create reproducible environments. This allows you to install the required R packages in two ways:</p>
<ul>
<li>Run <code><a href="https://rstudio.github.io/renv/reference/install.html" class="external-link">renv::install()</a></code> to install the most recent version of the packages listed in the <code>renv.lock</code> file.</li>
<li>For full reproducibility, run <code><a href="https://rstudio.github.io/renv/reference/restore.html" class="external-link">renv::restore()</a></code> to install the exact package versions specified in the <code>renv.lock</code> file. Note that this process may take longer.</li>
</ul>
</div>
<div class="section level3">
<h3 id="required-inputs">Required inputs<a class="anchor" aria-label="anchor" href="#required-inputs"></a>
</h3>
<p>Cluefish requires two key inputs:</p>
<ol style="list-style-type: decimal">
<li>
<strong>A background transcript list</strong>: Typically, this includes the identifiers for all detected transcripts in the experiment.</li>
<li>
<strong>A deregulated transcript list</strong>: A subset of the background list, containing the identifiers of significantly deregulated transcripts. This list can be derived using any selection method.</li>
</ol>
</div>
<div class="section level3">
<h3 id="recommended-selection-method">Recommended Selection Method<a class="anchor" aria-label="anchor" href="#recommended-selection-method"></a>
</h3>
<p>While the inputs can be derived from any selection method, Cluefish was optimised to work seamlessly with the results from <code>DRomics</code>, a tool tailored for dose-response modelling of omics data.</p>
<p>Although using <code>DRomics</code> is optional, Cluefish leverages some of its visualization functions and modelling metrics to provide deeper insights into the biological interpretation of the data.</p>
<p><em>For more information on DRomics, please refer to their <a href="https://lbbe-software.github.io/DRomics/" class="external-link">documentation</a></em>.</p>
<p align="right">
</p>
<p>(<a href="#readme-top">back to top</a>)</p>

<!-- WORKFLOW -->
</div>
</div>
<div class="section level2">
<h2 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h2>
<p>A schematic overview of the Cluefish workflow is shown below. For a full, step-by-step guide, refer to the vignette, <a href="https://ellfran-7.github.io/cluefish/articles/cluefish.html">Introduction to Cluefish</a>, which provides instructions using the <em>ZebrafishDBP</em> example dataset. The raw count data is publicly available on NCBI GEO and can be accessed with <strong>GSE283957</strong>.</p>
<center>
<p><img src="reference/figures/cluefish-schematic-2024-09-27-rm-asterisk.png" alt="Cluefish schematic" width="600"></p>
</center>
<p align="right">
</p>
<p>(<a href="#readme-top">back to top</a>)</p>

<!-- CITATION -->
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>If you use Cluefish, for the time being please cite it as follows:</p>
<blockquote>
<p>Franklin E, Billoir E, Delignette-Muller M, Prud‚Äôhomme S (2024). cluefish: Fishing for Clues in the Depths of Transcriptomic Data Series. R package version 1.0.0, <a href="https://ellfran-7.github.io/cluefish/" class="uri">https://ellfran-7.github.io/cluefish/</a>.</p>
</blockquote>
<p>A paper will shortly be associated with the tool.</p>
<!-- CONTRIBUTING -->
</div>
<div class="section level2">
<h2 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h2>
<p>Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are <strong>greatly appreciated</strong>.</p>
<p>If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag ‚Äúenhancement‚Äù. Don‚Äôt forget to give the project a star! Thanks again!</p>
<ol style="list-style-type: decimal">
<li>Fork the Project</li>
<li>Create your Feature Branch (<code>git checkout -b feature/AmazingIdea</code>)</li>
<li>Commit your Changes (<code>git commit -m 'Add some AmazingIdea'</code>)</li>
<li>Push to the Branch (<code>git push origin feature/AmazingIdea</code>)</li>
<li>Open a Pull Request</li>
</ol>
<p align="right">
</p>
<p>(<a href="#readme-top">back to top</a>)</p>

<!-- LICENSE -->
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>Distributed under the CeCILL Free Software License Agreement v2.1. See <code>LICENSE.txt</code> for more information.</p>
<p>CECILL-2.1 is compatible with GNU GPL. See the <a href="http://www.cecill.eu/index.en.html" class="external-link">official CeCILL site</a> for more information.</p>
<p align="right">
</p>
<p>(<a href="#readme-top">back to top</a>)</p>

<!-- CONTACT -->
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>If you have any need that is not yet covered, any feedback on Cluefish, or anything other question, feel free to contact me !</p>
<p>Ellis Franklin - <a href="https://twitter.com/elffran7" class="external-link">@elffran7</a> - <a href="mailto:ellis.franklin@univ-lorraine.fr" class="email">ellis.franklin@univ-lorraine.fr</a></p>
<p>Project Link: <a href="https://github.com/ellfran-7/cluefish" class="external-link uri">https://github.com/ellfran-7/cluefish</a></p>
<p align="right">
</p>
<p>(<a href="#readme-top">back to top</a>)</p>

<!-- ACKNOWLEDGMENTS -->
</div>
<div class="section level2">
<h2 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<ul>
<li><a href="https://github.com/othneildrew/Best-README-Template" class="external-link">Othneil Drew‚Äôs README template</a></li>
<li><a href="https://flexbox.malven.co/" class="external-link">Malven‚Äôs Flexbox Cheatsheet</a></li>
<li><a href="https://grid.malven.co/" class="external-link">Malven‚Äôs Grid Cheatsheet</a></li>
<li><a href="https://shields.io/" class="external-link">Img Shields</a></li>
</ul>
<p align="right">
</p>
<p>(<a href="#readme-top">back to top</a>)</p>

<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/ellfran-7/cluefish/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/ellfran-7/cluefish" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.cecill.info/licences/Licence_CeCILL_V2.1-en.txt" class="external-link">CECILL-2.1</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing cluefish</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Ellis Franklin <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0000-0002-6614-4109" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Elise Billoir <br><small class="roles"> Author, reviewer </small> <a href="https://orcid.org/0000-0001-9012-3298" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Marie Laure Delignette-Muller <br><small class="roles"> Author, reviewer </small> <a href="https://orcid.org/0000-0001-5453-3994" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Sophie Prud'homme <br><small class="roles"> Author, reviewer </small> <a href="https://orcid.org/0000-0002-7199-1839" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ellis Franklin, Elise Billoir, Marie Laure Delignette-Muller, Sophie Prud‚Äôhomme.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
