<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to cluefish • cluefish</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to cluefish">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cluefish</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/cluefish.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ellfran-7/cluefish/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="cluefish_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="cluefish_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="cluefish_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="cluefish_files/datatables-binding-0.33/datatables.js"></script><link href="cluefish_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="cluefish_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="cluefish_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="cluefish_files/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="cluefish_files/nouislider-7.0.10/jquery.nouislider.min.js"></script><link href="cluefish_files/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="cluefish_files/selectize-0.12.0/selectize.min.js"></script><link href="cluefish_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="cluefish_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to cluefish</h1>
                        <h4 data-toc-skip class="author">Ellis
Franklin</h4>
                        
            <h4 data-toc-skip class="date">2025-06-06</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ellfran-7/cluefish/blob/HEAD/vignettes/cluefish.Rmd" class="external-link"><code>vignettes/cluefish.Rmd</code></a></small>
      <div class="d-none name"><code>cluefish.Rmd</code></div>
    </div>

    
    
<!-- Introduction  -->
<!--------------- Add Cluefish graphical abstract from paper -->
<center>
<img src="figures%2Fcluefish-graphical-abstract-2025-06-06.png" alt="Cluefish graphical abstract"><p align="center" style="color: grey;">
<em>Graphical abstract of the Cluefish workflow.</em>
</p>
</center>
<div class="section level2">
<h2 id="preface">Preface<a class="anchor" aria-label="anchor" href="#preface"></a>
</h2>
<p>This vignette is intended to help users who wish to use Cluefish. The
first part of this vignette is dedicated to an introduction of the
context in which Cluefish was developped, followed by an overview of the
workflow. The second part is dedicated to a step-by-step guide of the
main Cluefish workflow using an example dose-response (DR) dataset.</p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="a-bit-of-context">A bit of context<a class="anchor" aria-label="anchor" href="#a-bit-of-context"></a>
</h3>
<p>In transcriptomics, analyses yield extensive transcript lists that
require impractical manual literature reviews. To address this,
<strong>functional enrichment (FE) analysis</strong>, also known as
<strong>pathway enrichment analysis</strong>, has become the standard
approach. This method condenses large gene lists into more manageable
and interpretable sets of biological functions or pathways <span class="citation">(Das, McClain, and Rai 2020)</span>. FE relies on
predefined gene sets representing molecular functions, biological
processes or pathways, as outlined by resources like the Gene Ontology
(GO) <span class="citation">(Ashburner et al. 2000)</span> or pathway
databases such as KEGG <span class="citation">(Kanehisa and Goto
2000)</span> and Wikipathways <span class="citation">(Martens et al.
2021)</span>.</p>
<p>FE approaches have evolved in four generations: Over-Representation
Analysis (ORA), Functional Class Scoring (FCS), Pathway Topology (PT),
and Network Topology (NT)-based methods<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Multiple literature reviews provide healthy descriptions
and comparisons of these approaches: &lt;span class="citation"&gt;Khatri,
Sirota, and Butte (2012)&lt;/span&gt;; &lt;span class="citation"&gt;Ihnatova,
Popovici, and Budinska (2018)&lt;/span&gt;; &lt;span class="citation"&gt;Ma,
Shojaie, and Michailidis (2019)&lt;/span&gt;; &lt;span class="citation"&gt;García-Campos, Espinal-Enríquez, and Hernández-Lemus
(2015)&lt;/span&gt;; &lt;span class="citation"&gt;Bayerlová et al. (2015)&lt;/span&gt;;
&lt;span class="citation"&gt;Huang, Sherman, and Lempicki (2009)&lt;/span&gt;;.&lt;/p&gt;'><sup>1</sup></a>.</p>
<p>Each generation introduced improvements by accounting for gene
co-expression, pathway topology, and network crosstalk. However, PT and
NT-based approaches face limitations due to incomplete pathway and
network information for lesser studied organisms, making ORA and FCS the
most widely used methods <span class="citation">(Liu, Wei, and Ruan
2017)</span>.</p>
<p>While these earlier methods are commonly applied to transcriptomic
data with two or three conditions, dealing with data involving many
ordered conditions—referred to as “data series”—poses substantial
complexity.</p>
<p>In dose-response (DR) studies, which explore the relationship between
exposure and biological effect, DR omics analyses play a critical role
<span class="citation">(Lovell 2016)</span>. Unlike standard
differential expression analyses, DR methods, such as the <a href="https://lbbe-software.github.io/DRomics/" class="external-link"><code>DRomics</code></a>
(Dose-Response for Omics) R package <span class="citation">(Larras et
al. 2018)</span>, model the full DR relationship of each transcript
rather than just fold-change between conditions. From these models, a
benchmark dose (BMD) is calculated for each transcript : a dose that
represents a specific change in response, indicating a gene’s
sensitivity to a stressor <span class="citation">(Committee et al.
2017)</span>.</p>
<p>While differential gene expression analysis provides fold-change
values and their associated p-values, modelling approaches, such as
<code>DRomics</code>, characterise the entire DR relationship. PT-based
enrichment and FCS, which rely on fold-change values, are therefore not
adapted for functional enrichment in a DR context <span class="citation">(Khatri, Sirota, and Butte 2012)</span>. ORA is
currently the only suitable method available to complement DR modelling,
as it exclusively uses deregulated gene lists. However, ORA comes with
several other limitations. The enrichment results, often based on
Fisher’s Exact Tests, can be biased towards broad biological functions,
since using all deregulated genes makes it harder to reach statistical
significance for smaller or specialised functions. This tendency is
intrinsically tied to the granularity of gene sets in certain databases
(e.g., GO). Gene sets associated to a biological process/pathway may be
overly specific, while others are too general, potentially compromising
the relevance of the analysis <span class="citation">(Karp et al. 2021;
Mubeen et al. 2022)</span>. Additionally, gene sets may vary in
specificity, and results often reflect only part of the deregulated gene
list. Therefore, ORA alone (hereafter referred to as the
“<strong>standard approach</strong>”) is insufficient for interpreting
complex DR data, and combining additional methods is essential to
capture a fuller and more precise view of the biological context.</p>
</div>
<div class="section level3">
<h3 id="overview-of-cluefish">Overview of Cluefish<a class="anchor" aria-label="anchor" href="#overview-of-cluefish"></a>
</h3>
<p>To address these limitations, we developed <a href="https://github.com/ellfran-7/cluefish" class="external-link">Cluefish</a> <a href="https://www.biorxiv.org/content/10.1101/2024.12.18.627334v1" class="external-link">Franklin
<em>et al.</em> (submitted)</a>, a free and open-source, semi-automated
workflow designed for comprehensive and untargeted exploration of
transcriptomic data series. Its name reflects the three key concepts
driving the workflow: <strong>Clustering</strong>,
<strong>Enrichment</strong>, and <strong>Fishing</strong>—metaphorically
aligned with “<em>fishing for clues</em>”🎣 in complex biological
data.</p>
<p>Cluefish is composed of 10 main steps:</p>
<ol style="list-style-type: decimal">
<li>Download transcription factor (TF) and co-factor (CoTF) gene
annotations (<em>optional</em>) (<a href="#step1">Step 1</a>).</li>
<li>Load background and deregulated transcript lists (<a href="#step2">Step 2</a>).</li>
<li>Retrieve gene identifiers (<a href="#step3">Step 3</a>).</li>
<li>Assign regulatory status (<a href="#step4">Step 4</a>) (<em>only
possible if Step 1 is performed</em>).</li>
<li>Construct, cluster the PPI network of the deregulated gene list and
then retrieve the results (<a href="#step5">Step 5</a>).</li>
<li>Filter gene clusters based on gene set size (<a href="#step6">Step
6</a>).</li>
<li>Perform cluster-wise functional enrichment (<a href="#step7">Step
7</a>).</li>
<li>Merge clusters sharing enriched biological functions (<a href="#step8">Step 8</a>).</li>
<li>Fish <em>lonely genes</em> into existing clusters based on shared
functional annotations (<a href="#step9">Step 9</a>).</li>
<li>Perform simple functional enrichment on the remaining lonely genes,
forming the <em>lonely cluster</em> (<a href="#step10">Step
10</a>).</li>
</ol>
<center>
<img src="figures%2Fcluefish-schematic-2024-09-27-rm-asterisk.png" alt="Cluefish schematic" width="600"><p align="center" style="color: grey;">
<em>Schematic of the Cluefish workflow.</em>
</p>
</center>
<p>Cluefish integrates valuable information from various biological
functional/pathway databases (e.g., GO, KEGG, WP), the AnimalTFDB4
database of transcription factors and co-factors for &gt;180 animal
genomes, and the STRING database of known an predicted PPIs for
&gt;12,000. Despite drawing from multiple complementary resources, these
databases collectively provide a broad coverage of organisms, ensuring
that Cluefish is applicable to a wide range of both model and non-model
species.</p>
<p><br></p>
<!-- Installation -->
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can use Cluefish locally in one of two ways:</p>
<ol style="list-style-type: decimal">
<li>
<p>Clone the repository via a terminal:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/ellfran-7/cluefish.git</span></code></pre></div>
</li>
<li>
<p>Install the developmental version of Cluefish from GitHub in R
(<code>remotes</code> needed):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ellfran-7/cluefish"</span><span class="op">)</span></span></code></pre></div>
</li>
</ol>
<p><br></p>
<!-- Additional Requirements -->
</div>
<div class="section level2">
<h2 id="additional-requirements">Additional Requirements<a class="anchor" aria-label="anchor" href="#additional-requirements"></a>
</h2>
<p>The Cluefish workflow is developed in <strong>R</strong>, so having R
installed is a prerequisite. You can download it <a href="https://posit.co/download/rstudio-desktop/" class="external-link">here</a>.</p>
<p>For an enhanced experience, we recommend using the
<strong>RStudio</strong> integrated development environment (IDE), which
is available for download at the same link, <a href="https://posit.co/download/rstudio-desktop/" class="external-link">here</a>.</p>
<p>Additionally, Cluefish relies on external open source software for an
intermediate step within its workflow. Please ensure the following tools
are installed:</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Cytoscape</strong>:</p>
<p>Cluefish uses Cytoscape in order to visualise PPI networks. Install
Cytoscape from their <a href="https://cytoscape.org/download.html" class="external-link">download page</a>.</p>
</li>
<li>
<p><strong>Required Cytoscape Apps</strong>:</p>
<p>Within Cytoscape, install the <strong>StringApp</strong> and
<strong>clusterMaker2</strong> apps. To do this:</p>
<ul>
<li>Open Cytoscape</li>
<li>Navigate to <code>Apps</code> &gt; <code>App Store</code> &gt;
<code>Show App Store</code>
</li>
<li>Search for and install “StringApp” (for retrieving STRING protein
interactions) and “clusterMaker2” (for clustering network data).</li>
</ul>
</li>
</ol>
<p><strong>You can also view more about these apps on the <a href="https://apps.cytoscape.org/" class="external-link">Cytoscape App Store</a>.</strong></p>
<p><br></p>
<!-- Naming Convention -->
</div>
<div class="section level2">
<h2 id="naming-convention">Naming Convention<a class="anchor" aria-label="anchor" href="#naming-convention"></a>
</h2>
<p>To enhance consistency and readability, we have incorporated a
standardised naming convention for data files, inputs, intermediate
objects, and outputs:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Data type</strong>:
<ul>
<li>
<code>dr</code>: deregulated data (only deregulated items)</li>
<li>
<code>bg</code>: background data (all experimental items)</li>
</ul>
</li>
<li>
<strong>Data format</strong>:
<ul>
<li>
<code>t</code>: transcript-per-row data, where each row corresponds
to a single transcript</li>
<li>
<code>g</code>: gene-per-row data, where each row corresponds to a
single gene</li>
</ul>
</li>
<li>
<strong>Annotation</strong>: For each of the aforementioned
patterns, an optional biological annotation ‘<code>a</code>’ is
appended. This annotation facilitates a specific delineation, providing
one row per annotation. For example, ‘<code>t_a</code>’ signifies that
the data is formatted with one row corresponding to a single transcript
alongside one of its corresponding annotations.</li>
</ol>
<p><u>Examples</u>:</p>
<ul>
<li>
<code>bg_t_data</code>: background data with each row corresponding
to a single transcript</li>
<li>
<code>dr_g_a_data</code>: deregulated data with each row
corresponding to a specific combination of gene and annotation</li>
</ul>
<p><br></p>
<!-- Getting started -->
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>To run the Cluefish workflow, you can use the <code>make.R</code>
script, which serves as the ‘master’ script for the entire process. We
recommend using this script as a template to ensure smooth and
sequential execution of the workflow steps.</p>
<p>Here we will go through the main steps one by one !</p>
<div class="section level3">
<h3 id="the-case-study">The case study<a class="anchor" aria-label="anchor" href="#the-case-study"></a>
</h3>
<p>In order to showcase Cluefish, we will use an example dataset. This
is a dose-response transcriptomic dataset, where zebrafish embryos were
exposed to a five dose-gradient of dibutyl phthalate. The experiment and
resulting dataset is further described in the associated paper <a href="https://www.biorxiv.org/content/10.1101/2024.12.18.627334v1" class="external-link">Franklin
<em>et al.</em> (submitted)</a> and you may access the count data at
GEO, under the accession number <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE283957" class="external-link">GSE283957</a>.</p>
<div class="note">
<p>In addition to the zebrafish dataset used in this vignette, two
external transcriptomic datasets were analysed in the associated study:
the rat liver dataset from Gwinn <em>et al.</em> (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE147072" class="external-link">GSE147072</a>)
and the poplar root dataset from Gréau <em>et al.</em> (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE263776" class="external-link">GSE263776</a>).
Both datasets are publicly available through the GEO database.You may
find their analysis files in the repository itself.</p>
</div>
</div>
<div class="section level3">
<h3 id="before-walking">Before walking<a class="anchor" aria-label="anchor" href="#before-walking"></a>
</h3>
<p>A key feature of Cluefish is the integration of <code>renv</code>
<span class="citation">(Ushey and Wickham 2024)</span> to create
reproducible environments. This enables you to install the required R
packages in two ways:</p>
<ol style="list-style-type: decimal">
<li>Install the latest package versions with
<code><a href="https://rstudio.github.io/renv/reference/install.html" class="external-link">renv::install()</a></code>.</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>For full reproducibility, install the exact package versions
specified in the <code>renv.lock</code> file by running
<code><a href="https://rstudio.github.io/renv/reference/restore.html" class="external-link">renv::restore()</a></code>. Note that this process may take
longer.</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/restore.html" class="external-link">restore</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Once the packages installed or restored, you can then load the
project R functions with:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/load_all.html" class="external-link">load_all</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>If no errors have appeared, you are all set to go !</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step1">Step 1: Download transcription (co-)factor annotations<a class="anchor" aria-label="anchor" href="#step1"></a>
</h3>
<p>Before proceeding, check if your organism of interest is available in
the AnimalTFDB database. A list of supported species can be found on
their website: <a href="https://guolab.wchscu.cn/AnimalTFDB4/#/Species" class="external-link">Go to
AnimalTFDB</a>.</p>
<ol style="list-style-type: decimal">
<li><p>If you organism is not listed in the database, you may skip this
step.</p></li>
<li>
<p>If your organism is listed, retrieve the URLs for both the
transcription (TF) and co-transcription factors (CoTF). For example, the
URLs for zebrafish (<em>Danio rerio</em>) are:</p>
<ul>
<li>TF: “<a href="https://guolab.wchscu.cn/AnimalTFDB4_static/download/TF_list_final/Danio_rerio_TF" class="external-link uri">https://guolab.wchscu.cn/AnimalTFDB4_static/download/TF_list_final/Danio_rerio_TF</a>”</li>
<li>CoTF: “<a href="https://guolab.wchscu.cn/AnimalTFDB4_static/download/Cof_list_final/Danio_rerio_Cof" class="external-link uri">https://guolab.wchscu.cn/AnimalTFDB4_static/download/Cof_list_final/Danio_rerio_Cof</a>”</li>
</ul>
</li>
</ol>
<p>To download data for a different species, replace
“<code>Danio_rerio</code>” in the URL with the Latin name of your
organism.</p>
<p>Use the <span class="homemade-function"><code><a href="../reference/dl_regulation_data.html">dl_regulation_data()</a></code></span>
function downloads the TF and CoTF data files to a specified directory.
Customise the file paths and names by setting the function arguments as
shown below:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">dl_regulation_data</span>(</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>  <span class="at">url_tf =</span> <span class="st">"https://guolab.wchscu.cn/AnimalTFDB4_static/download/TF_list_final/Danio_rerio_TF"</span>,</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>  <span class="at">url_cof =</span> <span class="st">"https://guolab.wchscu.cn/AnimalTFDB4_static/download/Cof_list_final/Danio_rerio_Cof"</span>,</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>  <span class="at">path =</span> <span class="st">"your/chosen/path"</span>, <span class="co"># the path where you want to save the files</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>  <span class="at">filename_tf =</span> <span class="st">"TF_filename.txt"</span>, <span class="co"># the chosen TF filename </span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="at">filename_cof =</span> <span class="st">"CoTF_filename.txt"</span>), <span class="co"># the chosen CoTF filename</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  overwrite <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="er">)</span></span></code></pre></div>
<hr>
<p><span class="italic-grey"> The <code>overwrite</code> argument
controls whether existing files with the same name in the target
directory are replaced. Setting <code>overwrite = TRUE</code> allows the
function to overwrite files, while <code>overwrite = FALSE</code>
preserves existing files to prevent accidental loss of data.</span></p>
<hr>
<p>For more info on each argument of the function, run
<code>?cluefish::dl_regulation_data()</code>.</p>
<p>For simplicity in this vignette, we will use two pre-saved example
files: <code>example_TF_file.txt</code> and
<code>example_CoTF_file.txt</code>.</p>
<p>The transcription (co)factor data will be used in <a href="#step3">Step 3</a> to determine the regulatory status of the
deregulated genes.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step2">Step 2: Load background and deregulated transcripts lists<a class="anchor" aria-label="anchor" href="#step2"></a>
</h3>
<p>For this step, load the <strong>background</strong> and
<strong>deregulated transcript lists</strong>. These must be acquired
beforehand.</p>
<p>The <strong>background transcript list</strong> corresponds to a
character vector of all transcript IDs detected in the experiment,
whereas the <strong>deregulated transcript list</strong> contains only
the IDs of significantly deregulated transcripts.</p>
<p>Multiple identifiers are available for transcripts and genes.</p>
<p>Pay special attention to the type of transcript identifier used, as
this will be crucial for the next step. For instance,
“<strong>ENSDART00000080481.6</strong>” is the <em>Ensembl transcript
stable ID</em> for the <em>rxraa-201</em>.</p>
<hr>
<p><span class="italic-grey">While the inputs can be derived from any
selection method, Cluefish was optimised to work seamlessly with the
results from <code>DRomics</code>. In addition, Cluefish leverages some
of <code>DRomics</code> visualisation functions and modelling metrics to
provide deeper insights into the biological interpretation of the
data.</span></p>
<hr>
<p>With the <code>DRomics</code> analysis performed beforehand in our
case, we can retrieve the outputs of:</p>
<ul>
<li>the <code>drcfit()</code> function: holding the background
transcript list</li>
<li>the <code>bmdboot()</code> function: holding the deregulated
transcript list</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the DRomics "drcfit" object</span></span>
<span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/example_fitres.rds"</span><span class="op">)</span></span>
<span><span class="co"># Load the DRomics "bmdboot" object results ($res)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/example_bootres.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the background transcript identifiers as a character vector</span></span>
<span><span class="va">bg_transcript_list</span> <span class="op">&lt;-</span> <span class="va">f</span><span class="op">$</span><span class="va">omicdata</span><span class="op">$</span><span class="va">item</span></span>
<span><span class="co"># Extract the deregulated transcript identifiers as a character vector</span></span>
<span><span class="va">dr_transcript_list</span> <span class="op">&lt;-</span> <span class="va">b</span><span class="op">$</span><span class="va">id</span></span></code></pre></div>
<p>We can get a look of the transcript identifiers and the count for
each list with either <code><a href="https://rdrr.io/r/utils/str.html" class="external-link">str()</a></code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">bg_transcript_list</span><span class="op">)</span></span>
<span><span class="co">#&gt;  chr [1:39890] "ENSDART00000139936.2" "ENSDART00000192840.1" ...</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">dr_transcript_list</span><span class="op">)</span></span>
<span><span class="co">#&gt;  chr [1:2433] "ENSDART00000139096.2" "ENSDART00000127420.3" ...</span></span></code></pre></div>
<p>Or if the <code>dplyr</code> package is installed:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">bg_transcript_list</span><span class="op">)</span></span>
<span><span class="co">#&gt;  chr [1:39890] "ENSDART00000139936.2" "ENSDART00000192840.1" ...</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">dr_transcript_list</span><span class="op">)</span></span>
<span><span class="co">#&gt;  chr [1:2433] "ENSDART00000139096.2" "ENSDART00000127420.3" ...</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step3">Step 3: Retrieve gene identifiers<a class="anchor" aria-label="anchor" href="#step3"></a>
</h3>
<p>Next, retrieve specific gene identifiers. Before proceeding, we
recommend gathering some essential elements from Ensembl BioMart:</p>
<ul>
<li>The Ensembl BioMart service where your organism’s dataset is
located</li>
<li>The specific dataset name for your species within the BioMart
service.</li>
<li>The identifiers you want to retrieve (e.g, gene)</li>
</ul>
<p>It’s important to gather this information in advance, as each element
has unique naming conventions specific to Ensembl BioMart. To do so
:</p>
<ol style="list-style-type: decimal">
<li>
<p>Identify the Ensembl BioMart services currently available. To
list all active services, use the <code>listEnsembl()</code> function
(or <code>listMarts()</code>):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">biomaRt</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listMarts.html" class="external-link">listMarts</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">biomaRt</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listEnsembl.html" class="external-link">listEnsembl</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The main Ensembl BioMart service provides annotations for vertebrate
genomes. For annotations of Protists, Metazoa, and Fungi, specialised
BioMart interfaces are also available. We can use the
<code>listEnsemblGenomes()</code>function to list these options:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">biomaRt</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listEnsembl.html" class="external-link">listEnsemblGenomes</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>Once the appropriate Ensembl BioMart service is identified,
connect to it using <code>useMart()</code>, <code>useEnsembl(</code>),
or <code>useEnsemblGenomes()</code>. Specify the <code>biomart</code>
argument with a name from the previous step’s output:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="fu">biomaRt</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html" class="external-link">useMart</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>Within the selected BioMart service, each species has its own
dataset. List the available datasets using the
<code>listDatasets()</code> function:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">biomaRt</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listDatasets.html" class="external-link">listDatasets</a></span><span class="op">(</span>mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>BioMart allows you to retrieve various types of information
(e.g., gene or transcript IDs). Use either <code>listAttributes()</code>
or <code>listFilters()</code> to view available identifiers and
filters:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">biomaRt</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listAttributes.html" class="external-link">listAttributes</a></span><span class="op">(</span>mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span></span>
<span><span class="co">#OR </span></span>
<span><span class="fu">biomaRt</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listFilters.html" class="external-link">listFilters</a></span><span class="op">(</span>mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span></span></code></pre></div>
</li>
</ol>
<p>With the required elements gathered, employ the <span class="homemade-function"><code><a href="../reference/getids.html">getids()</a></code></span> function to
retrieve gene (<code>gene_id</code>) and gene name
(<code>gene_name</code>) identifiers corresponding to the transcript
identifiers (<code>transcript_id</code>) associated to the background
transcript list (<code>bg_transcripts_list</code>).</p>
<p><span class="italic-red"><u>Important</u>: To ensure compatibility
with the databases and tools used later in Cluefish—specifically the
<strong>STRING</strong> database and <strong>g:Profiler</strong>—it is
important that the output dataframe from the <code><a href="../reference/getids.html">getids()</a></code>
function includes <em>at least one identifier</em> supported by both
platforms. However, if no single identifier meets this requirement,
<em>two separate identifiers</em> may be used, ensuring that one is
valid for STRING and the other for g:Profiler.</span></p>
<p>For example, in the case of <em>Danio rerio</em> (Zebrafish), the
<strong>Ensembl gene ID</strong> is supported by both platforms,
eliminating the need for additional identifier conversion. However, in
the case of <em>Daphnia magna</em> (Water flea), the <strong>Ensembl
gene ID</strong> is supported by g:Profiler but not by the STRING
database. In this case, the <strong>UniProt Protein identifier</strong>
is supported by STRING, so both identifiers should be retrieved to
ensure compatibility. </p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bg_t_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getids.html">getids</a></span><span class="op">(</span></span>
<span>  id_query <span class="op">=</span> <span class="va">bg_transcripts_list</span>, </span>
<span>  biomart_db <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>,</span>
<span>  species_dataset <span class="op">=</span> <span class="st">"drerio_gene_ensembl"</span>,</span>
<span>  transcript_id <span class="op">=</span> <span class="st">"ensembl_transcript_id_version"</span>,</span>
<span>  gene_id <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span>  gene_name <span class="op">=</span> <span class="st">"external_gene_name"</span>,</span>
<span>  other_id <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For more information on each argument of the function, run
<code>?cluefish::getids()</code>.</p>
<p><br></p>
<details><summary>
Understand <code>gene_id</code>, <code>gene_name</code> and
<code>other_id</code> usage
</summary><ul>
<li>
<strong><code>gene_id</code></strong>: This identifier is essential
for downstream steps, such as querying the protein-protein interaction
(PPI) network (<span id="step5">Step 5</span>) and performing functional
enrichment using g:profiler (<span id="step7">Step 7</span>). For
instance, with <em>Danio rerio</em> as our organism, the
<code>ensembl_gene_id</code> is supported by both the STRING and
g:profiler database, minimising the need for additional ID conversions
and ensuring compatibility across resources.</li>
<li>
<strong><code>gene_name</code></strong>: This identifier enhances
biological interpretability by providing descriptive labels, such as the
<code>external_gene_name</code> or <code>description</code>. These
labels offer intuitive gene names that clarify gene functions or roles
in biological processes, aiding in result interpretation and
annotation.</li>
<li>
<strong><code>other_id</code></strong>: Additional identifiers can
be rerquested if there are other annotation needs, such as protein IDs
or alternative identifiers related to specific transcripts or gene
annotations.</li>
</ul></details><p><br></p>
<details><summary>
Effort to remove redundancy between <code>transcript_id</code>,
<code>gene_id</code> and <code>gene_name</code>
</summary><p>In many cases, a single gene can have multiple transcripts, and the
same gene symbol (like <code>external_gene_name</code>) may correspond
to different Ensembl gene IDs, creating redundancy in the data. The
<span class="homemade-function"><code><a href="../reference/getids.html">getids()</a></code></span> function
handles this redundancy by:</p>
<ol style="list-style-type: decimal">
<li>Checking if the query includes the “<code>external_gene_name</code>”
attribute in the <code>gene_name</code> argument.</li>
<li>Identifying any duplicate <code>transcript_id</code> and/or
<code>gene_id</code> entries across rows.</li>
</ol>
<p>When duplicates are found, the function assigns unique names to
<code>gene_name</code> using a <strong><code>_g#t#</code></strong>
format:</p>
<ul>
<li>
<strong><code>g#</code></strong> represents the gene index</li>
<li>
<strong><code>t#</code></strong> represents the transcript
index</li>
</ul>
<p><u>For example</u>:</p>
<ul>
<li>If a single <code>gene_id</code> has multiple transcripts with the
same <code>gene_name</code> (like “<code>agrn</code>”), the function
labels them as “<strong><code>agrn_g1t1</code></strong>”,
“<strong><code>agrn_g1t2</code></strong>”, and so on. This indicates
different transcripts for the same gene.</li>
<li>If another <code>gene_id</code> with the same <code>gene_name</code>
appears, it gets a new gene index, resulting in labels like
“<strong><code>agrn_g2t1</code></strong>, indicating it belongs to a
second gene.</li>
</ul>
<p>If no <code>gene_name</code> is found for certain transcripts, the
function assigns “<code>Unknown</code>” and applies the same
<strong><code>_g#t#</code></strong> numbering (e.g.,
<strong><code>Unknown_g1t1</code></strong>,
<strong><code>Unknown_g1t2</code></strong>).</p>
<p>This systematic labeling helps make each transcript-gene relationship
unique, aiding in clear interpretation and visualisation of results.</p>
</details><p><br></p>
<p>We can get a look at the structure of the output:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">bg_t_ids</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 39,890</span></span>
<span><span class="co">#&gt; Columns: 3</span></span>
<span><span class="co">#&gt; $ transcript_id <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSDART00000000678.7", "ENSDART00000000877.7", "ENSDART…</span></span>
<span><span class="co">#&gt; $ gene_id       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSDARG00000000638", "ENSDARG00000044279", "ENSDARG0000…</span></span>
<span><span class="co">#&gt; $ gene_name     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "opn1mw4", "opn1mw3", "psmb13a", "unknown.1", "drd2l_t1"…</span></span></code></pre></div>
<p>As the result consists of the background list identifiers, you need
to subset the deregulated data. This can be done as shown below:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dr_t_ids</span> <span class="op">&lt;-</span> <span class="va">bg_t_ids</span><span class="op">[</span><span class="va">bg_t_ids</span><span class="op">$</span><span class="va">transcript_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">dr_transcript_list</span>,<span class="op">]</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step4">Step 4: Determine deregulated gene regulatory status<a class="anchor" aria-label="anchor" href="#step4"></a>
</h3>
<p>This step involves determining the regulatory status of deregulated
genes, using the TF and CoTF files downloaded in <a href="#step1">Step
1</a>. if your organism was not supported in AnimalTFDB, you can skip
this step!</p>
<p>If the TF and CoTF files are available, use the <span class="homemade-function"><code><a href="../reference/getregs.html">getregs()</a></code></span> function to add
a regulatory status column (<code>TF</code>). This column will contain a
logical value (<code>TRUE/FALSE</code>), indicating whether each gene is
a TF or CoTF.</p>
<p>The function requires the output of the previous
<code><a href="../reference/getids.html">getids()</a></code> function and the file paths for both the TF and
CoTF files.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dr_t_regs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getregs.html">getregs</a></span><span class="op">(</span></span>
<span>  getids_data <span class="op">=</span> <span class="va">dr_t_ids</span>,</span>
<span>  regulator_file <span class="op">=</span> <span class="st">"data/example_TF_file.txt"</span>,</span>
<span>  coregulator_file <span class="op">=</span> <span class="st">"data/example_CoTF_file.txt"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For more information on each argument of the function, run
<code>?cluefish::getregs()</code>.</p>
<p>We can look at the structure of the output:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">dr_t_regs</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 2,433</span></span>
<span><span class="co">#&gt; Columns: 4</span></span>
<span><span class="co">#&gt; $ transcript_id <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSDART00000135999.2", "ENSDART00000173686.2", "ENSDART…</span></span>
<span><span class="co">#&gt; $ gene_id       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSDARG00000092550", "ENSDARG00000033683", "ENSDARG0000…</span></span>
<span><span class="co">#&gt; $ gene_name     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "unknown.4894", "tpma_t3", "adgrl4_t2", "rnf144aa", "SLC…</span></span>
<span><span class="co">#&gt; $ TF            <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, …</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3 tabset">
<h3 class="tabset" id="step5">Step 5: Construct and retrieve the clustered
protein-protein interaction (PPI) network<a class="anchor" aria-label="anchor" href="#step5"></a>
</h3>
<p>This step involves creating a clustered protein-protein interaction
(PPI) network using Cytoscape, an external visualisation tool. Start by
saving the R output to a file:</p>

<p>Once the file is saved, open Cytoscape and follow the steps below to
import, create, and cluster the PPI network, then return to the R
environment to continue the analysis.</p>










<ul class="nav nav-tabs" id="step5" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">dr_t_regs</span>, file <span class="op">=</span> <span class="st">"your/chosen/path/filename.txt"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span></code></button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#start" id="start-tab" type="button" role="tab" aria-controls="start" aria-selected="false" class="nav-link">Start</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#section" id="section-tab" type="button" role="tab" aria-controls="section" aria-selected="false" class="nav-link">1</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#section-1" id="section-1-tab" type="button" role="tab" aria-controls="section-1" aria-selected="false" class="nav-link">2</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#section-2" id="section-2-tab" type="button" role="tab" aria-controls="section-2" aria-selected="false" class="nav-link">3</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#section-3" id="section-3-tab" type="button" role="tab" aria-controls="section-3" aria-selected="false" class="nav-link">4</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#section-4" id="section-4-tab" type="button" role="tab" aria-controls="section-4" aria-selected="false" class="nav-link">5</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#section-5" id="section-5-tab" type="button" role="tab" aria-controls="section-5" aria-selected="false" class="nav-link">6</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#section-6" id="section-6-tab" type="button" role="tab" aria-controls="section-6" aria-selected="false" class="nav-link">7</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#section-7" id="section-7-tab" type="button" role="tab" aria-controls="section-7" aria-selected="false" class="nav-link">8</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#section-8" id="section-8-tab" type="button" role="tab" aria-controls="section-8" aria-selected="false" class="nav-link">9</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab"></div>
<div class="tab-pane" id="start" role="tabpanel" aria-labelledby="start-tab">

<center>
<div class="float">
<img src="figures%2Fcytoscape-step0.png" style="width:75.0%" alt="Home page of the Cytoscape App"><div class="figcaption">Home page of the Cytoscape App</div>
</div>
</center>
<p><br></p>
</div>
<div class="tab-pane" id="section" role="tabpanel" aria-labelledby="section-tab">

<p>In Cytoscape, go to <code>File</code> &gt; <code>Import</code> &gt;
<code>Table from File</code></p>
<center>
<div class="float">
<img src="figures%2Fcytoscape-step1.png" style="width:50.0%" alt="File to Import"><div class="figcaption">File to Import</div>
</div>
</center>
<p><br></p>
</div>
<div class="tab-pane" id="section-1" role="tabpanel" aria-labelledby="section-1-tab">

<p>Select the saved text file, click <code>Open</code>, name the table
and select <code>OK</code></p>
<center>
<div class="float">
<img src="figures%2Fcytoscape-step2.png" style="width:65.0%" alt="Name and confirm import"><div class="figcaption">Name and confirm import</div>
</div>
</center>
<p><br></p>
</div>
<div class="tab-pane" id="section-2" role="tabpanel" aria-labelledby="section-2-tab">

<p>Right-click the column with ids for querying proteins (e.g.,
<code>gene_id</code>) (1), then select <code>Copy Column Values</code>
(2)</p>
<center>
<div class="float">
<img src="figures%2Fcytoscape-step3.png" style="width:65.0%" alt="Copy gene id column values"><div class="figcaption">Copy gene id column values</div>
</div>
</center>
<p><br></p>
</div>
<div class="tab-pane" id="section-3" role="tabpanel" aria-labelledby="section-3-tab">

<p>Ensure you are using the <em>Protein Query</em> feature with
StringApp.</p>
<center>
<div class="float">
<img src="figures%2Fcytoscape-step4.png" style="width:65.0%" alt="Check “Protein Query”"><div class="figcaption">Check “Protein Query”</div>
</div>
</center>
<p><br></p>
</div>
<div class="tab-pane" id="section-4" role="tabpanel" aria-labelledby="section-4-tab">

<p>In the StringApp query section, paste
(<code>Ctrl</code>+<code>V</code>) the copied values</p>
<center>
<div class="float">
<img src="figures%2Fcytoscape-step5.png" style="width:65.0%" alt="Paste values into query"><div class="figcaption">Paste values into query</div>
</div>
</center>
<p><br></p>
</div>
<div class="tab-pane" id="section-5" role="tabpanel" aria-labelledby="section-5-tab">

<p>Adjust the query settings:</p>
<ol style="list-style-type: decimal">
<li>Click on the “<strong>More options…</strong>” icon</li>
<li>Select the species (e.g., <em>Danio rerio</em>)</li>
<li>Select the network type<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;The choice is either “&lt;strong&gt;full STRING
network&lt;/strong&gt;”, meaning both direct (physical) and indirect
(functional) interactions or “&lt;strong&gt;physical network&lt;/strong&gt;” for
just direct interactions &lt;span class="citation"&gt;(Szklarczyk et al.
2019)&lt;/span&gt;.&lt;/p&gt;'><sup>2</sup></a>
</li>
<li>Select a confidence score cutoff<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;All scores rank from 0 to 1, with 1 being the highest
possible confidence. A score of 0.5 would indicate that roughly every
second interaction might be erroneous (i.e., a false positive) (&lt;a href="https://string-db.org/cgi/info" class="external-link"&gt;See STRING info&lt;/a&gt;).&lt;/p&gt;'><sup>3</sup></a>
</li>
<li>Make sure that maximum additional interactors is at “0”!<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Adding interactors meaning adding supplementary
information than what we give STRING. This does not make sense for our
analysis as we are focusing solely on the deregulated items.&lt;/p&gt;"><sup>4</sup></a>
</li>
</ol>
<center>
<p>Visit the <a href="https://string-db.org/" class="external-link">STRING documentation</a>
to getting a fuller understand of these concepts.</p>
<div class="float">
<img src="figures%2Fcytoscape-step6.png" style="width:70.0%" alt="Modify the query options"><div class="figcaption">Modify the query options</div>
</div>
</center>
<p><br></p>
</div>
<div class="tab-pane" id="section-6" role="tabpanel" aria-labelledby="section-6-tab">

<p>Run the query (1), view the PPI network, then click on
“<strong>Clustering</strong>” (2)</p>
<center>
<div class="float">
<img src="figures%2Fcytoscape-step7.png" style="width:80.0%" alt="Resulting PPI network"><div class="figcaption">Resulting PPI network</div>
</div>
</center>
<p><br></p>
</div>
<div class="tab-pane" id="section-7" role="tabpanel" aria-labelledby="section-7-tab">

<p>Choose a granularity (inflation) value to adjust cluster sizes<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Inflation affects the granularity or resolution of the
clustering outcome, with low values (1.3, 1.4) leading to fewer and
larger clusters and high values (5, 6) leading to more and smaller
clusters; the default value of 2 is a good starting point &lt;span class="citation"&gt;(Van Dongen 2008)&lt;/span&gt;. To get a better understanding
of MCL clustering, visit &lt;a href="http://micans.org/mcl/" class="external-link uri"&gt;http://micans.org/mcl/&lt;/a&gt;.&lt;/p&gt;'><sup>5</sup></a>, then
click “<strong>OK</strong>”</p>
<center>
<div class="float">
<img src="figures%2Fcytoscape-step8.png" style="width:65.0%" alt="Options for MCL Clustering"><div class="figcaption">Options for MCL Clustering</div>
</div>
</center>
<p><br></p>
</div>
<div class="tab-pane" id="section-8" role="tabpanel" aria-labelledby="section-8-tab">

<p>Open the <strong><code>Node Table</code></strong> (1) and click on
the <code>Export Tables to File...</code> icon (2). Finally, save the
table in .csv format within your working directory.</p>
<center>
<p>With the node table saved, we can now return to the R environment and
import the table.</p>
<div class="float">
<img src="figures%2Fcytoscape-step9.png" style="width:75.0%" alt="Selection and exporting the Node Table"><div class="figcaption">Selection and exporting the Node Table</div>
</div>
</center>
</div>
</div>
</div>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="section-9"></h3>
<p>Using the <span class="homemade-function"><code><a href="../reference/getclustrs.html">getclustrs()</a></code></span> function,
retrieve and merge the node table with the output of the <span class="homemade-function"><code><a href="../reference/getregs.html">getregs()</a></code></span> function.
Therefor it requires the output from the function used in <a href="#step3">Step 3</a>, the column name to use for merging and the
nodetable path and filename, as shown below :</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dr_t_clustrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getclustrs.html">getclustrs</a></span><span class="op">(</span></span>
<span>  gene_data <span class="op">=</span> <span class="va">dr_t_regs</span>,</span>
<span>  colname_for_merge <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>  path <span class="op">=</span> <span class="st">"your/chosen/path"</span>,</span>
<span>  nodetable_filename <span class="op">=</span> <span class="st">"filename.csv"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For more information on each argument of the function, run
<code>?cluefish::getclustrs()</code>.</p>
<p>We can look at the structure of this output:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">dr_t_clustrs</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 775</span></span>
<span><span class="co">#&gt; Columns: 5</span></span>
<span><span class="co">#&gt; $ gene_id       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSDARG00000000068", "ENSDARG00000000086", "ENSDARG0000…</span></span>
<span><span class="co">#&gt; $ transcript_id <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSDART00000000069.8", "ENSDART00000132378.4", "ENSDART…</span></span>
<span><span class="co">#&gt; $ gene_name     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "nherf1a_g2t1", "itsn1", "pms1_t2", "asnsd1_t2", "unknow…</span></span>
<span><span class="co">#&gt; $ TF            <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> FALSE, FALSE, TRUE, FALSE, FALSE, FALSE, FALSE, FALSE, F…</span></span>
<span><span class="co">#&gt; $ clustr        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 139, 140, 27, 172, 159, 56, 47, 36, 187, 164, 144, 52, 6…</span></span></code></pre></div>
<p>We can also get an idea of the size of the clusters:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">dr_t_clustrs</span><span class="op">$</span><span class="va">clustr</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20 </span></span>
<span><span class="co">#&gt;  41  35  22  18  17  14  12  12   8   8   7   7   7   8   6   6   6   6   7   6 </span></span>
<span><span class="co">#&gt;  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36  37  38  39  40 </span></span>
<span><span class="co">#&gt;   6   6   5   5   6   5   5   5   7   5   5   5   5   5   5   5   5   5   4   4 </span></span>
<span><span class="co">#&gt;  41  42  43  44  45  46  47  48  49  50  51  52  53  54  55  56  57  58  59  60 </span></span>
<span><span class="co">#&gt;   5   5   4   4   4   4   4   4   5   4   4   4   4   3   3   4   3   3   3   3 </span></span>
<span><span class="co">#&gt;  61  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76  77  78  79  80 </span></span>
<span><span class="co">#&gt;   3   3   3   3   3   5   3   3   3   3   3   3   3   3   3   3   3   3   3   3 </span></span>
<span><span class="co">#&gt;  81  82  83  84  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99 100 </span></span>
<span><span class="co">#&gt;   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3 </span></span>
<span><span class="co">#&gt; 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 </span></span>
<span><span class="co">#&gt;   3   3   2   2   3   3   2   2   3   2   4   2   2   2   2   2   2   2   2   2 </span></span>
<span><span class="co">#&gt; 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 </span></span>
<span><span class="co">#&gt;   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   3 </span></span>
<span><span class="co">#&gt; 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 </span></span>
<span><span class="co">#&gt;   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2 </span></span>
<span><span class="co">#&gt; 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 </span></span>
<span><span class="co">#&gt;   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2 </span></span>
<span><span class="co">#&gt; 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 </span></span>
<span><span class="co">#&gt;   2   2   2   3   2   2   2   2   2   2   3   2   2   2   3   2   3   2   2   2 </span></span>
<span><span class="co">#&gt; 201 202 203 204 </span></span>
<span><span class="co">#&gt;   2   2   2   2</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step6">Step 6: Filter clusters<a class="anchor" aria-label="anchor" href="#step6"></a>
</h3>
<p>In this step, we filter clusters based on their gene set sizes using
the <strong>lower cluster size filter</strong>. This filter retains only
clusters meeting or exceeding a specified minimum size, helping to focus
on clusters that are likely biologically relevant.</p>
<p>The function <span class="homemade-function"><code><a href="../reference/clustrfiltr.html">clustrfiltr()</a></code></span> applies this
filter, requiring two main arguments:</p>
<ul>
<li>
<code>getclustrs_data</code>: the output from the <span class="homemade-function"><code><a href="../reference/getclustrs.html">getclustrs()</a></code></span> function</li>
<li>the desired <strong>lower cluster size filter</strong> value.</li>
</ul>
<details><summary>
Choosing an appropriate lower cluster size limit
</summary><p>The choice of size limit depends on factors such as the type of
study, the organism, and prior steps in the PPI network creation and
clustering. Setting a higher size limit will retain fewer, larger
clusters, while a lower limit will include both large and small
clusters. Note that:</p>
<ul>
<li>
<strong>Large</strong> clusters may lack specific biological context
due to their broad inclusivity.</li>
<li>
<strong>Small</strong> clusters may suffer from limited statistical
support and might not represent distinct biological complexes. Instead,
they could reflect case-specific or transient interactions.</li>
</ul>
<p>Consider these factors when deciding on an appropriate size
limit.</p>
</details><div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dr_t_clustrs_filtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustrfiltr.html">clustrfiltr</a></span><span class="op">(</span></span>
<span>  getclustrs_data <span class="op">=</span> <span class="va">dr_t_clustrs</span>,</span>
<span>  size_filtr <span class="op">=</span> <span class="fl">4</span> <span class="co"># default value</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Total clusters kept: 53 / 204</span></span></code></pre></div>
<p>For more information on each argument of the function, run
<code>?cluefish::clustrfiltr()</code>.</p>
<p>The function outputs a named list with three components:</p>
<ul>
<li>
<strong><code>kept</code></strong>: A dataframe of type
<strong>t</strong> holding clusters retained after filtering</li>
<li>
<strong><code>removed</code></strong>: A dataframe of type
<strong>t</strong> holding clusters filtered out</li>
<li>
<strong><code>params</code></strong>: A list of the main parameters
used; in this case the choice for <code>size_filtr</code>
</li>
</ul>
<p>We can take a look at the output’s structure:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">dr_t_clustrs_filtr</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 3</span></span>
<span><span class="co">#&gt;  $ kept   :'data.frame': 411 obs. of  5 variables:</span></span>
<span><span class="co">#&gt;   ..$ gene_id      : chr [1:411] "ENSDARG00000003564" "ENSDARG00000003599" "ENSDARG00000005791" "ENSDARG00000006413" ...</span></span>
<span><span class="co">#&gt;   ..$ transcript_id: chr [1:411] "ENSDART00000018515.5" "ENSDART00000011251.10" "ENSDART00000175435.2" "ENSDART00000021069.8" ...</span></span>
<span><span class="co">#&gt;   ..$ gene_name    : chr [1:411] "dohh" "rpl3_t2" "rpl28_t2" "rpl38" ...</span></span>
<span><span class="co">#&gt;   ..$ TF           : logi [1:411] FALSE FALSE FALSE FALSE FALSE FALSE ...</span></span>
<span><span class="co">#&gt;   ..$ clustr       : int [1:411] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;  $ removed:'data.frame': 364 obs. of  5 variables:</span></span>
<span><span class="co">#&gt;   ..$ gene_id      : chr [1:364] "ENSDARG00000023152" "ENSDARG00000041155" "ENSDARG00000101216" "ENSDARG00000077822" ...</span></span>
<span><span class="co">#&gt;   ..$ transcript_id: chr [1:364] "ENSDART00000030984.4" "ENSDART00000139441.2" "ENSDART00000172204.2" "ENSDART00000113752.3" ...</span></span>
<span><span class="co">#&gt;   ..$ gene_name    : chr [1:364] "wdr73_g2t1" "morf4l1" "meaf6_t2" "si:dkey-6i22.5" ...</span></span>
<span><span class="co">#&gt;   ..$ TF           : logi [1:364] FALSE TRUE TRUE FALSE FALSE FALSE ...</span></span>
<span><span class="co">#&gt;   ..$ clustr       : int [1:364] 54 54 54 55 55 55 56 56 56 56 ...</span></span>
<span><span class="co">#&gt;  $ params :List of 1</span></span>
<span><span class="co">#&gt;   ..$ size_filtr: num 4</span></span>
<span><span class="co">#&gt;  - attr(*, "class")= chr "clustrfiltres"</span></span></code></pre></div>
<p>We can also see which clusters were retained:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">dr_t_clustrs_filtr</span><span class="op">$</span><span class="va">kept</span><span class="op">$</span><span class="va">clustr</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 </span></span>
<span><span class="co">#&gt; 41 35 22 18 17 14 12 12  8  8  7  7  7  8  6  6  6  6  7  6  6  6  5  5  6  5 </span></span>
<span><span class="co">#&gt; 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 </span></span>
<span><span class="co">#&gt;  5  5  7  5  5  5  5  5  5  5  5  5  4  4  5  5  4  4  4  4  4  4  5  4  4  4 </span></span>
<span><span class="co">#&gt; 53 </span></span>
<span><span class="co">#&gt;  4</span></span></code></pre></div>
<hr>
<p><span class="italic-grey">Genes that are not retained in this step
have now become <em>lonely genes</em>.</span></p>
<hr>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step7">Step 7: Perform cluster-wise functional enrichment<a class="anchor" aria-label="anchor" href="#step7"></a>
</h3>
<p>This step applies Over-Representation Analysis (ORA) to characterise
each cluster using the <code>gprofiler2</code> R package <span class="citation">(Kolberg et al. 2020)</span>, which interfaces with
g:Profiler’s <span class="citation">(Kolberg et al. 2023)</span>
ressources and functions. The g:GOst function performs ORA across
multiple gene lists, analysing against sources such as GO, KEGG, and WP.
For more information on g:Profiler and g:GOst, visit the web interface
<a href="https://biit.cs.ut.ee/gprofiler/gost" class="external-link">here</a>.</p>
<p>The <span class="homemade-function"><code><a href="../reference/clustrenrich.html">clustrenrich()</a></code></span> function,
built on g:GOst, performs cluster-wise ORA.</p>
<p>The function applies g:GOst’s highlighting GO driver terms by
considering the underlying topology of annotations, as described in
Kolberg et al. 2023.</p>
<p>Consider the following g:GOst key options:</p>
<ul>
<li>
<code>sources</code>: Functional data sources to analyse against
(e.g. “GO”, “KEGG”)</li>
<li>
<code>domain_scope</code>: Background choice (e.g. “annotated”,
“known”, “custom”)</li>
<li>
<code>correction_method</code>: Multiple testing correction method,
e.g., FDR</li>
<li>
<code>user_threshold</code>: P-value threshold for significance,
e.g. 0.05</li>
</ul>
<p>These are a few key options, but many others are available. Some
require careful thought to select the best value, method, or scope. For
more details, check the documentation on their <a href="https://biit.cs.ut.ee/gprofiler/gost" class="external-link">site</a> or reading the help
of the function by running <code>?gprofiler2::gost()</code> in R.</p>
<p>Additionally, two new filters have been integrated with the
construction of the <span class="homemade-function"><code><a href="../reference/clustrenrich.html">clustrenrich()</a></code></span>
function:</p>
<ul>
<li>
<p><strong>Biological function size filter</strong>
(<code>min_term_size</code> and <code>max_term_size</code>): Filter
based on upper and lower gene set size limits to control for biological
function specificity</p>
<hr>
<p><span class="italic-grey">Very large gene sets, which are dependent
on the organism, are often associated with broad biological processes.
These can introduce noise and obscure more specific functions.
Conversely, very small gene sets can be overly specific and may not
capture the full biological context of a pathway. </span></p>
<hr>
</li>
<li>
<p><strong>Enrichment gene count filter</strong>
(<code>ngenes_enrich_filtr</code>): Filters based on the number of genes
contributing to enrichment of a cluster.</p>
<hr>
<p><span class="italic-grey">Small clusters can lead to a higher rate of
false positives, and enrichments based on very few genes may not provide
substantial biological insight into the cluster itself. </span></p>
<hr>
</li>
</ul>
<p>The function data inputs include:</p>
<ul>
<li>
<strong><code>clustrfiltr_data</code></strong>: The named list
output from the <span class="homemade-function"><code><a href="../reference/clustrfiltr.html">clustrfiltr()</a></code></span>
function</li>
<li>
<strong><code>dr_genes</code></strong>: The character vector of
deregulated genes that can correspond to the
<strong><code>gene_id</code></strong> column in the output of the <span class="homemade-function"><code><a href="../reference/getids.html">getids()</a></code></span> or <span class="homemade-function"><code><a href="../reference/getregs.html">getregs()</a></code></span> function.</li>
<li>
<strong><code>bg_genes</code></strong>: The character vector of
background genes (preferably from the experiment) that typically
corresponds to the ‘gene_id’ column in the output of the <span class="homemade-function"><code><a href="../reference/getids.html">getids()</a></code></span> function.</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clustr_enrichres</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustrenrich.html">clustrenrich</a></span><span class="op">(</span></span>
<span>  clustrfiltr_data <span class="op">=</span> <span class="va">dr_t_clustrs_filtr</span>,</span>
<span>  dr_genes <span class="op">=</span> <span class="va">dr_t_regs</span><span class="op">$</span><span class="va">gene_id</span>,</span>
<span>  bg_genes <span class="op">=</span> <span class="va">bg_t_ids</span><span class="op">$</span><span class="va">gene_id</span>,</span>
<span>  bg_type <span class="op">=</span> <span class="st">"custom_annotated"</span>,</span>
<span>  sources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:BP"</span>, <span class="st">"KEGG"</span>, <span class="st">"WP"</span><span class="op">)</span>, </span>
<span>  organism <span class="op">=</span> <span class="st">"drerio"</span>,</span>
<span>  user_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  correction_method <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>  exclude_iea <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  min_term_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  max_term_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  only_highlighted_GO <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ngenes_enrich_filtr <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  path <span class="op">=</span> <span class="st">"your/chosen/path"</span>,</span>
<span>  output_filename <span class="op">=</span> <span class="st">"filename.rds"</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For more information on each argument of the function, run
<code>?cluefish::clustrenrich()</code>.</p>
<p>The output is a named list with four components:</p>
<ul>
<li>
<strong><code>dr_g_a_enrich</code></strong>: A dataframe of type
<strong>g_a</strong> holding the cluster-wise enrichment results</li>
<li>
<strong><code>gostres</code></strong>: A named list corresponding to
the original result output of g:GOst</li>
<li>
<strong><code>dr_g_a_whole</code></strong>: A dataframe of type
<strong>g_a</strong> holding all the biological function annotations
found in the g:profiler database for all the deregulated genes.</li>
<li>
<strong><code>c_simplify_log</code></strong>: A dataframe of tracing
the number of biological functions enriched per cluster before and after
each filtering step for each source.</li>
<li>
<strong><code>params</code></strong>: A list of the main parameters
used; in this case the choice for <code>monoterm_fusion</code>
</li>
</ul>
<p>We can take a look at the structure of the
<strong><code>dr_g_a_enrich</code></strong> dataframe:</p>
<pre><code><span><span class="co">#&gt; Rows: 754</span></span>
<span><span class="co">#&gt; Columns: 5</span></span>
<span><span class="co">#&gt; $ gene_id   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSDARG00000051783", "ENSDARG00000009285", "ENSDARG00000020…</span></span>
<span><span class="co">#&gt; $ clustr    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, …</span></span>
<span><span class="co">#&gt; $ term_name <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Cytoplasmic ribosomal proteins", "Cytoplasmic ribosomal pro…</span></span>
<span><span class="co">#&gt; $ term_id   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "WP:WP324", "WP:WP324", "WP:WP324", "WP:WP324", "WP:WP324", …</span></span>
<span><span class="co">#&gt; $ source    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "WP", "WP", "WP", "WP", "WP", "WP", "WP", "WP", "WP", "WP", …</span></span></code></pre>
<p>After running the function, we recommend checking if the filters need
adjustment, as finding the right lower and upper limits for gene set
sizes can be quite a challenge. These limits depend on the organism, the
type of transcriptomic data, and what constitutes ‘general’ terms for
this study.</p>
<p>To help determine if the size limits are suitable, the following code
lists all terms associated with deregulated genes in descending order by
size, giving insight into the generality of terms and guiding any needed
adjustments:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clustr_enrichres</span><span class="op">$</span><span class="va">dr_g_a_whole</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">term_name</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>count <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="co"># Number of rows to print, adjustable based on the study</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3,646 × 2</span></span></span>
<span><span class="co">#&gt;    term_name                           count</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                               <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> biological_process                   <span style="text-decoration: underline;">1</span>509</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> cellular process                     <span style="text-decoration: underline;">1</span>299</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> metabolic process                     903</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> organic substance metabolic process   855</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> primary metabolic process             832</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> nitrogen compound metabolic process   776</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> cellular metabolic process            771</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> macromolecule metabolic process       718</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> KEGG root term                        707</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> biological regulation                 556</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3,636 more rows</span></span></span></code></pre></div>
<p>If you have the <code>DT</code> package installed, you can view this
data interactively in the viewer pane:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clustr_enrichres</span><span class="op">$</span><span class="va">dr_g_a_whole</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">term_name</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>count <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span></span>
<span>    options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pageLength <span class="op">=</span> <span class="fl">10</span></span>
<span>    <span class="op">)</span>,</span>
<span>    filter <span class="op">=</span> <span class="st">'top'</span>,</span>
<span>    class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"compact"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"top","vertical":false,"filterHTML":"<tr>\n  <td><\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"integer\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none;position: absolute;width: 200px;opacity: 1\">\n      <div data-min=\"1\" data-max=\"1509\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n<\/tr>","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284","285","286","287","288","289","290","291","292","293","294","295","296","297","298","299","300","301","302","303","304","305","306","307","308","309","310","311","312","313","314","315","316","317","318","319","320","321","322","323","324","325","326","327","328","329","330","331","332","333","334","335","336","337","338","339","340","341","342","343","344","345","346","347","348","349","350","351","352","353","354","355","356","357","358","359","360","361","362","363","364","365","366","367","368","369","370","371","372","373","374","375","376","377","378","379","380","381","382","383","384","385","386","387","388","389","390","391","392","393","394","395","396","397","398","399","400","401","402","403","404","405","406","407","408","409","410","411","412","413","414","415","416","417","418","419","420","421","422","423","424","425","426","427","428","429","430","431","432","433","434","435","436","437","438","439","440","441","442","443","444","445","446","447","448","449","450","451","452","453","454","455","456","457","458","459","460","461","462","463","464","465","466","467","468","469","470","471","472","473","474","475","476","477","478","479","480","481","482","483","484","485","486","487","488","489","490","491","492","493","494","495","496","497","498","499","500","501","502","503","504","505","506","507","508","509","510","511","512","513","514","515","516","517","518","519","520","521","522","523","524","525","526","527","528","529","530","531","532","533","534","535","536","537","538","539","540","541","542","543","544","545","546","547","548","549","550","551","552","553","554","555","556","557","558","559","560","561","562","563","564","565","566","567","568","569","570","571","572","573","574","575","576","577","578","579","580","581","582","583","584","585","586","587","588","589","590","591","592","593","594","595","596","597","598","599","600","601","602","603","604","605","606","607","608","609","610","611","612","613","614","615","616","617","618","619","620","621","622","623","624","625","626","627","628","629","630","631","632","633","634","635","636","637","638","639","640","641","642","643","644","645","646","647","648","649","650","651","652","653","654","655","656","657","658","659","660","661","662","663","664","665","666","667","668","669","670","671","672","673","674","675","676","677","678","679","680","681","682","683","684","685","686","687","688","689","690","691","692","693","694","695","696","697","698","699","700","701","702","703","704","705","706","707","708","709","710","711","712","713","714","715","716","717","718","719","720","721","722","723","724","725","726","727","728","729","730","731","732","733","734","735","736","737","738","739","740","741","742","743","744","745","746","747","748","749","750","751","752","753","754","755","756","757","758","759","760","761","762","763","764","765","766","767","768","769","770","771","772","773","774","775","776","777","778","779","780","781","782","783","784","785","786","787","788","789","790","791","792","793","794","795","796","797","798","799","800","801","802","803","804","805","806","807","808","809","810","811","812","813","814","815","816","817","818","819","820","821","822","823","824","825","826","827","828","829","830","831","832","833","834","835","836","837","838","839","840","841","842","843","844","845","846","847","848","849","850","851","852","853","854","855","856","857","858","859","860","861","862","863","864","865","866","867","868","869","870","871","872","873","874","875","876","877","878","879","880","881","882","883","884","885","886","887","888","889","890","891","892","893","894","895","896","897","898","899","900","901","902","903","904","905","906","907","908","909","910","911","912","913","914","915","916","917","918","919","920","921","922","923","924","925","926","927","928","929","930","931","932","933","934","935","936","937","938","939","940","941","942","943","944","945","946","947","948","949","950","951","952","953","954","955","956","957","958","959","960","961","962","963","964","965","966","967","968","969","970","971","972","973","974","975","976","977","978","979","980","981","982","983","984","985","986","987","988","989","990","991","992","993","994","995","996","997","998","999","1000","1001","1002","1003","1004","1005","1006","1007","1008","1009","1010","1011","1012","1013","1014","1015","1016","1017","1018","1019","1020","1021","1022","1023","1024","1025","1026","1027","1028","1029","1030","1031","1032","1033","1034","1035","1036","1037","1038","1039","1040","1041","1042","1043","1044","1045","1046","1047","1048","1049","1050","1051","1052","1053","1054","1055","1056","1057","1058","1059","1060","1061","1062","1063","1064","1065","1066","1067","1068","1069","1070","1071","1072","1073","1074","1075","1076","1077","1078","1079","1080","1081","1082","1083","1084","1085","1086","1087","1088","1089","1090","1091","1092","1093","1094","1095","1096","1097","1098","1099","1100","1101","1102","1103","1104","1105","1106","1107","1108","1109","1110","1111","1112","1113","1114","1115","1116","1117","1118","1119","1120","1121","1122","1123","1124","1125","1126","1127","1128","1129","1130","1131","1132","1133","1134","1135","1136","1137","1138","1139","1140","1141","1142","1143","1144","1145","1146","1147","1148","1149","1150","1151","1152","1153","1154","1155","1156","1157","1158","1159","1160","1161","1162","1163","1164","1165","1166","1167","1168","1169","1170","1171","1172","1173","1174","1175","1176","1177","1178","1179","1180","1181","1182","1183","1184","1185","1186","1187","1188","1189","1190","1191","1192","1193","1194","1195","1196","1197","1198","1199","1200","1201","1202","1203","1204","1205","1206","1207","1208","1209","1210","1211","1212","1213","1214","1215","1216","1217","1218","1219","1220","1221","1222","1223","1224","1225","1226","1227","1228","1229","1230","1231","1232","1233","1234","1235","1236","1237","1238","1239","1240","1241","1242","1243","1244","1245","1246","1247","1248","1249","1250","1251","1252","1253","1254","1255","1256","1257","1258","1259","1260","1261","1262","1263","1264","1265","1266","1267","1268","1269","1270","1271","1272","1273","1274","1275","1276","1277","1278","1279","1280","1281","1282","1283","1284","1285","1286","1287","1288","1289","1290","1291","1292","1293","1294","1295","1296","1297","1298","1299","1300","1301","1302","1303","1304","1305","1306","1307","1308","1309","1310","1311","1312","1313","1314","1315","1316","1317","1318","1319","1320","1321","1322","1323","1324","1325","1326","1327","1328","1329","1330","1331","1332","1333","1334","1335","1336","1337","1338","1339","1340","1341","1342","1343","1344","1345","1346","1347","1348","1349","1350","1351","1352","1353","1354","1355","1356","1357","1358","1359","1360","1361","1362","1363","1364","1365","1366","1367","1368","1369","1370","1371","1372","1373","1374","1375","1376","1377","1378","1379","1380","1381","1382","1383","1384","1385","1386","1387","1388","1389","1390","1391","1392","1393","1394","1395","1396","1397","1398","1399","1400","1401","1402","1403","1404","1405","1406","1407","1408","1409","1410","1411","1412","1413","1414","1415","1416","1417","1418","1419","1420","1421","1422","1423","1424","1425","1426","1427","1428","1429","1430","1431","1432","1433","1434","1435","1436","1437","1438","1439","1440","1441","1442","1443","1444","1445","1446","1447","1448","1449","1450","1451","1452","1453","1454","1455","1456","1457","1458","1459","1460","1461","1462","1463","1464","1465","1466","1467","1468","1469","1470","1471","1472","1473","1474","1475","1476","1477","1478","1479","1480","1481","1482","1483","1484","1485","1486","1487","1488","1489","1490","1491","1492","1493","1494","1495","1496","1497","1498","1499","1500","1501","1502","1503","1504","1505","1506","1507","1508","1509","1510","1511","1512","1513","1514","1515","1516","1517","1518","1519","1520","1521","1522","1523","1524","1525","1526","1527","1528","1529","1530","1531","1532","1533","1534","1535","1536","1537","1538","1539","1540","1541","1542","1543","1544","1545","1546","1547","1548","1549","1550","1551","1552","1553","1554","1555","1556","1557","1558","1559","1560","1561","1562","1563","1564","1565","1566","1567","1568","1569","1570","1571","1572","1573","1574","1575","1576","1577","1578","1579","1580","1581","1582","1583","1584","1585","1586","1587","1588","1589","1590","1591","1592","1593","1594","1595","1596","1597","1598","1599","1600","1601","1602","1603","1604","1605","1606","1607","1608","1609","1610","1611","1612","1613","1614","1615","1616","1617","1618","1619","1620","1621","1622","1623","1624","1625","1626","1627","1628","1629","1630","1631","1632","1633","1634","1635","1636","1637","1638","1639","1640","1641","1642","1643","1644","1645","1646","1647","1648","1649","1650","1651","1652","1653","1654","1655","1656","1657","1658","1659","1660","1661","1662","1663","1664","1665","1666","1667","1668","1669","1670","1671","1672","1673","1674","1675","1676","1677","1678","1679","1680","1681","1682","1683","1684","1685","1686","1687","1688","1689","1690","1691","1692","1693","1694","1695","1696","1697","1698","1699","1700","1701","1702","1703","1704","1705","1706","1707","1708","1709","1710","1711","1712","1713","1714","1715","1716","1717","1718","1719","1720","1721","1722","1723","1724","1725","1726","1727","1728","1729","1730","1731","1732","1733","1734","1735","1736","1737","1738","1739","1740","1741","1742","1743","1744","1745","1746","1747","1748","1749","1750","1751","1752","1753","1754","1755","1756","1757","1758","1759","1760","1761","1762","1763","1764","1765","1766","1767","1768","1769","1770","1771","1772","1773","1774","1775","1776","1777","1778","1779","1780","1781","1782","1783","1784","1785","1786","1787","1788","1789","1790","1791","1792","1793","1794","1795","1796","1797","1798","1799","1800","1801","1802","1803","1804","1805","1806","1807","1808","1809","1810","1811","1812","1813","1814","1815","1816","1817","1818","1819","1820","1821","1822","1823","1824","1825","1826","1827","1828","1829","1830","1831","1832","1833","1834","1835","1836","1837","1838","1839","1840","1841","1842","1843","1844","1845","1846","1847","1848","1849","1850","1851","1852","1853","1854","1855","1856","1857","1858","1859","1860","1861","1862","1863","1864","1865","1866","1867","1868","1869","1870","1871","1872","1873","1874","1875","1876","1877","1878","1879","1880","1881","1882","1883","1884","1885","1886","1887","1888","1889","1890","1891","1892","1893","1894","1895","1896","1897","1898","1899","1900","1901","1902","1903","1904","1905","1906","1907","1908","1909","1910","1911","1912","1913","1914","1915","1916","1917","1918","1919","1920","1921","1922","1923","1924","1925","1926","1927","1928","1929","1930","1931","1932","1933","1934","1935","1936","1937","1938","1939","1940","1941","1942","1943","1944","1945","1946","1947","1948","1949","1950","1951","1952","1953","1954","1955","1956","1957","1958","1959","1960","1961","1962","1963","1964","1965","1966","1967","1968","1969","1970","1971","1972","1973","1974","1975","1976","1977","1978","1979","1980","1981","1982","1983","1984","1985","1986","1987","1988","1989","1990","1991","1992","1993","1994","1995","1996","1997","1998","1999","2000","2001","2002","2003","2004","2005","2006","2007","2008","2009","2010","2011","2012","2013","2014","2015","2016","2017","2018","2019","2020","2021","2022","2023","2024","2025","2026","2027","2028","2029","2030","2031","2032","2033","2034","2035","2036","2037","2038","2039","2040","2041","2042","2043","2044","2045","2046","2047","2048","2049","2050","2051","2052","2053","2054","2055","2056","2057","2058","2059","2060","2061","2062","2063","2064","2065","2066","2067","2068","2069","2070","2071","2072","2073","2074","2075","2076","2077","2078","2079","2080","2081","2082","2083","2084","2085","2086","2087","2088","2089","2090","2091","2092","2093","2094","2095","2096","2097","2098","2099","2100","2101","2102","2103","2104","2105","2106","2107","2108","2109","2110","2111","2112","2113","2114","2115","2116","2117","2118","2119","2120","2121","2122","2123","2124","2125","2126","2127","2128","2129","2130","2131","2132","2133","2134","2135","2136","2137","2138","2139","2140","2141","2142","2143","2144","2145","2146","2147","2148","2149","2150","2151","2152","2153","2154","2155","2156","2157","2158","2159","2160","2161","2162","2163","2164","2165","2166","2167","2168","2169","2170","2171","2172","2173","2174","2175","2176","2177","2178","2179","2180","2181","2182","2183","2184","2185","2186","2187","2188","2189","2190","2191","2192","2193","2194","2195","2196","2197","2198","2199","2200","2201","2202","2203","2204","2205","2206","2207","2208","2209","2210","2211","2212","2213","2214","2215","2216","2217","2218","2219","2220","2221","2222","2223","2224","2225","2226","2227","2228","2229","2230","2231","2232","2233","2234","2235","2236","2237","2238","2239","2240","2241","2242","2243","2244","2245","2246","2247","2248","2249","2250","2251","2252","2253","2254","2255","2256","2257","2258","2259","2260","2261","2262","2263","2264","2265","2266","2267","2268","2269","2270","2271","2272","2273","2274","2275","2276","2277","2278","2279","2280","2281","2282","2283","2284","2285","2286","2287","2288","2289","2290","2291","2292","2293","2294","2295","2296","2297","2298","2299","2300","2301","2302","2303","2304","2305","2306","2307","2308","2309","2310","2311","2312","2313","2314","2315","2316","2317","2318","2319","2320","2321","2322","2323","2324","2325","2326","2327","2328","2329","2330","2331","2332","2333","2334","2335","2336","2337","2338","2339","2340","2341","2342","2343","2344","2345","2346","2347","2348","2349","2350","2351","2352","2353","2354","2355","2356","2357","2358","2359","2360","2361","2362","2363","2364","2365","2366","2367","2368","2369","2370","2371","2372","2373","2374","2375","2376","2377","2378","2379","2380","2381","2382","2383","2384","2385","2386","2387","2388","2389","2390","2391","2392","2393","2394","2395","2396","2397","2398","2399","2400","2401","2402","2403","2404","2405","2406","2407","2408","2409","2410","2411","2412","2413","2414","2415","2416","2417","2418","2419","2420","2421","2422","2423","2424","2425","2426","2427","2428","2429","2430","2431","2432","2433","2434","2435","2436","2437","2438","2439","2440","2441","2442","2443","2444","2445","2446","2447","2448","2449","2450","2451","2452","2453","2454","2455","2456","2457","2458","2459","2460","2461","2462","2463","2464","2465","2466","2467","2468","2469","2470","2471","2472","2473","2474","2475","2476","2477","2478","2479","2480","2481","2482","2483","2484","2485","2486","2487","2488","2489","2490","2491","2492","2493","2494","2495","2496","2497","2498","2499","2500","2501","2502","2503","2504","2505","2506","2507","2508","2509","2510","2511","2512","2513","2514","2515","2516","2517","2518","2519","2520","2521","2522","2523","2524","2525","2526","2527","2528","2529","2530","2531","2532","2533","2534","2535","2536","2537","2538","2539","2540","2541","2542","2543","2544","2545","2546","2547","2548","2549","2550","2551","2552","2553","2554","2555","2556","2557","2558","2559","2560","2561","2562","2563","2564","2565","2566","2567","2568","2569","2570","2571","2572","2573","2574","2575","2576","2577","2578","2579","2580","2581","2582","2583","2584","2585","2586","2587","2588","2589","2590","2591","2592","2593","2594","2595","2596","2597","2598","2599","2600","2601","2602","2603","2604","2605","2606","2607","2608","2609","2610","2611","2612","2613","2614","2615","2616","2617","2618","2619","2620","2621","2622","2623","2624","2625","2626","2627","2628","2629","2630","2631","2632","2633","2634","2635","2636","2637","2638","2639","2640","2641","2642","2643","2644","2645","2646","2647","2648","2649","2650","2651","2652","2653","2654","2655","2656","2657","2658","2659","2660","2661","2662","2663","2664","2665","2666","2667","2668","2669","2670","2671","2672","2673","2674","2675","2676","2677","2678","2679","2680","2681","2682","2683","2684","2685","2686","2687","2688","2689","2690","2691","2692","2693","2694","2695","2696","2697","2698","2699","2700","2701","2702","2703","2704","2705","2706","2707","2708","2709","2710","2711","2712","2713","2714","2715","2716","2717","2718","2719","2720","2721","2722","2723","2724","2725","2726","2727","2728","2729","2730","2731","2732","2733","2734","2735","2736","2737","2738","2739","2740","2741","2742","2743","2744","2745","2746","2747","2748","2749","2750","2751","2752","2753","2754","2755","2756","2757","2758","2759","2760","2761","2762","2763","2764","2765","2766","2767","2768","2769","2770","2771","2772","2773","2774","2775","2776","2777","2778","2779","2780","2781","2782","2783","2784","2785","2786","2787","2788","2789","2790","2791","2792","2793","2794","2795","2796","2797","2798","2799","2800","2801","2802","2803","2804","2805","2806","2807","2808","2809","2810","2811","2812","2813","2814","2815","2816","2817","2818","2819","2820","2821","2822","2823","2824","2825","2826","2827","2828","2829","2830","2831","2832","2833","2834","2835","2836","2837","2838","2839","2840","2841","2842","2843","2844","2845","2846","2847","2848","2849","2850","2851","2852","2853","2854","2855","2856","2857","2858","2859","2860","2861","2862","2863","2864","2865","2866","2867","2868","2869","2870","2871","2872","2873","2874","2875","2876","2877","2878","2879","2880","2881","2882","2883","2884","2885","2886","2887","2888","2889","2890","2891","2892","2893","2894","2895","2896","2897","2898","2899","2900","2901","2902","2903","2904","2905","2906","2907","2908","2909","2910","2911","2912","2913","2914","2915","2916","2917","2918","2919","2920","2921","2922","2923","2924","2925","2926","2927","2928","2929","2930","2931","2932","2933","2934","2935","2936","2937","2938","2939","2940","2941","2942","2943","2944","2945","2946","2947","2948","2949","2950","2951","2952","2953","2954","2955","2956","2957","2958","2959","2960","2961","2962","2963","2964","2965","2966","2967","2968","2969","2970","2971","2972","2973","2974","2975","2976","2977","2978","2979","2980","2981","2982","2983","2984","2985","2986","2987","2988","2989","2990","2991","2992","2993","2994","2995","2996","2997","2998","2999","3000","3001","3002","3003","3004","3005","3006","3007","3008","3009","3010","3011","3012","3013","3014","3015","3016","3017","3018","3019","3020","3021","3022","3023","3024","3025","3026","3027","3028","3029","3030","3031","3032","3033","3034","3035","3036","3037","3038","3039","3040","3041","3042","3043","3044","3045","3046","3047","3048","3049","3050","3051","3052","3053","3054","3055","3056","3057","3058","3059","3060","3061","3062","3063","3064","3065","3066","3067","3068","3069","3070","3071","3072","3073","3074","3075","3076","3077","3078","3079","3080","3081","3082","3083","3084","3085","3086","3087","3088","3089","3090","3091","3092","3093","3094","3095","3096","3097","3098","3099","3100","3101","3102","3103","3104","3105","3106","3107","3108","3109","3110","3111","3112","3113","3114","3115","3116","3117","3118","3119","3120","3121","3122","3123","3124","3125","3126","3127","3128","3129","3130","3131","3132","3133","3134","3135","3136","3137","3138","3139","3140","3141","3142","3143","3144","3145","3146","3147","3148","3149","3150","3151","3152","3153","3154","3155","3156","3157","3158","3159","3160","3161","3162","3163","3164","3165","3166","3167","3168","3169","3170","3171","3172","3173","3174","3175","3176","3177","3178","3179","3180","3181","3182","3183","3184","3185","3186","3187","3188","3189","3190","3191","3192","3193","3194","3195","3196","3197","3198","3199","3200","3201","3202","3203","3204","3205","3206","3207","3208","3209","3210","3211","3212","3213","3214","3215","3216","3217","3218","3219","3220","3221","3222","3223","3224","3225","3226","3227","3228","3229","3230","3231","3232","3233","3234","3235","3236","3237","3238","3239","3240","3241","3242","3243","3244","3245","3246","3247","3248","3249","3250","3251","3252","3253","3254","3255","3256","3257","3258","3259","3260","3261","3262","3263","3264","3265","3266","3267","3268","3269","3270","3271","3272","3273","3274","3275","3276","3277","3278","3279","3280","3281","3282","3283","3284","3285","3286","3287","3288","3289","3290","3291","3292","3293","3294","3295","3296","3297","3298","3299","3300","3301","3302","3303","3304","3305","3306","3307","3308","3309","3310","3311","3312","3313","3314","3315","3316","3317","3318","3319","3320","3321","3322","3323","3324","3325","3326","3327","3328","3329","3330","3331","3332","3333","3334","3335","3336","3337","3338","3339","3340","3341","3342","3343","3344","3345","3346","3347","3348","3349","3350","3351","3352","3353","3354","3355","3356","3357","3358","3359","3360","3361","3362","3363","3364","3365","3366","3367","3368","3369","3370","3371","3372","3373","3374","3375","3376","3377","3378","3379","3380","3381","3382","3383","3384","3385","3386","3387","3388","3389","3390","3391","3392","3393","3394","3395","3396","3397","3398","3399","3400","3401","3402","3403","3404","3405","3406","3407","3408","3409","3410","3411","3412","3413","3414","3415","3416","3417","3418","3419","3420","3421","3422","3423","3424","3425","3426","3427","3428","3429","3430","3431","3432","3433","3434","3435","3436","3437","3438","3439","3440","3441","3442","3443","3444","3445","3446","3447","3448","3449","3450","3451","3452","3453","3454","3455","3456","3457","3458","3459","3460","3461","3462","3463","3464","3465","3466","3467","3468","3469","3470","3471","3472","3473","3474","3475","3476","3477","3478","3479","3480","3481","3482","3483","3484","3485","3486","3487","3488","3489","3490","3491","3492","3493","3494","3495","3496","3497","3498","3499","3500","3501","3502","3503","3504","3505","3506","3507","3508","3509","3510","3511","3512","3513","3514","3515","3516","3517","3518","3519","3520","3521","3522","3523","3524","3525","3526","3527","3528","3529","3530","3531","3532","3533","3534","3535","3536","3537","3538","3539","3540","3541","3542","3543","3544","3545","3546","3547","3548","3549","3550","3551","3552","3553","3554","3555","3556","3557","3558","3559","3560","3561","3562","3563","3564","3565","3566","3567","3568","3569","3570","3571","3572","3573","3574","3575","3576","3577","3578","3579","3580","3581","3582","3583","3584","3585","3586","3587","3588","3589","3590","3591","3592","3593","3594","3595","3596","3597","3598","3599","3600","3601","3602","3603","3604","3605","3606","3607","3608","3609","3610","3611","3612","3613","3614","3615","3616","3617","3618","3619","3620","3621","3622","3623","3624","3625","3626","3627","3628","3629","3630","3631","3632","3633","3634","3635","3636","3637","3638","3639","3640","3641","3642","3643","3644","3645","3646"],["'de novo' GDP-L-fucose biosynthetic process","'de novo' IMP biosynthetic process","'de novo' UMP biosynthetic process","'de novo' post-translational protein folding","'de novo' protein folding","'de novo' pyrimidine nucleobase biosynthetic process","10-formyltetrahydrofolate catabolic process","10-formyltetrahydrofolate metabolic process","2'-deoxyribonucleotide biosynthetic process","2'-deoxyribonucleotide metabolic process","2-Oxocarboxylic acid metabolism","3'-UTR-mediated mRNA destabilization","5-methylcytosine catabolic process","5-methylcytosine metabolic process","ABC transporters","ADP biosynthetic process","ADP catabolic process","ADP metabolic process","AMP biosynthetic process","AMP metabolic process","AMP salvage","ARF protein signal transduction","ATP biosynthetic process","ATP metabolic process","ATP synthesis coupled electron transport","ATP-dependent chromatin remodeling","Acetylcholine synthesis","Adherens junction","Adipocytokine signaling pathway","Adipogenesis","Adrenergic signaling in cardiomyocytes","Alanine, aspartate and glutamate metabolism","Amino sugar and nucleotide sugar metabolism","Aminoacyl-tRNA biosynthesis","Apelin signaling pathway","Apoptosis","Arachidonic acid metabolism","Arginine and proline metabolism","Arginine biosynthesis","Arp2/3 complex-mediated actin nucleation","Ascorbate and aldarate metabolism","Autophagy - animal","Autophagy - other","B cell activation","B cell differentiation","BMP signaling pathway","Basal transcription factors","Base excision repair","Biogenic amine synthesis","Biosynthesis of amino acids","Biosynthesis of cofactors","Biosynthesis of nucleotide sugars","Biosynthesis of unsaturated fatty acids","Butanoate metabolism","C-5 methylation of cytosine","C-type lectin receptor signaling pathway","C4-dicarboxylate transport","CDP-diacylglycerol biosynthetic process","CDP-diacylglycerol metabolic process","CENP-A containing chromatin assembly","COPII vesicle coating","COPII-coated vesicle budding","COPII-coated vesicle cargo loading","CRD motif and sFRPs influence on Wnt signaling","CRD-mediated mRNA stabilization","CTP biosynthetic process","CTP metabolic process","Calcium signaling pathway","Carbon metabolism","Cardiac muscle contraction","Cell adhesion molecules","Cell cycle","Cellular senescence","Cholesterol biosynthesis","Citrate cycle (TCA cycle)","Cysteine and methionine metabolism","Cytokine-cytokine receptor interaction","Cytoplasmic ribosomal proteins","Cytosolic DNA-sensing pathway","DNA alkylation","DNA biosynthetic process","DNA clamp unloading","DNA conformation change","DNA damage checkpoint signaling","DNA damage response","DNA dealkylation","DNA demethylation","DNA duplex unwinding","DNA geometric change","DNA integrity checkpoint signaling","DNA metabolic process","DNA methylation","DNA methylation on cytosine","DNA methylation on cytosine within a CG sequence","DNA methylation-dependent heterochromatin formation","DNA modification","DNA recombination","DNA repair","DNA replication","DNA replication checkpoint signaling","DNA replication initiation","DNA replication, Okazaki fragment processing","DNA replication, synthesis of RNA primer","DNA strand elongation","DNA strand elongation involved in DNA replication","DNA synthesis involved in DNA repair","DNA topological change","DNA-templated DNA replication","DNA-templated DNA replication maintenance of fidelity","DNA-templated transcription","DNA-templated transcription elongation","DNA-templated transcription initiation","DNA-templated transcription termination","Drug metabolism - cytochrome P450","Drug metabolism - other enzymes","ECM-receptor interaction","EPO receptor signaling","ER-nucleus signaling pathway","ERBB signaling pathway","ERK1 ERK2 MAPK cascade","ERK1 and ERK2 cascade","Effect of L carnitine on metabolism","Efferocytosis","Eicosanoid synthesis","Electron transport chain","Endocytosis","ErbB signaling pathway","Estrogen signaling","Ether lipid metabolism","Exercise induced circadian regulation","FGF signaling pathway","Fanconi anemia pathway","Fatty acid biosynthesis","Fatty acid degradation","Fatty acid elongation","Fatty acid metabolism","Ferroptosis","Focal adhesion","Folate biosynthesis","FoxO signaling pathway","Fructose and mannose metabolism","G protein-coupled acetylcholine receptor signaling pathway","G protein-coupled adenosine receptor signaling pathway","G protein-coupled purinergic receptor signaling pathway","G protein-coupled receptor signaling pathway","G protein-coupled receptor signaling pathway, coupled to cyclic nucleotide second messenger","G protein-coupled serotonin receptor signaling pathway","G1 to S cell cycle control","G1/S transition of mitotic cell cycle","G2/M transition of mitotic cell cycle","GDP-L-fucose biosynthetic process","GDP-L-fucose metabolic process","GDP-mannose metabolic process","GMP biosynthetic process","GMP metabolic process","GMP salvage","GPI anchor biosynthetic process","GPI anchor metabolic process","GTP biosynthetic process","GTP metabolic process","Galactose metabolism","Gap junction","Glutathione metabolism","Glycerolipid metabolism","Glycerophospholipid metabolism","Glycine, serine and threonine metabolism","Glycogen metabolism","Glycolysis / Gluconeogenesis","Glycolysis and gluconeogenesis","Glycosaminoglycan biosynthesis - chondroitin sulfate / dermatan sulfate","Glycosaminoglycan biosynthesis - heparan sulfate / heparin","Glycosaminoglycan degradation","Glycosphingolipid biosynthesis - ganglio series","Glycosphingolipid biosynthesis - globo and isoglobo series","Glyoxylate and dicarboxylate metabolism","GnRH signaling pathway","Golgi organization","Golgi to endosome transport","Golgi to plasma membrane protein transport","Golgi to plasma membrane transport","Golgi to vacuole transport","Golgi vesicle transport","Hedgehog signaling pathway","Herpes simplex virus 1 infection","Histidine metabolism","Homologous recombination","Hypertrophy model","IL 4 signaling pathway","IMP biosynthetic process","IMP metabolic process","IMP salvage","Inositol phosphate metabolism","Insulin signaling","Insulin signaling pathway","Integrin mediated cell adhesion","Intestinal immune network for IgA production","JNK cascade","KEGG root term","Kupffer's vesicle development","L-alanine catabolic process","L-alanine import across plasma membrane","L-alanine metabolic process","L-alanine transmembrane transport","L-alanine transport","L-alpha-amino acid transmembrane transport","L-amino acid biosynthetic process","L-amino acid catabolic process","L-amino acid metabolic process","L-amino acid transport","L-arginine import across plasma membrane","L-arginine transmembrane transport","L-aspartate transmembrane transport","L-glutamate import","L-glutamate transmembrane transport","L-leucine import across plasma membrane","L-lysine transmembrane transport","L-lysine transport","L-methionine biosynthetic process","L-methionine salvage","L-methionine salvage from methylthioadenosine","L-ornithine transmembrane transport","L-proline biosynthetic process","L-serine biosynthetic process","L-serine metabolic process","L-serine transport","Linoleic acid metabolism","Lipoic acid metabolism","Lysine degradation","Lysosome","MAPK cascade","MAPK signaling pathway","Mannose type O-glycan biosynthesis","Melanogenesis","Metabolic pathways","Metabolism of xenobiotics by cytochrome P450","Mismatch repair","Mitochondrial long chain fatty acid beta oxidation","Mitophagy - animal","Model of Fgf and Wnt crosstalk in tailbud ","Motor proteins","Mucin type O-glycan biosynthesis","N-Glycan biosynthesis","N-glycan processing","N-terminal protein amino acid modification","N-terminal protein lipidation","N-terminal protein myristoylation","NAD metabolic process","NADH dehydrogenase complex assembly","NADH metabolic process","NLS-bearing protein import into nucleus","NOD-like receptor signaling pathway","Necroptosis","Neural crest development","Neuroactive ligand-receptor interaction","Nicotinate and nicotinamide metabolism","Nitrogen metabolism","Nodal signaling pathway","Non canonical Wnt pathway","Non-homologous end-joining","Notch signaling pathway","Nuclear receptors in lipid metabolism and toxicity","Nucleocytoplasmic transport","Nucleotide excision repair","Nucleotide metabolism","O-glycan processing","One carbon metabolism","One carbon pool by folate","Oocyte meiosis","Other glycan degradation","Other types of O-glycan biosynthesis","Oxidative phosphorylation","Oxidative stress response","P-body assembly","PPAR signaling pathway","Pantothenate and CoA biosynthesis","Pentose and glucuronate interconversions","Peroxisome","Phagosome","Phenylalanine metabolism","Phenylalanine, tyrosine and tryptophan biosynthesis","Phosphatidylinositol signaling system","Phosphonate and phosphinate metabolism","Phototransduction","Polycomb repressive complex","Porphyrin metabolism","Primary bile acid biosynthesis","Progesterone-mediated oocyte maturation","Propanoate metabolism","Proteasome","Protein export","Protein processing in endoplasmic reticulum","Purine metabolism","Pyrimidine metabolism","Pyruvate metabolism","RIG-I-like receptor signaling pathway","RNA 3'-end processing","RNA 5'-end processing","RNA biosynthetic process","RNA catabolic process","RNA degradation","RNA destabilization","RNA export from nucleus","RNA localization","RNA localization to chromatin","RNA metabolic process","RNA methylation","RNA modification","RNA polymerase","RNA polymerase I assembly","RNA polymerase II core complex assembly","RNA polymerase II preinitiation complex assembly","RNA polymerase III assembly","RNA processing","RNA splicing","RNA splicing, via endonucleolytic cleavage and ligation","RNA splicing, via transesterification reactions","RNA splicing, via transesterification reactions with bulged adenosine as nucleophile","RNA stabilization","RNA transport","RNA-templated DNA biosynthetic process","Rab protein signal transduction","Rap protein signal transduction","Ras protein signal transduction","Regulation of actin cytoskeleton","Retinol metabolism","Rho protein signal transduction","Riboflavin metabolism","Ribosome","Ribosome biogenesis in eukaryotes","SCF complex assembly","SCF-dependent proteasomal ubiquitin-dependent protein catabolic process","SNARE interactions in vesicular transport","SREBP signaling pathway","SREBP-SCAP complex retention in endoplasmic reticulum","SRP-dependent cotranslational protein targeting to membrane","SRP-dependent cotranslational protein targeting to membrane, translocation","Salmonella infection","Schwann cell development","Schwann cell differentiation","Selenium metabolism selenoproteins","Selenocompound metabolism","Spemann organizer formation","Sphingolipid metabolism","Spliceosome","Starch and sucrose metabolism","Steroid biosynthesis","Steroid hormone biosynthesis","Sulfur metabolism","T cell activation","T cell chemotaxis","T cell costimulation","T cell differentiation","T cell differentiation in thymus","T cell migration","T cell proliferation","TCA cycle","TGF-beta signaling pathway","TOR signaling","TORC1 signaling","TORC2 signaling","Terpenoid backbone biosynthesis","Thiamine metabolism","Tight junction","Toll like receptor signaling pathway","Toll signaling pathway","Toll-like receptor signaling pathway","Tryptophan metabolism","Tyrosine metabolism","U6 snRNA 3'-end processing","UMP biosynthetic process","UMP metabolic process","UTP biosynthetic process","UTP metabolic process","Ubiquitin mediated proteolysis","VEGF signaling pathway","Valine, leucine and isoleucine biosynthesis","Valine, leucine and isoleucine degradation","Various types of N-glycan biosynthesis","Vascular smooth muscle contraction","Viral life cycle - HIV-1","Virion - Herpesvirus","Vitamin B6 metabolism","WIKIPATHWAYS","Wnt Signaling Pathway","Wnt Signaling Pathway and Pluripotency","Wnt signaling pathway","Wnt signaling pathway, planar cell polarity pathway","acetate metabolic process","acetyl-CoA biosynthetic process","acetyl-CoA biosynthetic process from acetate","acetyl-CoA metabolic process","acetylcholine receptor signaling pathway","acidic amino acid transport","actin cortical patch localization","actin crosslink formation","actin cytoskeleton organization","actin filament bundle assembly","actin filament bundle organization","actin filament capping","actin filament depolymerization","actin filament organization","actin filament polymerization","actin filament severing","actin filament-based movement","actin filament-based process","actin filament-based transport","actin nucleation","actin polymerization or depolymerization","actin-mediated cell contraction","actin-myosin filament sliding","action potential","action potential propagation","activation of GTPase activity","activation of immune response","activation of innate immune response","activation of store-operated calcium channel activity","activin receptor signaling pathway","actomyosin structure organization","acyl carnitine transmembrane transport","acyl carnitine transport","acyl-CoA biosynthetic process","acyl-CoA metabolic process","acylglycerol biosynthetic process","acylglycerol metabolic process","adenosine 5'-(hexahydrogen pentaphosphate) catabolic process","adenosine 5'-(hexahydrogen pentaphosphate) metabolic process","adenylate cyclase-activating G protein-coupled receptor signaling pathway","adenylate cyclase-activating dopamine receptor signaling pathway","adenylate cyclase-inhibiting G protein-coupled receptor signaling pathway","adenylate cyclase-inhibiting serotonin receptor signaling pathway","adenylate cyclase-modulating G protein-coupled receptor signaling pathway","adherens junction organization","adhesion of symbiont to host","adhesion of symbiont to host cell","adiponectin-activated signaling pathway","adipose tissue development","adult behavior","adult feeding behavior","aerobic electron transport chain","aerobic respiration","alanine catabolic process","alanine metabolic process","alanine transport","alanyl-tRNA aminoacylation","alcohol biosynthetic process","alcohol metabolic process","aldehyde catabolic process","alditol phosphate metabolic process","alkanesulfonate transport","alpha-Linolenic acid metabolism","alpha-amino acid biosynthetic process","alpha-amino acid catabolic process","alpha-amino acid metabolic process","alternative mRNA splicing, via spliceosome","ameboidal-type cell migration","amide biosynthetic process","amide metabolic process","amide transport","amine biosynthetic process","amine catabolic process","amine metabolic process","amino acid activation","amino acid biosynthetic process","amino acid catabolic process","amino acid import across plasma membrane","amino acid metabolic process","amino acid salvage","amino acid transmembrane transport","amino acid transport","amino-acid betaine biosynthetic process","amino-acid betaine metabolic process","amino-acid betaine transport","aminoacyl-tRNA metabolism involved in translational fidelity","aminoglycan biosynthetic process","aminoglycan metabolic process","aminophospholipid transport","ammonium transmembrane transport","amyloid precursor protein biosynthetic process","amyloid precursor protein metabolic process","amyloid-beta clearance","amyloid-beta clearance by cellular catabolic process","anaphase-promoting complex-dependent catabolic process","anatomical structure development","anatomical structure formation involved in morphogenesis","anatomical structure homeostasis","anatomical structure maturation","anatomical structure morphogenesis","angioblast cell migration","angiogenesis","animal organ development","animal organ formation","animal organ maturation","animal organ morphogenesis","anterior commissure morphogenesis","anterior lateral line development","anterior lateral line neuromast development","anterior lateral line neuromast hair cell development","anterior lateral line neuromast hair cell differentiation","anterior lateral line system development","anterior/posterior axis specification","anterior/posterior axis specification, embryo","anterior/posterior axon guidance","anterior/posterior pattern specification","anterograde trans-synaptic signaling","antibacterial humoral response","antigen processing and presentation","antigen processing and presentation of endogenous antigen","antigen processing and presentation of endogenous peptide antigen","antigen processing and presentation of endogenous peptide antigen via MHC class I","antigen processing and presentation of peptide antigen","antigen processing and presentation of peptide antigen via MHC class I","antigen processing and presentation of peptide antigen via MHC class II","antigen processing and presentation of peptide or polysaccharide antigen via MHC class II","antimicrobial humoral response","aorta development","apelin receptor signaling pathway","apical junction assembly","apoptotic process","apoptotic signaling pathway","appendage development","appendage morphogenesis","arachidonate transport","arachidonic acid metabolic process","arachidonic acid secretion","arginine catabolic process","arginine catabolic process to glutamate","arginine catabolic process to ornithine","arginine catabolic process to proline","arginine catabolic process to proline via ornithine","arginine metabolic process","aromatic amino acid metabolic process","aromatic amino acid transport","aromatic compound biosynthetic process","aromatic compound catabolic process","artery development","artery morphogenesis","asparagine biosynthetic process","asparagine metabolic process","asparaginyl-tRNA aminoacylation","aspartate biosynthetic process","aspartate family amino acid biosynthetic process","aspartate family amino acid metabolic process","aspartate metabolic process","aspartate transmembrane transport","aspartyl-tRNA aminoacylation","atrioventricular canal development","attachment of mitotic spindle microtubules to kinetochore","attachment of spindle microtubules to kinetochore","auditory behavior","auditory receptor cell fate commitment","auditory receptor cell fate determination","autonomic nervous system development","autophagosome assembly","autophagosome maturation","autophagosome organization","autophagy","autophagy of mitochondrion","autophagy of peroxisome","axis specification","axo-dendritic transport","axon development","axon ensheathment","axon extension","axon extension involved in axon guidance","axon guidance","axon regeneration","axon regeneration at neuromuscular junction","axonal fasciculation","axonal transport","axonogenesis","barbed-end actin filament capping","base conversion or substitution editing","base-excision repair","basement membrane assembly","basement membrane organization","basic amino acid transmembrane transport","basic amino acid transport","basophil activation","basophil activation involved in immune response","behavior","behavioral response to nicotine","beta-Alanine metabolism","bicellular tight junction assembly","bile acid and bile salt transport","biogenic amine biosynthetic process","biogenic amine metabolic process","biological process involved in interaction with host","biological process involved in interspecies interaction between organisms","biological process involved in symbiotic interaction","biological regulation","biological_process","biomineral tissue development","biosynthetic process","blastoderm segmentation","blood circulation","blood coagulation","blood vessel development","blood vessel diameter maintenance","blood vessel endothelial cell differentiation","blood vessel endothelial cell fate commitment","blood vessel endothelial cell migration","blood vessel lumenization","blood vessel morphogenesis","body morphogenesis","bone development","bone mineralization","bone morphogenesis","bone remodeling","bone resorption","brain development","brain morphogenesis","branched-chain amino acid transport","branching involved in blood vessel morphogenesis","branching morphogenesis of an epithelial tube","bulbus arteriosus development","cAMP-mediated signaling","cGMP biosynthetic process","cGMP metabolic process","calcineurin-mediated signaling","calcitriol biosynthetic process from calciol","calcium import into the mitochondrion","calcium ion homeostasis","calcium ion import","calcium ion import across plasma membrane","calcium ion import into cytosol","calcium ion transmembrane import into cytosol","calcium ion transmembrane transport","calcium ion transport","calcium-dependent cell-cell adhesion via plasma membrane cell adhesion molecules","calcium-ion regulated exocytosis","calcium-mediated signaling","camera-type eye development","camera-type eye morphogenesis","camera-type eye photoreceptor cell differentiation","canonical NF-kappaB signal transduction","canonical Wnt signaling pathway","canonical Wnt signaling pathway involved in negative regulation of apoptotic process","carbohydrate biosynthetic process","carbohydrate catabolic process","carbohydrate derivative biosynthetic process","carbohydrate derivative catabolic process","carbohydrate derivative metabolic process","carbohydrate derivative transport","carbohydrate homeostasis","carbohydrate metabolic process","carbohydrate transport","carboxylic acid biosynthetic process","carboxylic acid catabolic process","carboxylic acid metabolic process","carboxylic acid transmembrane transport","carboxylic acid transport","cardiac cell development","cardiac chamber development","cardiac muscle cell action potential","cardiac muscle cell action potential involved in contraction","cardiac muscle cell contraction","cardiac muscle cell development","cardiac muscle cell differentiation","cardiac muscle cell membrane repolarization","cardiac muscle contraction","cardiac muscle myosin thick filament assembly","cardiac muscle tissue development","cardiac muscle tissue regeneration","cardiac myofibril assembly","cardiocyte differentiation","cardiolipin acyl-chain remodeling","cardiolipin biosynthetic process","cardiolipin metabolic process","carnitine biosynthetic process","carnitine metabolic process","carnitine transmembrane transport","carnitine transport","cartilage condensation","cartilage development","cartilage morphogenesis","catabolic process","catechol-containing compound metabolic process","catecholamine metabolic process","caveola assembly","cell activation","cell activation involved in immune response","cell adhesion","cell aggregation","cell chemotaxis","cell communication","cell communication by electrical coupling","cell cycle","cell cycle G1/S phase transition","cell cycle G2/M phase transition","cell cycle checkpoint signaling","cell cycle phase transition","cell cycle process","cell death","cell development","cell differentiation","cell differentiation in spinal cord","cell differentiation involved in kidney development","cell differentiation involved in pronephros development","cell division","cell fate commitment","cell fate determination","cell fate specification","cell growth","cell junction assembly","cell junction organization","cell maturation","cell migration","cell migration involved in gastrulation","cell migration involved in mesendoderm migration","cell migration involved in sprouting angiogenesis","cell morphogenesis","cell morphogenesis involved in neuron differentiation","cell motility","cell population proliferation","cell projection assembly","cell projection morphogenesis","cell projection organization","cell proliferation in hindbrain","cell proliferation in hindbrain ventricular zone","cell recognition","cell redox homeostasis","cell surface pattern recognition receptor signaling pathway","cell surface receptor signaling pathway","cell surface toll-like receptor signaling pathway","cell-cell adhesion","cell-cell adhesion mediated by cadherin","cell-cell adhesion via plasma-membrane adhesion molecules","cell-cell fusion","cell-cell junction assembly","cell-cell junction organization","cell-cell signaling","cell-cell signaling by wnt","cell-matrix adhesion","cell-substrate adhesion","cell-substrate junction assembly","cell-substrate junction organization","cellular aldehyde metabolic process","cellular anatomical entity morphogenesis","cellular aromatic compound metabolic process","cellular biogenic amine catabolic process","cellular biosynthetic process","cellular catabolic process","cellular component assembly","cellular component assembly involved in morphogenesis","cellular component biogenesis","cellular component disassembly","cellular component maintenance","cellular component organization","cellular component organization or biogenesis","cellular detoxification","cellular detoxification of aldehyde","cellular developmental process","cellular extravasation","cellular homeostasis","cellular ketone body metabolic process","cellular ketone metabolic process","cellular lipid catabolic process","cellular lipid metabolic process","cellular localization","cellular macromolecule localization","cellular metabolic compound salvage","cellular metabolic process","cellular modified amino acid biosynthetic process","cellular modified amino acid catabolic process","cellular modified amino acid metabolic process","cellular nitrogen compound biosynthetic process","cellular nitrogen compound catabolic process","cellular nitrogen compound metabolic process","cellular oxidant detoxification","cellular pigmentation","cellular process","cellular process involved in reproduction in multicellular organism","cellular respiration","cellular response to BMP stimulus","cellular response to acetylcholine","cellular response to aldehyde","cellular response to amino acid starvation","cellular response to biotic stimulus","cellular response to carbohydrate stimulus","cellular response to catecholamine stimulus","cellular response to chemical stimulus","cellular response to chemical stress","cellular response to chemokine","cellular response to corticotropin-releasing hormone stimulus","cellular response to cytokine stimulus","cellular response to dopamine","cellular response to endogenous stimulus","cellular response to estrogen stimulus","cellular response to external stimulus","cellular response to extracellular stimulus","cellular response to fibroblast growth factor stimulus","cellular response to glucagon stimulus","cellular response to glucose stimulus","cellular response to growth factor stimulus","cellular response to growth hormone stimulus","cellular response to heat","cellular response to hexose stimulus","cellular response to hormone stimulus","cellular response to insulin stimulus","cellular response to interleukin-1","cellular response to lipid","cellular response to lipopolysaccharide","cellular response to misfolded protein","cellular response to molecule of bacterial origin","cellular response to monoamine stimulus","cellular response to monosaccharide stimulus","cellular response to nitrogen compound","cellular response to nutrient levels","cellular response to organic cyclic compound","cellular response to organic substance","cellular response to organonitrogen compound","cellular response to oxidative stress","cellular response to oxygen-containing compound","cellular response to peptide","cellular response to peptide hormone stimulus","cellular response to starvation","cellular response to steroid hormone stimulus","cellular response to sterol","cellular response to sterol depletion","cellular response to stimulus","cellular response to stress","cellular response to topologically incorrect protein","cellular response to toxic substance","cellular response to transforming growth factor beta stimulus","cellular response to tumor necrosis factor","cellular response to type II interferon","cellular response to unfolded protein","cellular response to xenobiotic stimulus","central nervous system development","central nervous system morphogenesis","central nervous system neuron axonogenesis","central nervous system neuron development","central nervous system neuron differentiation","central nervous system projection neuron axonogenesis","centromere complex assembly","centromeric sister chromatid cohesion","centrosome cycle","centrosome duplication","centrosome-templated microtubule nucleation","ceramide biosynthetic process","ceramide catabolic process","ceramide metabolic process","cerebellum vasculature development","chaperone cofactor-dependent protein refolding","chaperone-mediated protein complex assembly","chaperone-mediated protein folding","chemical homeostasis","chemical synaptic transmission","chemical synaptic transmission, postsynaptic","chemokine-mediated signaling pathway","chemotaxis","chloride transmembrane transport","chloride transport","cholesterol biosynthetic process","cholesterol homeostasis","cholesterol metabolic process","cholesterol transport","chondrocyte differentiation","chondrocyte proliferation","chondroitin sulfate biosynthetic process","chondroitin sulfate metabolic process","chondroitin sulfate proteoglycan biosynthetic process","chondroitin sulfate proteoglycan metabolic process","chordate embryonic development","chromatin organization","chromatin remodeling","chromatin remodeling at centromere","chromosome condensation","chromosome localization","chromosome organization","chromosome segregation","cilium assembly","cilium movement","cilium organization","circadian rhythm","circulatory system development","circulatory system process","citrate metabolic process","clathrin-dependent endocytosis","clathrin-dependent synaptic vesicle endocytosis","closure of optic fissure","co-transcriptional RNA 3'-end processing, cleavage and polyadenylation pathway","co-transcriptional mRNA 3'-end processing, cleavage and polyadenylation pathway","coagulation","collagen fibril organization","collagen metabolic process","columnar/cuboidal epithelial cell development","columnar/cuboidal epithelial cell differentiation","common bile duct development","complement activation","cone photoresponse recovery","connective tissue development","convergent extension","convergent extension involved in gastrulation","copper ion transmembrane transport","copper ion transport","cortical actin cytoskeleton organization","cortical cytoskeleton organization","cotranslational protein targeting to membrane","cranial ganglion development","cranial nerve development","cranial nerve formation","cranial nerve morphogenesis","cranial skeletal system development","cyclic nucleotide biosynthetic process","cyclic nucleotide metabolic process","cyclic purine nucleotide metabolic process","cyclic-nucleotide-mediated signaling","cysteine biosynthetic process","cysteine biosynthetic process via cystathionine","cysteine metabolic process","cytidine to uridine editing","cytochrome complex assembly","cytokine production","cytokine-mediated signaling pathway","cytokinesis","cytolysis","cytoplasm organization","cytoplasmic microtubule organization","cytoplasmic pattern recognition receptor signaling pathway","cytoplasmic translation","cytoplasmic translational elongation","cytoplasmic translational initiation","cytoskeleton organization","cytoskeleton-dependent cytokinesis","cytoskeleton-dependent intracellular transport","cytosolic transport","dTMP biosynthetic process","dTMP metabolic process","dTTP biosynthetic process","dTTP metabolic process","dUMP biosynthetic process","dUMP metabolic process","dUTP catabolic process","dUTP metabolic process","defense response","defense response to bacterium","defense response to other organism","defense response to symbiont","definitive hemopoiesis","delamination","demethylation","dendrite development","dendrite morphogenesis","dendritic spine development","dendritic spine morphogenesis","dendritic spine organization","deoxyribonucleoside monophosphate biosynthetic process","deoxyribonucleoside monophosphate metabolic process","deoxyribonucleotide biosynthetic process","deoxyribonucleotide catabolic process","deoxyribonucleotide metabolic process","deoxyribose phosphate biosynthetic process","deoxyribose phosphate catabolic process","deoxyribose phosphate metabolic process","dephosphorylation","detection of abiotic stimulus","detection of external stimulus","detection of light stimulus","detection of mechanical stimulus","detection of mechanical stimulus involved in sensory perception","detection of stimulus","detection of stimulus involved in sensory perception","detection of visible light","determination of bilateral symmetry","determination of digestive tract left/right asymmetry","determination of heart left/right asymmetry","determination of left/right asymmetry in lateral mesoderm","determination of left/right symmetry","detoxification","detoxification of cadmium ion","detoxification of inorganic compound","developmental cell growth","developmental growth","developmental growth involved in morphogenesis","developmental induction","developmental maturation","developmental pigmentation","developmental process","developmental process involved in reproduction","diadenosine hexaphosphate catabolic process","diadenosine hexaphosphate metabolic process","diadenosine pentaphosphate catabolic process","diadenosine pentaphosphate metabolic process","diadenosine polyphosphate biosynthetic process","diadenosine polyphosphate catabolic process","diadenosine polyphosphate metabolic process","diadenosine tetraphosphate biosynthetic process","diadenosine tetraphosphate metabolic process","dicarboxylic acid biosynthetic process","dicarboxylic acid catabolic process","dicarboxylic acid metabolic process","dicarboxylic acid transport","diencephalon development","digestion","digestive system development","digestive system process","digestive tract development","digestive tract morphogenesis","diphosphoinositol polyphosphate metabolic process","direct ossification","disaccharide transport","diterpenoid metabolic process","dolichol biosynthetic process","dolichol metabolic process","dolichol-linked oligosaccharide biosynthetic process","dopamine receptor signaling pathway","dopaminergic neuron differentiation","dorsal aorta development","dorsal motor nucleus of vagus nerve development","dorsal/ventral axis specification","dorsal/ventral pattern formation","double-strand break repair","double-strand break repair via break-induced replication","double-strand break repair via homologous recombination","ear development","ear morphogenesis","early endosome to late endosome transport","ectodermal placode development","ectodermal placode formation","ectodermal placode morphogenesis","elastic fiber assembly","electron transport chain","embryo development","embryo development ending in birth or egg hatching","embryonic axis specification","embryonic body morphogenesis","embryonic camera-type eye development","embryonic camera-type eye formation","embryonic camera-type eye morphogenesis","embryonic cranial skeleton morphogenesis","embryonic epithelial tube formation","embryonic eye morphogenesis","embryonic heart tube development","embryonic heart tube morphogenesis","embryonic hemopoiesis","embryonic morphogenesis","embryonic neurocranium morphogenesis","embryonic organ development","embryonic organ morphogenesis","embryonic pattern specification","embryonic retina morphogenesis in camera-type eye","embryonic skeletal system development","embryonic skeletal system morphogenesis","embryonic viscerocranium morphogenesis","endocardium development","endocrine pancreas development","endocrine system development","endocytic recycling","endocytosis","endoderm development","endoderm formation","endolysosomal toll-like receptor signaling pathway","endomembrane system organization","endonucleolytic cleavage of tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)","endoplasmic reticulum membrane organization","endoplasmic reticulum organization","endoplasmic reticulum to Golgi vesicle-mediated transport","endoplasmic reticulum tubular network maintenance","endoplasmic reticulum tubular network organization","endoplasmic reticulum unfolded protein response","endosomal transport","endosome to lysosome transport","endosome to plasma membrane protein transport","endothelial cell activation","endothelial cell activation involved in immune response","endothelial cell development","endothelial cell differentiation","endothelial cell fate commitment","endothelial cell migration","endothelial to hematopoietic transition","endothelial tube morphogenesis","endothelium development","energy derivation by oxidation of organic compounds","energy reserve metabolic process","ensheathment of neurons","enteroendocrine cell differentiation","entrainment of circadian clock","enzyme-linked receptor protein signaling pathway","epiboly","epiboly involved in gastrulation with mouth forming second","epidermal cell differentiation","epidermal growth factor receptor signaling pathway","epidermis development","epigenetic regulation of gene expression","epithalamus development","epithelial cell development","epithelial cell differentiation","epithelial cell differentiation involved in kidney development","epithelial cell fate commitment","epithelial cell migration","epithelial cell proliferation","epithelial structure maintenance","epithelial to mesenchymal transition","epithelial tube formation","epithelial tube morphogenesis","epithelium development","epithelium migration","epithelium regeneration","error-free translesion synthesis","error-prone translesion synthesis","erythrocyte development","erythrocyte differentiation","erythrocyte homeostasis","establishment of RNA localization","establishment of chromosome localization","establishment of endothelial barrier","establishment of localization","establishment of localization in cell","establishment of melanosome localization","establishment of mitotic spindle localization","establishment of organelle localization","establishment of pigment granule localization","establishment of protein localization","establishment of protein localization to endoplasmic reticulum","establishment of protein localization to extracellular region","establishment of protein localization to membrane","establishment of protein localization to mitochondrial membrane","establishment of protein localization to mitochondrion","establishment of protein localization to organelle","establishment of protein localization to peroxisome","establishment of protein localization to plasma membrane","establishment of protein localization to postsynaptic membrane","establishment of protein localization to vacuole","establishment of spindle localization","establishment of vesicle localization","establishment or maintenance of apical/basal cell polarity","establishment or maintenance of bipolar cell polarity","establishment or maintenance of cell polarity","establishment or maintenance of cytoskeleton polarity","establishment or maintenance of epithelial cell apical/basal polarity","establishment or maintenance of microtubule cytoskeleton polarity","establishment or maintenance of transmembrane electrochemical gradient","excitatory postsynaptic potential","exocrine pancreas development","exocrine system development","exocyst localization","exocytic process","exocytosis","exonucleolytic trimming to generate mature 3'-end of 5.8S rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)","export across plasma membrane","export from cell","external encapsulating structure organization","extracellular matrix assembly","extracellular matrix organization","extracellular structure organization","extrinsic apoptotic signaling pathway","extrinsic apoptotic signaling pathway in absence of ligand","eye development","eye morphogenesis","eye photoreceptor cell development","eye photoreceptor cell differentiation","facultative heterochromatin formation","farnesyl diphosphate biosynthetic process","farnesyl diphosphate biosynthetic process, mevalonate pathway","farnesyl diphosphate metabolic process","fat pad development","fat-soluble vitamin biosynthetic process","fat-soluble vitamin metabolic process","fatty acid beta-oxidation","fatty acid beta-oxidation using acyl-CoA oxidase","fatty acid biosynthetic process","fatty acid catabolic process","fatty acid derivative biosynthetic process","fatty acid derivative catabolic process","fatty acid derivative metabolic process","fatty acid metabolic process","fatty acid oxidation","fatty acid transport","fatty-acyl-CoA biosynthetic process","fatty-acyl-CoA metabolic process","feeding behavior","female gamete generation","fibroblast growth factor receptor signaling pathway","filopodium assembly","fin development","fin morphogenesis","fin regeneration","floor plate development","floor plate formation","floor plate morphogenesis","fluid transport","folic acid-containing compound catabolic process","folic acid-containing compound metabolic process","forebrain development","forebrain generation of neurons","forebrain neuron development","forebrain neuron differentiation","formaldehyde catabolic process","formaldehyde metabolic process","formation of primary germ layer","galactolipid biosynthetic process","galactolipid metabolic process","galactosylceramide biosynthetic process","galactosylceramide metabolic process","gamete generation","gamma-aminobutyric acid signaling pathway","gamma-aminobutyric acid transport","ganglion development","ganglioside biosynthetic process","ganglioside catabolic process","ganglioside metabolic process","gap junction-mediated intercellular transport","gas transport","gastric inhibitory peptide signaling pathway","gastrulation","gastrulation with mouth forming second","gene expression","generation of neurons","generation of precursor metabolites and energy","germ cell development","gland development","glandular epithelial cell development","glandular epithelial cell differentiation","glial cell development","glial cell differentiation","gliogenesis","glucan biosynthetic process","glucan catabolic process","glucan metabolic process","glucose homeostasis","glucose metabolic process","glucosylceramide catabolic process","glucosylceramide metabolic process","glutamate metabolic process","glutamine biosynthetic process","glutamine family amino acid biosynthetic process","glutamine family amino acid catabolic process","glutamine family amino acid metabolic process","glutamine metabolic process","glutamyl-tRNA aminoacylation","glutathione biosynthetic process","glutathione metabolic process","glutathione transmembrane transport","glutathione transport","glycerol-3-phosphate metabolic process","glycerolipid biosynthetic process","glycerolipid metabolic process","glycerophospholipid biosynthetic process","glycerophospholipid metabolic process","glycine catabolic process","glycine metabolic process","glycogen biosynthetic process","glycogen catabolic process","glycogen metabolic process","glycolipid biosynthetic process","glycolipid catabolic process","glycolipid metabolic process","glycolytic process","glycoprotein biosynthetic process","glycoprotein metabolic process","glycosaminoglycan biosynthetic process","glycosaminoglycan metabolic process","glycosphingolipid biosynthetic process","glycosphingolipid catabolic process","glycosphingolipid metabolic process","glycosyl compound biosynthetic process","glycosyl compound metabolic process","glycosylation","glycosylceramide biosynthetic process","glycosylceramide catabolic process","glycosylceramide metabolic process","glycyl-tRNA aminoacylation","granulocyte activation","granulocyte chemotaxis","granulocyte differentiation","granulocyte migration","growth","growth hormone receptor signaling pathway","growth hormone receptor signaling pathway via JAK-STAT","guanine biosynthetic process","guanine metabolic process","guanine salvage","habenula development","hair cell differentiation","hatching","hatching gland development","head development","heart contraction","heart development","heart formation","heart jogging","heart looping","heart morphogenesis","heart process","hemangioblast cell differentiation","hematopoietic or lymphoid organ development","hematopoietic progenitor cell differentiation","hematopoietic stem cell differentiation","hematopoietic stem cell homeostasis","heme export","heme metabolic process","heme transmembrane transport","heme transport","hemoglobin biosynthetic process","hemoglobin metabolic process","hemopoiesis","hemostasis","heparan sulfate proteoglycan biosynthetic process","hepaticobiliary system development","hepatoblast differentiation","hepoxilin biosynthetic process","hepoxilin metabolic process","heterochromatin formation","heterochromatin organization","heterocycle biosynthetic process","heterocycle catabolic process","heterocycle metabolic process","heterophilic cell-cell adhesion via plasma membrane cell adhesion molecules","heterotypic cell-cell adhesion","hexose metabolic process","hindbrain development","hippo signaling","histidyl-tRNA aminoacylation","homeostasis of number of cells","homeostatic process","homocysteine metabolic process","homologous recombination","homophilic cell adhesion via plasma membrane adhesion molecules","homoserine metabolic process","hormone metabolic process","hormone secretion","hormone transport","hormone-mediated signaling pathway","humoral immune response","hyaluronan biosynthetic process","hyaluronan metabolic process","hydrogen peroxide catabolic process","hydrogen peroxide metabolic process","hypotonic response","hypotonic salinity response","hypoxanthine metabolic process","icosanoid metabolic process","icosanoid secretion","icosanoid transport","immune effector process","immune response","immune response-activating cell surface receptor signaling pathway","immune response-activating signaling pathway","immune response-regulating cell surface receptor signaling pathway","immune response-regulating signaling pathway","immune system development","immune system process","import across plasma membrane","import into cell","import into nucleus","import into the mitochondrion","inflammatory response","innate immune response","innate immune response activating cell surface receptor signaling pathway","innate immune response-activating signaling pathway","inner ear auditory receptor cell differentiation","inner ear development","inner ear morphogenesis","inner ear receptor cell development","inner ear receptor cell differentiation","inner ear receptor cell fate commitment","inner mitochondrial membrane organization","inorganic anion transmembrane transport","inorganic anion transport","inorganic cation import across plasma membrane","inorganic cation transmembrane transport","inorganic ion homeostasis","inorganic ion import across plasma membrane","inorganic ion transmembrane transport","inositol phosphate biosynthetic process","inositol phosphate metabolic process","insulin receptor signaling pathway","insulin secretion","insulin secretion involved in cellular response to glucose stimulus","insulin-like growth factor receptor signaling pathway","integrin-mediated signaling pathway","intein-mediated protein splicing","intercellular transport","interleukin-13 production","intermediate filament cytoskeleton organization","intermediate filament organization","intermediate filament-based process","intermembrane lipid transfer","interstrand cross-link repair","intestinal absorption","intestinal cholesterol absorption","intestinal epithelial structure maintenance","intestinal lipid absorption","intestinal motility","intracellular calcium ion homeostasis","intracellular chemical homeostasis","intracellular cholesterol transport","intracellular glucose homeostasis","intracellular iron ion homeostasis","intracellular lipid transport","intracellular mRNA localization","intracellular mRNA localization involved in anterior/posterior axis specification","intracellular mRNA localization involved in pattern specification process","intracellular monoatomic cation homeostasis","intracellular monoatomic ion homeostasis","intracellular potassium ion homeostasis","intracellular protein transmembrane transport","intracellular protein transport","intracellular receptor signaling pathway","intracellular sequestering of iron ion","intracellular signal transduction","intracellular signaling cassette","intracellular sodium ion homeostasis","intracellular steroid hormone receptor signaling pathway","intracellular sterol transport","intracellular transport","intracellular zinc ion homeostasis","intraciliary transport","intrahepatic bile duct development","intramembranous ossification","intrinsic apoptotic signaling pathway","intrinsic apoptotic signaling pathway by p53 class mediator","iris morphogenesis","iron coordination entity transport","iron ion transmembrane transport","iron ion transport","iron-sulfur cluster assembly","isocitrate metabolic process","isoleucyl-tRNA aminoacylation","isoprenoid biosynthetic process","isoprenoid metabolic process","ketone biosynthetic process","ketone body catabolic process","ketone body metabolic process","kidney development","kidney epithelium development","kidney morphogenesis","kinetochore assembly","kinetochore organization","lagging strand elongation","lamellipodium assembly","lamellipodium morphogenesis","lamellipodium organization","larval behavior","larval locomotory behavior","late endosome to lysosome transport","lateral line development","lateral line ganglion development","lateral line nerve development","lateral line system development","lateral mesoderm development","left/right pattern formation","lens development in camera-type eye","lens fiber cell development","lens fiber cell differentiation","lens fiber cell morphogenesis","lens morphogenesis in camera-type eye","leucine import across plasma membrane","leucine transport","leucyl-tRNA aminoacylation","leukocyte activation","leukocyte activation involved in immune response","leukocyte cell-cell adhesion","leukocyte chemotaxis","leukocyte differentiation","leukocyte homeostasis","leukocyte migration","leukocyte proliferation","limb development","limb morphogenesis","linoleic acid metabolic process","lipid biosynthetic process","lipid catabolic process","lipid digestion","lipid droplet organization","lipid homeostasis","lipid import into cell","lipid localization","lipid metabolic process","lipid modification","lipid oxidation","lipid storage","lipid translocation","lipid transport","lipoprotein biosynthetic process","lipoprotein catabolic process","lipoprotein localization","lipoprotein metabolic process","lipoprotein transport","liposaccharide metabolic process","lipoxygenase pathway","liver development","localization","localization within membrane","locomotion","locomotory behavior","long-chain fatty acid biosynthetic process","long-chain fatty acid import into cell","long-chain fatty acid metabolic process","long-chain fatty acid transport","long-chain fatty-acyl-CoA biosynthetic process","long-chain fatty-acyl-CoA metabolic process","low-density lipoprotein particle clearance","lymph vessel development","lymph vessel morphogenesis","lymphangiogenesis","lymphatic endothelial cell differentiation","lymphatic endothelial cell fate commitment","lymphocyte activation","lymphocyte chemotaxis","lymphocyte costimulation","lymphocyte differentiation","lymphocyte migration","lymphocyte proliferation","lysosomal transport","lysosome organization","lysyl-tRNA aminoacylation","lytic vacuole organization","mRNA 3'-end processing","mRNA catabolic process","mRNA destabilization","mRNA metabolic process","mRNA modification","mRNA polyadenylation","mRNA processing","mRNA splice site recognition","mRNA splicing, via spliceosome","mRNA stabilization","mRNA surveillance pathway","mRNA transport","mTOR signaling pathway","macroautophagy","macromolecule biosynthetic process","macromolecule catabolic process","macromolecule deacylation","macromolecule depalmitoylation","macromolecule glycosylation","macromolecule localization","macromolecule metabolic process","macromolecule methylation","macromolecule modification","macrophage activation","macrophage differentiation","macrophage migration","maintenance of apical/basal cell polarity","maintenance of cell number","maintenance of cell polarity","maintenance of epithelial cell apical/basal polarity","maintenance of gastrointestinal epithelium","maintenance of lens transparency","maintenance of location","maintenance of location in cell","maintenance of protein localization in endoplasmic reticulum","maintenance of protein localization in organelle","maintenance of protein location","maintenance of protein location in cell","malate metabolic process","malate-aspartate shuttle","male gamete generation","mannose metabolic process","mannosylation","mast cell activation","maternal determination of anterior/posterior axis, embryo","maturation of 5.8S rRNA","maturation of 5.8S rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)","maturation of LSU-rRNA","maturation of LSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)","maturation of SSU-rRNA","maturation of SSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)","mechanoreceptor differentiation","mechanoreceptor differentiation involved in mechanosensory epithelium regeneration","mechanosensory behavior","mechanosensory epithelium regeneration","mechanosensory lateral line system development","mediolateral intercalation","medulla oblongata development","meiosis I","meiosis I cell cycle process","meiotic DNA double-strand break formation","meiotic cell cycle","meiotic cell cycle process","meiotic chromosome segregation","meiotic nuclear division","meiotic sister chromatid cohesion","meiotic sister chromatid cohesion, centromeric","melanocyte differentiation","melanocyte proliferation","melanosome localization","melanosome transport","membrane assembly","membrane biogenesis","membrane fusion","membrane lipid biosynthetic process","membrane lipid catabolic process","membrane lipid metabolic process","membrane organization","membrane raft assembly","membrane raft organization","membrane repolarization","mesenchymal cell differentiation","mesenchymal cell migration","mesenchyme development","mesendoderm development","mesendoderm migration","mesoderm development","mesoderm formation","mesoderm morphogenesis","mesodermal cell differentiation","mesodermal cell migration","metabolic process","metal ion transport","metallo-sulfur cluster assembly","metaphase chromosome alignment","metaphase/anaphase transition of cell cycle","metaphase/anaphase transition of mitotic cell cycle","methionine biosynthetic process","methionine metabolic process","methionyl-tRNA aminoacylation","methylation","miRNA processing","miRNA-mediated gene silencing by inhibition of translation","miRNA-mediated post-transcriptional gene silencing","microtubule anchoring","microtubule bundle formation","microtubule cytoskeleton organization","microtubule cytoskeleton organization involved in mitosis","microtubule depolymerization","microtubule nucleation","microtubule nucleation by microtubule organizing center","microtubule organizing center organization","microtubule polymerization","microtubule polymerization or depolymerization","microtubule-based movement","microtubule-based process","microtubule-based transport","midbrain development","mismatch repair","mitochondrial ATP synthesis coupled electron transport","mitochondrial RNA metabolic process","mitochondrial RNA modification","mitochondrial RNA processing","mitochondrial calcium ion homeostasis","mitochondrial calcium ion transmembrane transport","mitochondrial cytochrome c oxidase assembly","mitochondrial electron transport, NADH to ubiquinone","mitochondrial electron transport, cytochrome c to oxygen","mitochondrial electron transport, succinate to ubiquinone","mitochondrial gene expression","mitochondrial membrane organization","mitochondrial protein processing","mitochondrial respirasome assembly","mitochondrial respiratory chain complex I assembly","mitochondrial respiratory chain complex III assembly","mitochondrial respiratory chain complex assembly","mitochondrial tRNA modification","mitochondrial tRNA processing","mitochondrial tRNA wobble uridine modification","mitochondrial translation","mitochondrial transmembrane transport","mitochondrial transport","mitochondrion organization","mitophagy","mitotic DNA damage checkpoint signaling","mitotic DNA integrity checkpoint signaling","mitotic DNA replication checkpoint signaling","mitotic G1 DNA damage checkpoint signaling","mitotic G1/S transition checkpoint signaling","mitotic G2/M transition checkpoint","mitotic cell cycle","mitotic cell cycle checkpoint signaling","mitotic cell cycle phase transition","mitotic cell cycle process","mitotic chromosome condensation","mitotic cytokinesis","mitotic metaphase chromosome alignment","mitotic nuclear division","mitotic sister chromatid segregation","mitotic spindle assembly","mitotic spindle organization","modification-dependent macromolecule catabolic process","modification-dependent protein catabolic process","modified amino acid transport","modulation of chemical synaptic transmission","monoatomic anion transmembrane transport","monoatomic anion transport","monoatomic cation homeostasis","monoatomic cation transmembrane transport","monoatomic cation transport","monoatomic ion homeostasis","monoatomic ion transmembrane transport","monoatomic ion transport","monocarboxylic acid biosynthetic process","monocarboxylic acid catabolic process","monocarboxylic acid metabolic process","monocarboxylic acid transport","monocyte chemotaxis","mononuclear cell differentiation","mononuclear cell migration","mononuclear cell proliferation","monosaccharide metabolic process","monounsaturated fatty acid biosynthetic process","monounsaturated fatty acid metabolic process","morphogenesis of a branching epithelium","morphogenesis of a branching structure","morphogenesis of an endothelium","morphogenesis of an epithelial sheet","morphogenesis of an epithelium","morphogenesis of embryonic epithelium","motor neuron axon guidance","mucopolysaccharide metabolic process","multicellular organism development","multicellular organism reproduction","multicellular organismal movement","multicellular organismal process","multicellular organismal reproductive process","multicellular organismal-level homeostasis","muscle attachment","muscle cell development","muscle cell differentiation","muscle cell proliferation","muscle contraction","muscle organ development","muscle organ morphogenesis","muscle structure development","muscle system process","muscle tissue development","muscle tissue morphogenesis","musculoskeletal movement","myelin maintenance","myelination","myelination in peripheral nervous system","myeloid cell activation involved in immune response","myeloid cell development","myeloid cell differentiation","myeloid cell homeostasis","myeloid leukocyte activation","myeloid leukocyte differentiation","myeloid leukocyte migration","myoblast fusion","myofibril assembly","myosin filament assembly","myosin filament organization","myotube cell development","myotube differentiation","ncRNA 5'-end processing","ncRNA metabolic process","ncRNA processing","ncRNA transcription","necroptotic process","negative chemotaxis","negative regulation of BMP signaling pathway","negative regulation of DNA biosynthetic process","negative regulation of DNA metabolic process","negative regulation of DNA-binding transcription factor activity","negative regulation of DNA-templated transcription","negative regulation of DNA-templated transcription, elongation","negative regulation of ERK1 and ERK2 cascade","negative regulation of G1/S transition of mitotic cell cycle","negative regulation of G2/M transition of mitotic cell cycle","negative regulation of MAPK cascade","negative regulation of NF-kappaB transcription factor activity","negative regulation of Notch signaling pathway","negative regulation of RNA biosynthetic process","negative regulation of RNA catabolic process","negative regulation of RNA metabolic process","negative regulation of RNA splicing","negative regulation of Ras protein signal transduction","negative regulation of Rho protein signal transduction","negative regulation of SREBP signaling pathway","negative regulation of TOR signaling","negative regulation of TORC1 signaling","negative regulation of Wnt signaling pathway","negative regulation of actin filament depolymerization","negative regulation of actin filament polymerization","negative regulation of activin receptor signaling pathway","negative regulation of adenylate cyclase activity","negative regulation of amide metabolic process","negative regulation of amyloid precursor protein biosynthetic process","negative regulation of angiogenesis","negative regulation of apoptotic process","negative regulation of apoptotic signaling pathway","negative regulation of appetite","negative regulation of axon extension","negative regulation of axon extension involved in axon guidance","negative regulation of axonogenesis","negative regulation of biological process","negative regulation of biosynthetic process","negative regulation of blood coagulation","negative regulation of blood vessel morphogenesis","negative regulation of cAMP-mediated signaling","negative regulation of canonical NF-kappaB signal transduction","negative regulation of canonical Wnt signaling pathway","negative regulation of catabolic process","negative regulation of catalytic activity","negative regulation of cell activation","negative regulation of cell adhesion","negative regulation of cell communication","negative regulation of cell cycle","negative regulation of cell cycle G1/S phase transition","negative regulation of cell cycle G2/M phase transition","negative regulation of cell cycle phase transition","negative regulation of cell cycle process","negative regulation of cell development","negative regulation of cell differentiation","negative regulation of cell growth","negative regulation of cell migration","negative regulation of cell motility","negative regulation of cell population proliferation","negative regulation of cell projection organization","negative regulation of cell-cell adhesion","negative regulation of cell-substrate adhesion","negative regulation of cellular biosynthetic process","negative regulation of cellular catabolic process","negative regulation of cellular component organization","negative regulation of cellular metabolic process","negative regulation of cellular process","negative regulation of cellular response to growth factor stimulus","negative regulation of cellular response to insulin stimulus","negative regulation of chemotaxis","negative regulation of chromosome organization","negative regulation of coagulation","negative regulation of cyclase activity","negative regulation of cytoskeleton organization","negative regulation of defense response","negative regulation of developmental growth","negative regulation of developmental process","negative regulation of epithelial cell differentiation","negative regulation of fibroblast growth factor receptor signaling pathway","negative regulation of gene expression","negative regulation of gene expression, epigenetic","negative regulation of glycoprotein biosynthetic process","negative regulation of glycoprotein metabolic process","negative regulation of growth","negative regulation of hematopoietic progenitor cell differentiation","negative regulation of hematopoietic stem cell differentiation","negative regulation of hemostasis","negative regulation of hydrolase activity","negative regulation of immune system process","negative regulation of inflammatory response","negative regulation of insulin receptor signaling pathway","negative regulation of intracellular signal transduction","negative regulation of intrinsic apoptotic signaling pathway","negative regulation of intrinsic apoptotic signaling pathway by p53 class mediator","negative regulation of leukocyte activation","negative regulation of locomotion","negative regulation of lyase activity","negative regulation of mRNA catabolic process","negative regulation of mRNA metabolic process","negative regulation of mRNA processing","negative regulation of mRNA splicing, via spliceosome","negative regulation of macromolecule biosynthetic process","negative regulation of macromolecule metabolic process","negative regulation of macrophage activation","negative regulation of metabolic process","negative regulation of mitotic cell cycle","negative regulation of mitotic cell cycle phase transition","negative regulation of molecular function","negative regulation of multicellular organismal process","negative regulation of muscle contraction","negative regulation of nervous system development","negative regulation of neurogenesis","negative regulation of neuroinflammatory response","negative regulation of neuron projection development","negative regulation of nitrogen compound metabolic process","negative regulation of nodal signaling pathway","negative regulation of nucleobase-containing compound metabolic process","negative regulation of organelle organization","negative regulation of pancreatic A cell differentiation","negative regulation of peptidase activity","negative regulation of phosphate metabolic process","negative regulation of phosphorus metabolic process","negative regulation of phosphorylation","negative regulation of programmed cell death","negative regulation of protein catabolic process","negative regulation of protein depolymerization","negative regulation of protein metabolic process","negative regulation of protein modification process","negative regulation of protein phosphorylation","negative regulation of protein polymerization","negative regulation of protein-containing complex assembly","negative regulation of protein-containing complex disassembly","negative regulation of proteolysis","negative regulation of response to external stimulus","negative regulation of response to extracellular stimulus","negative regulation of response to food","negative regulation of response to nutrient levels","negative regulation of response to stimulus","negative regulation of response to wounding","negative regulation of sequestering of calcium ion","negative regulation of signal transduction","negative regulation of signal transduction by p53 class mediator","negative regulation of signaling","negative regulation of small GTPase mediated signal transduction","negative regulation of smoothened signaling pathway","negative regulation of stem cell differentiation","negative regulation of substrate adhesion-dependent cell spreading","negative regulation of supramolecular fiber organization","negative regulation of telomere maintenance","negative regulation of telomere maintenance via telomerase","negative regulation of telomere maintenance via telomere lengthening","negative regulation of transcription by RNA polymerase II","negative regulation of transcription elongation by RNA polymerase II","negative regulation of transforming growth factor beta receptor signaling pathway","negative regulation of translation","negative regulation of translational initiation","negative regulation of transmembrane receptor protein serine/threonine kinase signaling pathway","negative regulation of type B pancreatic cell development","negative regulation of vasculature development","negative regulation of wound healing","nephron development","nephron epithelium development","nephron tubule development","nephron tubule epithelial cell differentiation","nerve development","nervous system development","nervous system process","neural crest cell development","neural crest cell differentiation","neural crest cell fate commitment","neural crest cell fate specification","neural crest cell migration","neural crest formation","neural nucleus development","neural precursor cell proliferation","neural retina development","neural tube closure","neural tube development","neural tube formation","neuroblast proliferation","neurogenesis","neuroinflammatory response","neuromast deposition","neuromast development","neuromast hair cell development","neuromast hair cell differentiation","neuromast hair cell differentiation involved in neuromast regeneration","neuromast primordium migration","neuromast regeneration","neuromuscular junction development","neuromuscular process","neuromuscular process controlling balance","neuron development","neuron differentiation","neuron fate commitment","neuron fate determination","neuron fate specification","neuron migration","neuron projection development","neuron projection extension","neuron projection extension involved in neuron projection guidance","neuron projection fasciculation","neuron projection guidance","neuron projection maintenance","neuron projection morphogenesis","neuron projection organization","neuron projection regeneration","neuron recognition","neuronal action potential","neuronal action potential propagation","neuropeptide signaling pathway","neurotransmitter receptor transport","neurotransmitter receptor transport postsynaptic membrane to endosome","neurotransmitter receptor transport to plasma membrane","neurotransmitter receptor transport to postsynaptic membrane","neurotransmitter receptor transport, endosome to plasma membrane","neurotransmitter receptor transport, endosome to postsynaptic membrane","neurotransmitter secretion","neurotransmitter transport","neutral amino acid transport","neutral lipid biosynthetic process","neutral lipid metabolic process","neutrophil chemotaxis","neutrophil differentiation","neutrophil homeostasis","neutrophil migration","nicotinamide nucleotide metabolic process","nitrogen compound metabolic process","nitrogen compound transport","nitrogen cycle metabolic process","nodal signaling pathway","non-canonical Wnt signaling pathway","non-membrane-bounded organelle assembly","non-proteinogenic amino acid metabolic process","nonribosomal peptide biosynthetic process","notochord development","nuclear chromosome segregation","nuclear division","nuclear envelope organization","nuclear export","nuclear membrane organization","nuclear migration","nuclear pore localization","nuclear transport","nuclear-transcribed mRNA catabolic process","nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay","nuclear-transcribed mRNA poly(A) tail shortening","nucleic acid metabolic process","nucleic acid transport","nucleobase biosynthetic process","nucleobase metabolic process","nucleobase-containing compound biosynthetic process","nucleobase-containing compound catabolic process","nucleobase-containing compound metabolic process","nucleobase-containing compound transport","nucleobase-containing small molecule biosynthetic process","nucleobase-containing small molecule metabolic process","nucleocytoplasmic transport","nucleoside biosynthetic process","nucleoside bisphosphate biosynthetic process","nucleoside bisphosphate metabolic process","nucleoside diphosphate biosynthetic process","nucleoside diphosphate catabolic process","nucleoside diphosphate metabolic process","nucleoside metabolic process","nucleoside monophosphate biosynthetic process","nucleoside monophosphate metabolic process","nucleoside monophosphate phosphorylation","nucleoside phosphate biosynthetic process","nucleoside phosphate catabolic process","nucleoside phosphate metabolic process","nucleoside salvage","nucleoside transmembrane transport","nucleoside transport","nucleoside triphosphate biosynthetic process","nucleoside triphosphate catabolic process","nucleoside triphosphate metabolic process","nucleosome assembly","nucleosome organization","nucleotide biosynthetic process","nucleotide catabolic process","nucleotide metabolic process","nucleotide salvage","nucleotide-excision repair","nucleotide-sugar biosynthetic process","nucleotide-sugar metabolic process","nucleus localization","nucleus organization","odontogenesis","olefinic compound metabolic process","olfactory bulb axon guidance","olfactory bulb development","olfactory lobe development","olfactory placode development","oligodendrocyte development","oligodendrocyte differentiation","oligopeptide transmembrane transport","oligopeptide transport","oligosaccharide metabolic process","oligosaccharide transport","oligosaccharide-lipid intermediate biosynthetic process","one-carbon metabolic process","oocyte anterior/posterior axis specification","oocyte axis specification","oocyte construction","oocyte development","oocyte differentiation","oocyte maturation","oogenesis","optic cup morphogenesis involved in camera-type eye development","optic nerve development","optic nerve formation","optic nerve morphogenesis","optokinetic behavior","organ growth","organelle assembly","organelle disassembly","organelle fission","organelle fusion","organelle localization","organelle membrane fusion","organelle organization","organelle transport along microtubule","organic acid biosynthetic process","organic acid catabolic process","organic acid metabolic process","organic acid transmembrane transport","organic acid transport","organic anion transport","organic cation transport","organic cyclic compound biosynthetic process","organic cyclic compound catabolic process","organic cyclic compound metabolic process","organic hydroxy compound biosynthetic process","organic hydroxy compound metabolic process","organic hydroxy compound transport","organic substance biosynthetic process","organic substance catabolic process","organic substance metabolic process","organic substance transport","organism emergence from protective structure","organonitrogen compound biosynthetic process","organonitrogen compound catabolic process","organonitrogen compound metabolic process","organophosphate biosynthetic process","organophosphate catabolic process","organophosphate ester transport","organophosphate metabolic process","ornithine metabolic process","ornithine transport","ossification","otic placode formation","otic vesicle development","otic vesicle formation","otic vesicle morphogenesis","otolith development","otolith morphogenesis","oxalate transport","oxaloacetate metabolic process","oxidative phosphorylation","oxoacid metabolic process","oxygen transport","p53 signaling pathway","pancreas development","pancreatic A cell differentiation","pancreatic epsilon cell differentiation","parasympathetic nervous system development","pattern recognition receptor signaling pathway","pattern specification process","pectoral fin development","peptide biosynthetic process","peptide cross-linking","peptide hormone processing","peptide hormone secretion","peptide metabolic process","peptide secretion","peptide transport","peptidyl-amino acid modification","peptidyl-arginine methylation","peptidyl-arginine modification","peptidyl-glutamic acid carboxylation","peptidyl-glutamic acid modification","peptidyl-lysine methylation","peptidyl-lysine modification","peptidyl-lysine modification to peptidyl-hypusine","peptidyl-lysine trimethylation","peptidyl-proline modification","peptidyl-tyrosine modification","peptidyl-tyrosine phosphorylation","perichondral ossification","peripheral nervous system axon ensheathment","peripheral nervous system development","peripheral nervous system myelin maintenance","peripheral nervous system neuron development","peripheral nervous system neuron differentiation","peroxisomal membrane transport","peroxisomal transport","peroxisome organization","phagocytosis","phagolysosome assembly","phagosome maturation","phagosome-lysosome fusion","pharyngeal arch artery morphogenesis","pharyngeal system development","phenol-containing compound metabolic process","phenylalanine transport","phenylalanyl-tRNA aminoacylation","phosphagen biosynthetic process","phosphagen metabolic process","phosphate ion transport","phosphate-containing compound metabolic process","phosphatidylethanolamine biosynthetic process","phosphatidylethanolamine metabolic process","phosphatidylglycerol biosynthetic process","phosphatidylglycerol metabolic process","phosphatidylinositol 3-kinase/protein kinase B signal transduction","phosphatidylinositol biosynthetic process","phosphatidylinositol dephosphorylation","phosphatidylinositol metabolic process","phosphatidylinositol phosphate biosynthetic process","phosphatidylinositol-3-phosphate biosynthetic process","phosphatidylinositol-mediated signaling","phosphatidylserine biosynthetic process","phosphatidylserine metabolic process","phosphocreatine biosynthetic process","phosphocreatine metabolic process","phospholipase C-activating G protein-coupled receptor signaling pathway","phospholipase C-activating dopamine receptor signaling pathway","phospholipid biosynthetic process","phospholipid catabolic process","phospholipid dephosphorylation","phospholipid metabolic process","phospholipid translocation","phospholipid transport","phosphorus metabolic process","phosphorylation","photoperiodism","photoreceptor cell development","photoreceptor cell differentiation","photoreceptor cell maintenance","photoreceptor cell outer segment organization","phototransduction","phototransduction, visible light","pigment cell development","pigment cell differentiation","pigment granule localization","pigment granule transport","pigment metabolic process","pigmentation","planar cell polarity pathway involved in gastrula mediolateral intercalation","plasma lipoprotein particle clearance","plasma membrane bounded cell projection assembly","plasma membrane bounded cell projection morphogenesis","plasma membrane bounded cell projection organization","plasma membrane organization","plasma membrane raft assembly","plasma membrane raft organization","plasma membrane tubulation","plus-end-directed organelle transport along microtubule","plus-end-directed vesicle transport along microtubule","pointed-end actin filament capping","pole plasm RNA localization","pole plasm assembly","pole plasm mRNA localization","polyamine acetylation","polyamine biosynthetic process","polyamine catabolic process","polyamine metabolic process","polyamine transmembrane transport","polyamine transport","polyol biosynthetic process","polyol metabolic process","polyprenol biosynthetic process","polyprenol metabolic process","polysaccharide biosynthetic process","polysaccharide catabolic process","polysaccharide metabolic process","porphyrin-containing compound metabolic process","positive regulation of DNA metabolic process","positive regulation of DNA repair","positive regulation of DNA-binding transcription factor activity","positive regulation of DNA-templated transcription","positive regulation of DNA-templated transcription initiation","positive regulation of DNA-templated transcription, elongation","positive regulation of ERK1 and ERK2 cascade","positive regulation of GTPase activity","positive regulation of Golgi to plasma membrane protein transport","positive regulation of MAPK cascade","positive regulation of NF-kappaB transcription factor activity","positive regulation of RNA biosynthetic process","positive regulation of RNA metabolic process","positive regulation of Ras protein signal transduction","positive regulation of Rho protein signal transduction","positive regulation of T cell activation","positive regulation of T cell chemotaxis","positive regulation of T cell migration","positive regulation of T cell proliferation","positive regulation of TOR signaling","positive regulation of TORC1 signaling","positive regulation of Wnt signaling pathway","positive regulation of actin filament polymerization","positive regulation of adherens junction organization","positive regulation of amide metabolic process","positive regulation of angiogenesis","positive regulation of apoptotic process","positive regulation of apoptotic signaling pathway","positive regulation of attachment of mitotic spindle microtubules to kinetochore","positive regulation of attachment of spindle microtubules to kinetochore","positive regulation of autophagosome maturation","positive regulation of autophagy","positive regulation of axon extension","positive regulation of axon regeneration","positive regulation of axonogenesis","positive regulation of base-excision repair","positive regulation of biological process","positive regulation of biosynthetic process","positive regulation of blood circulation","positive regulation of blood coagulation","positive regulation of blood vessel endothelial cell migration","positive regulation of calcium ion transmembrane transport","positive regulation of calcium ion transmembrane transporter activity","positive regulation of calcium ion transport","positive regulation of canonical NF-kappaB signal transduction","positive regulation of canonical Wnt signaling pathway","positive regulation of catabolic process","positive regulation of catalytic activity","positive regulation of cation channel activity","positive regulation of cation transmembrane transport","positive regulation of cell activation","positive regulation of cell adhesion","positive regulation of cell communication","positive regulation of cell cycle","positive regulation of cell cycle process","positive regulation of cell development","positive regulation of cell differentiation","positive regulation of cell division","positive regulation of cell growth","positive regulation of cell junction assembly","positive regulation of cell migration","positive regulation of cell migration involved in sprouting angiogenesis","positive regulation of cell morphogenesis","positive regulation of cell motility","positive regulation of cell population proliferation","positive regulation of cell projection organization","positive regulation of cell-cell adhesion","positive regulation of cell-substrate adhesion","positive regulation of cellular biosynthetic process","positive regulation of cellular catabolic process","positive regulation of cellular component biogenesis","positive regulation of cellular component organization","positive regulation of cellular extravasation","positive regulation of cellular metabolic process","positive regulation of cellular process","positive regulation of cellular response to transforming growth factor beta stimulus","positive regulation of centrosome duplication","positive regulation of chemotaxis","positive regulation of chromosome segregation","positive regulation of coagulation","positive regulation of cytoplasmic translation","positive regulation of cytoskeleton organization","positive regulation of defense response","positive regulation of dendrite morphogenesis","positive regulation of dendritic spine development","positive regulation of dendritic spine morphogenesis","positive regulation of dephosphorylation","positive regulation of developmental growth","positive regulation of developmental process","positive regulation of early endosome to late endosome transport","positive regulation of endoplasmic reticulum tubular network organization","positive regulation of endothelial cell development","positive regulation of endothelial cell migration","positive regulation of epithelial cell migration","positive regulation of establishment of endothelial barrier","positive regulation of establishment of protein localization","positive regulation of establishment of protein localization to mitochondrion","positive regulation of extrinsic apoptotic signaling pathway","positive regulation of gene expression","positive regulation of growth","positive regulation of heart contraction","positive regulation of hematopoietic progenitor cell differentiation","positive regulation of hematopoietic stem cell differentiation","positive regulation of hemostasis","positive regulation of hormone secretion","positive regulation of hydrolase activity","positive regulation of immune response","positive regulation of immune system process","positive regulation of innate immune response","positive regulation of insulin secretion","positive regulation of insulin secretion involved in cellular response to glucose stimulus","positive regulation of intracellular protein transport","positive regulation of intracellular signal transduction","positive regulation of intracellular transport","positive regulation of ion transmembrane transporter activity","positive regulation of kinase activity","positive regulation of leukocyte activation","positive regulation of leukocyte cell-cell adhesion","positive regulation of leukocyte chemotaxis","positive regulation of leukocyte migration","positive regulation of leukocyte proliferation","positive regulation of lipid localization","positive regulation of lipid storage","positive regulation of locomotion","positive regulation of lymphocyte activation","positive regulation of lymphocyte chemotaxis","positive regulation of lymphocyte migration","positive regulation of lymphocyte proliferation","positive regulation of mRNA catabolic process","positive regulation of mRNA metabolic process","positive regulation of macroautophagy","positive regulation of macromolecule biosynthetic process","positive regulation of macromolecule metabolic process","positive regulation of macrophage activation","positive regulation of macrophage migration","positive regulation of metabolic process","positive regulation of mitotic cell cycle","positive regulation of molecular function","positive regulation of monoatomic ion transmembrane transport","positive regulation of monoatomic ion transport","positive regulation of mononuclear cell migration","positive regulation of mononuclear cell proliferation","positive regulation of multicellular organismal process","positive regulation of muscle cell differentiation","positive regulation of myeloid cell differentiation","positive regulation of nervous system development","positive regulation of neural precursor cell proliferation","positive regulation of neuroblast proliferation","positive regulation of neurogenesis","positive regulation of neuromuscular junction development","positive regulation of neuron projection development","positive regulation of neuron projection regeneration","positive regulation of neutrophil migration","positive regulation of nitrogen compound metabolic process","positive regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay","positive regulation of nuclear-transcribed mRNA poly(A) tail shortening","positive regulation of nucleobase-containing compound metabolic process","positive regulation of organelle organization","positive regulation of pattern recognition receptor signaling pathway","positive regulation of peptide hormone secretion","positive regulation of peptide secretion","positive regulation of peptidyl-tyrosine phosphorylation","positive regulation of phosphate metabolic process","positive regulation of phosphatidylinositol 3-kinase/protein kinase B signal transduction","positive regulation of phosphorus metabolic process","positive regulation of phosphorylation","positive regulation of plasma membrane bounded cell projection assembly","positive regulation of polyamine transmembrane transport","positive regulation of programmed cell death","positive regulation of proteasomal protein catabolic process","positive regulation of proteasomal ubiquitin-dependent protein catabolic process","positive regulation of protein catabolic process","positive regulation of protein dephosphorylation","positive regulation of protein exit from endoplasmic reticulum","positive regulation of protein kinase A signaling","positive regulation of protein kinase activity","positive regulation of protein localization","positive regulation of protein localization to cell periphery","positive regulation of protein localization to early endosome","positive regulation of protein localization to endosome","positive regulation of protein localization to membrane","positive regulation of protein localization to plasma membrane","positive regulation of protein metabolic process","positive regulation of protein modification process","positive regulation of protein phosphorylation","positive regulation of protein polymerization","positive regulation of protein secretion","positive regulation of protein targeting to mitochondrion","positive regulation of protein transport","positive regulation of protein tyrosine kinase activity","positive regulation of protein-containing complex assembly","positive regulation of protein-containing complex disassembly","positive regulation of proteolysis","positive regulation of proteolysis involved in protein catabolic process","positive regulation of pseudopodium assembly","positive regulation of response to biotic stimulus","positive regulation of response to external stimulus","positive regulation of response to stimulus","positive regulation of response to wounding","positive regulation of sarcomere organization","positive regulation of secretion","positive regulation of secretion by cell","positive regulation of sequestering of triglyceride","positive regulation of signal transduction","positive regulation of signaling","positive regulation of small GTPase mediated signal transduction","positive regulation of sprouting angiogenesis","positive regulation of stem cell differentiation","positive regulation of store-operated calcium channel activity","positive regulation of striated muscle cell differentiation","positive regulation of supramolecular fiber organization","positive regulation of synapse assembly","positive regulation of synaptic assembly at neuromuscular junction","positive regulation of synaptic transmission","positive regulation of toll-like receptor 3 signaling pathway","positive regulation of toll-like receptor 4 signaling pathway","positive regulation of transcription by RNA polymerase II","positive regulation of transcription elongation by RNA polymerase II","positive regulation of transcription initiation by RNA polymerase II","positive regulation of transferase activity","positive regulation of transforming growth factor beta receptor signaling pathway","positive regulation of translation","positive regulation of translational elongation","positive regulation of translational termination","positive regulation of transmembrane receptor protein serine/threonine kinase signaling pathway","positive regulation of transmembrane transport","positive regulation of transport","positive regulation of transporter activity","positive regulation of ubiquitin-dependent protein catabolic process","positive regulation of vasculature development","positive regulation of wound healing","post-Golgi vesicle-mediated transport","post-transcriptional gene silencing","post-transcriptional regulation of gene expression","post-translational protein modification","post-translational protein targeting to endoplasmic reticulum membrane","post-translational protein targeting to membrane, translocation","posterior lateral line development","posterior lateral line ganglion development","posterior lateral line nerve development","posterior lateral line neuromast development","posterior lateral line neuromast hair cell development","posterior lateral line neuromast hair cell differentiation","posterior lateral line neuromast primordium migration","posterior lateral line system development","postreplication repair","postsynapse organization","postsynaptic actin cytoskeleton organization","postsynaptic cytoskeleton organization","postsynaptic signal transduction","potassium ion homeostasis","potassium ion import across plasma membrane","potassium ion transmembrane transport","potassium ion transport","pre-miRNA processing","preganglionic parasympathetic fiber development","prenylation","presynaptic endocytosis","primary alcohol metabolic process","primary metabolic process","primary neural tube formation","primitive erythrocyte differentiation","primitive hemopoiesis","process utilizing autophagic mechanism","proepicardium development","programmed cell death","programmed necrotic cell death","proline biosynthetic process","proline metabolic process","prolyl-tRNA aminoacylation","pronephric nephron development","pronephric nephron tubule development","pronephric nephron tubule epithelial cell differentiation","pronephros development","pronephros morphogenesis","proteasomal protein catabolic process","proteasome assembly","proteasome-mediated ubiquitin-dependent protein catabolic process","protein K11-linked ubiquitination","protein K29-linked ubiquitination","protein K48-linked ubiquitination","protein K63-linked ubiquitination","protein N-linked glycosylation","protein O-linked glycosylation","protein O-linked mannosylation","protein acylation","protein alkylation","protein autophosphorylation","protein autoprocessing","protein autoubiquitination","protein carboxylation","protein catabolic process","protein catabolic process in the vacuole","protein complex oligomerization","protein deacylation","protein depalmitoylation","protein dephosphorylation","protein depolymerization","protein deubiquitination","protein exit from endoplasmic reticulum","protein export from nucleus","protein folding","protein glycosylation","protein heterooligomerization","protein heterotrimerization","protein homooligomerization","protein import into mitochondrial matrix","protein import into nucleus","protein import into peroxisome matrix","protein import into peroxisome matrix, docking","protein insertion into ER membrane","protein insertion into membrane","protein insertion into membrane from inner side","protein insertion into mitochondrial inner membrane from matrix","protein insertion into mitochondrial membrane","protein kinase A signaling","protein kinase C signaling","protein lipidation","protein localization","protein localization to basolateral plasma membrane","protein localization to cell junction","protein localization to cell periphery","protein localization to cell surface","protein localization to centrosome","protein localization to chromosome","protein localization to chromosome, centromeric region","protein localization to condensed chromosome","protein localization to cytoskeleton","protein localization to early endosome","protein localization to endoplasmic reticulum","protein localization to endosome","protein localization to extracellular region","protein localization to kinetochore","protein localization to membrane","protein localization to microtubule","protein localization to microtubule cytoskeleton","protein localization to microtubule end","protein localization to microtubule organizing center","protein localization to microtubule plus-end","protein localization to mitochondrion","protein localization to nuclear envelope","protein localization to nucleus","protein localization to organelle","protein localization to peroxisome","protein localization to phagophore assembly site","protein localization to plasma membrane","protein localization to postsynapse","protein localization to postsynaptic membrane","protein localization to synapse","protein localization to vacuole","protein mannosylation","protein maturation","protein maturation by iron-sulfur cluster transfer","protein metabolic process","protein methylation","protein modification by small protein conjugation","protein modification by small protein conjugation or removal","protein modification by small protein removal","protein modification process","protein myristoylation","protein neddylation","protein peptidyl-prolyl isomerization","protein phosphorylation","protein polymerization","protein polyubiquitination","protein prenylation","protein processing","protein processing involved in protein targeting to mitochondrion","protein refolding","protein secretion","protein splicing","protein stabilization","protein sumoylation","protein targeting","protein targeting to ER","protein targeting to membrane","protein targeting to mitochondrion","protein targeting to vacuole","protein transmembrane import into intracellular organelle","protein transmembrane transport","protein transport","protein trimerization","protein ubiquitination","protein-DNA complex assembly","protein-DNA complex organization","protein-RNA complex assembly","protein-RNA complex organization","protein-containing complex assembly","protein-containing complex disassembly","protein-containing complex localization","protein-containing complex organization","proteinogenic amino acid biosynthetic process","proteinogenic amino acid catabolic process","proteinogenic amino acid metabolic process","proteoglycan biosynthetic process","proteoglycan metabolic process","proteolysis","proteolysis involved in protein catabolic process","proton motive force-driven ATP synthesis","proton transmembrane transport","proton-transporting V-type ATPase complex assembly","proton-transporting two-sector ATPase complex assembly","pseudopodium assembly","pseudopodium organization","pseudouridine synthesis","pteridine-containing compound catabolic process","pteridine-containing compound metabolic process","purine nucleobase biosynthetic process","purine nucleobase metabolic process","purine nucleobase salvage","purine nucleoside biosynthetic process","purine nucleoside bisphosphate biosynthetic process","purine nucleoside bisphosphate metabolic process","purine nucleoside diphosphate biosynthetic process","purine nucleoside diphosphate catabolic process","purine nucleoside diphosphate metabolic process","purine nucleoside metabolic process","purine nucleoside monophosphate biosynthetic process","purine nucleoside monophosphate metabolic process","purine nucleoside triphosphate biosynthetic process","purine nucleoside triphosphate metabolic process","purine nucleotide biosynthetic process","purine nucleotide catabolic process","purine nucleotide metabolic process","purine nucleotide salvage","purine ribonucleoside biosynthetic process","purine ribonucleoside diphosphate biosynthetic process","purine ribonucleoside diphosphate catabolic process","purine ribonucleoside diphosphate metabolic process","purine ribonucleoside metabolic process","purine ribonucleoside monophosphate biosynthetic process","purine ribonucleoside monophosphate metabolic process","purine ribonucleoside salvage","purine ribonucleoside triphosphate biosynthetic process","purine ribonucleoside triphosphate metabolic process","purine ribonucleotide biosynthetic process","purine ribonucleotide catabolic process","purine ribonucleotide metabolic process","purine ribonucleotide salvage","purine-containing compound biosynthetic process","purine-containing compound catabolic process","purine-containing compound metabolic process","purine-containing compound salvage","purinergic nucleotide receptor signaling pathway","putrescine biosynthetic process","putrescine biosynthetic process from ornithine","putrescine catabolic process","putrescine metabolic process","pyridine nucleotide catabolic process","pyridine nucleotide metabolic process","pyridine-containing compound catabolic process","pyridine-containing compound metabolic process","pyrimidine deoxyribonucleoside monophosphate biosynthetic process","pyrimidine deoxyribonucleoside monophosphate metabolic process","pyrimidine deoxyribonucleoside triphosphate biosynthetic process","pyrimidine deoxyribonucleoside triphosphate catabolic process","pyrimidine deoxyribonucleoside triphosphate metabolic process","pyrimidine deoxyribonucleotide biosynthetic process","pyrimidine deoxyribonucleotide catabolic process","pyrimidine deoxyribonucleotide metabolic process","pyrimidine nucleobase biosynthetic process","pyrimidine nucleobase metabolic process","pyrimidine nucleoside monophosphate biosynthetic process","pyrimidine nucleoside monophosphate metabolic process","pyrimidine nucleoside transport","pyrimidine nucleoside triphosphate biosynthetic process","pyrimidine nucleoside triphosphate catabolic process","pyrimidine nucleoside triphosphate metabolic process","pyrimidine nucleotide biosynthetic process","pyrimidine nucleotide catabolic process","pyrimidine nucleotide metabolic process","pyrimidine ribonucleoside monophosphate biosynthetic process","pyrimidine ribonucleoside monophosphate metabolic process","pyrimidine ribonucleoside triphosphate biosynthetic process","pyrimidine ribonucleoside triphosphate metabolic process","pyrimidine ribonucleotide biosynthetic process","pyrimidine ribonucleotide metabolic process","pyrimidine-containing compound biosynthetic process","pyrimidine-containing compound catabolic process","pyrimidine-containing compound metabolic process","pyruvate family amino acid catabolic process","pyruvate family amino acid metabolic process","pyruvate metabolic process","quaternary ammonium group transport","quinone biosynthetic process","quinone metabolic process","rRNA 3'-end processing","rRNA base methylation","rRNA metabolic process","rRNA methylation","rRNA modification","rRNA processing","rRNA transcription","reactive oxygen species metabolic process","receptor clustering","receptor guanylyl cyclase signaling pathway","receptor localization to synapse","receptor signaling pathway via JAK-STAT","receptor signaling pathway via STAT","receptor-mediated endocytosis","receptor-mediated endocytosis involved in cholesterol transport","reciprocal homologous recombination","reciprocal meiotic recombination","recombinational repair","recruitment of 3'-end processing factors to RNA polymerase II holoenzyme complex","regeneration","regionalization","regulated exocytosis","regulation of ARF protein signal transduction","regulation of Arp2/3 complex-mediated actin nucleation","regulation of BMP signaling pathway","regulation of COPII vesicle coating","regulation of DNA biosynthetic process","regulation of DNA metabolic process","regulation of DNA methylation","regulation of DNA repair","regulation of DNA replication","regulation of DNA-binding transcription factor activity","regulation of DNA-templated transcription","regulation of DNA-templated transcription elongation","regulation of DNA-templated transcription initiation","regulation of ERBB signaling pathway","regulation of ERK1 and ERK2 cascade","regulation of G1/S transition of mitotic cell cycle","regulation of G2/M transition of mitotic cell cycle","regulation of GTPase activity","regulation of Golgi to plasma membrane protein transport","regulation of JNK cascade","regulation of MAPK cascade","regulation of Notch signaling pathway","regulation of RNA biosynthetic process","regulation of RNA metabolic process","regulation of RNA splicing","regulation of RNA stability","regulation of Rab protein signal transduction","regulation of Ras protein signal transduction","regulation of Rho protein signal transduction","regulation of SREBP signaling pathway","regulation of T cell activation","regulation of T cell chemotaxis","regulation of T cell differentiation","regulation of T cell migration","regulation of T cell proliferation","regulation of TOR signaling","regulation of TORC1 signaling","regulation of Toll signaling pathway","regulation of Wnt signaling pathway","regulation of actin cytoskeleton organization","regulation of actin filament depolymerization","regulation of actin filament length","regulation of actin filament organization","regulation of actin filament polymerization","regulation of actin filament-based process","regulation of actin nucleation","regulation of actin polymerization or depolymerization","regulation of action potential","regulation of activin receptor signaling pathway","regulation of actomyosin structure organization","regulation of adenylate cyclase activity","regulation of adherens junction organization","regulation of alcohol biosynthetic process","regulation of alternative mRNA splicing, via spliceosome","regulation of amide metabolic process","regulation of amyloid precursor protein biosynthetic process","regulation of anatomical structure morphogenesis","regulation of anatomical structure size","regulation of angiogenesis","regulation of apoptotic process","regulation of apoptotic signaling pathway","regulation of appetite","regulation of attachment of mitotic spindle microtubules to kinetochore","regulation of attachment of spindle microtubules to kinetochore","regulation of autophagosome maturation","regulation of autophagy","regulation of axon extension","regulation of axon extension involved in axon guidance","regulation of axon regeneration","regulation of axonogenesis","regulation of base-excision repair","regulation of behavior","regulation of biological process","regulation of biological quality","regulation of biomineral tissue development","regulation of biosynthetic process","regulation of blood circulation","regulation of blood coagulation","regulation of blood pressure","regulation of blood vessel endothelial cell migration","regulation of body fluid levels","regulation of bone mineralization","regulation of bone remodeling","regulation of bone resorption","regulation of cAMP-mediated signaling","regulation of calcium ion transmembrane transport","regulation of calcium ion transmembrane transporter activity","regulation of calcium ion transport","regulation of calcium ion-dependent exocytosis","regulation of canonical NF-kappaB signal transduction","regulation of canonical Wnt signaling pathway","regulation of carbohydrate biosynthetic process","regulation of carbohydrate metabolic process","regulation of cardiac muscle cell differentiation","regulation of cardiac muscle cell membrane repolarization","regulation of cardiac muscle contraction","regulation of cardiac muscle contraction by calcium ion signaling","regulation of cardiac muscle fiber development","regulation of cardiocyte differentiation","regulation of cartilage development","regulation of catabolic process","regulation of catalytic activity","regulation of cation channel activity","regulation of cell activation","regulation of cell adhesion","regulation of cell communication","regulation of cell cycle","regulation of cell cycle G1/S phase transition","regulation of cell cycle G2/M phase transition","regulation of cell cycle phase transition","regulation of cell cycle process","regulation of cell development","regulation of cell differentiation","regulation of cell division","regulation of cell growth","regulation of cell junction assembly","regulation of cell migration","regulation of cell migration involved in sprouting angiogenesis","regulation of cell morphogenesis","regulation of cell motility","regulation of cell population proliferation","regulation of cell projection assembly","regulation of cell projection organization","regulation of cell shape","regulation of cell size","regulation of cell-cell adhesion","regulation of cell-substrate adhesion","regulation of cellular biosynthetic process","regulation of cellular catabolic process","regulation of cellular component biogenesis","regulation of cellular component organization","regulation of cellular component size","regulation of cellular extravasation","regulation of cellular ketone metabolic process","regulation of cellular localization","regulation of cellular metabolic process","regulation of cellular process","regulation of cellular response to growth factor stimulus","regulation of cellular response to insulin stimulus","regulation of cellular response to stress","regulation of cellular response to transforming growth factor beta stimulus","regulation of centrosome cycle","regulation of centrosome duplication","regulation of chemotaxis","regulation of cholesterol biosynthetic process","regulation of cholesterol metabolic process","regulation of chondrocyte differentiation","regulation of chromatin organization","regulation of chromosome organization","regulation of chromosome segregation","regulation of cilium assembly","regulation of cilium movement","regulation of circadian rhythm","regulation of coagulation","regulation of cyclase activity","regulation of cyclin-dependent protein kinase activity","regulation of cyclin-dependent protein serine/threonine kinase activity","regulation of cytokine production","regulation of cytoplasmic pattern recognition receptor signaling pathway","regulation of cytoplasmic translation","regulation of cytoskeleton organization","regulation of cytosolic calcium ion concentration","regulation of defense response","regulation of dendrite development","regulation of dendrite morphogenesis","regulation of dendritic spine development","regulation of dendritic spine morphogenesis","regulation of dephosphorylation","regulation of developmental growth","regulation of developmental process","regulation of early endosome to late endosome transport","regulation of embryonic development","regulation of endocytosis","regulation of endoplasmic reticulum tubular network organization","regulation of endothelial cell development","regulation of endothelial cell differentiation","regulation of endothelial cell migration","regulation of epidermal growth factor receptor signaling pathway","regulation of epithelial cell differentiation","regulation of epithelial cell migration","regulation of epithelial to mesenchymal transition","regulation of establishment of endothelial barrier","regulation of establishment of protein localization","regulation of establishment of protein localization to mitochondrion","regulation of exocytosis","regulation of extent of cell growth","regulation of extrinsic apoptotic signaling pathway","regulation of fatty acid beta-oxidation","regulation of fatty acid metabolic process","regulation of fatty acid oxidation","regulation of fibroblast growth factor receptor signaling pathway","regulation of filopodium assembly","regulation of gene expression","regulation of gene silencing by regulatory ncRNA","regulation of glucose metabolic process","regulation of glycoprotein biosynthetic process","regulation of glycoprotein metabolic process","regulation of granulocyte differentiation","regulation of growth","regulation of heart contraction","regulation of heart rate","regulation of hematopoietic progenitor cell differentiation","regulation of hematopoietic stem cell differentiation","regulation of hemopoiesis","regulation of hemostasis","regulation of hormone levels","regulation of hormone secretion","regulation of hyaluronan biosynthetic process","regulation of hydrolase activity","regulation of immune response","regulation of immune system process","regulation of inflammatory response","regulation of innate immune response","regulation of insulin receptor signaling pathway","regulation of insulin secretion","regulation of insulin secretion involved in cellular response to glucose stimulus","regulation of insulin-like growth factor receptor signaling pathway","regulation of interleukin-13 production","regulation of intracellular protein transport","regulation of intracellular signal transduction","regulation of intracellular transport","regulation of intrinsic apoptotic signaling pathway","regulation of intrinsic apoptotic signaling pathway by p53 class mediator","regulation of kainate selective glutamate receptor activity","regulation of kinase activity","regulation of lamellipodium assembly","regulation of lamellipodium morphogenesis","regulation of lamellipodium organization","regulation of late endosome to lysosome transport","regulation of leukocyte activation","regulation of leukocyte cell-cell adhesion","regulation of leukocyte chemotaxis","regulation of leukocyte differentiation","regulation of leukocyte migration","regulation of leukocyte proliferation","regulation of lipid biosynthetic process","regulation of lipid catabolic process","regulation of lipid localization","regulation of lipid metabolic process","regulation of lipid storage","regulation of localization","regulation of locomotion","regulation of long-term neuronal synaptic plasticity","regulation of lyase activity","regulation of lymphocyte activation","regulation of lymphocyte chemotaxis","regulation of lymphocyte differentiation","regulation of lymphocyte migration","regulation of lymphocyte proliferation","regulation of mRNA catabolic process","regulation of mRNA metabolic process","regulation of mRNA processing","regulation of mRNA splicing, via spliceosome","regulation of mRNA stability","regulation of macroautophagy","regulation of macromolecule biosynthetic process","regulation of macromolecule metabolic process","regulation of macrophage activation","regulation of macrophage migration","regulation of membrane lipid distribution","regulation of membrane potential","regulation of membrane repolarization","regulation of metabolic process","regulation of metal ion transport","regulation of metaphase plate congression","regulation of metaphase/anaphase transition of cell cycle","regulation of miRNA processing","regulation of microtubule cytoskeleton organization","regulation of microtubule depolymerization","regulation of microtubule polymerization or depolymerization","regulation of microtubule-based movement","regulation of microtubule-based process","regulation of mitochondrial mRNA stability","regulation of mitotic cell cycle","regulation of mitotic cell cycle phase transition","regulation of mitotic metaphase/anaphase transition","regulation of mitotic nuclear division","regulation of mitotic sister chromatid segregation","regulation of mitotic spindle assembly","regulation of mitotic spindle organization","regulation of molecular function","regulation of monoatomic cation transmembrane transport","regulation of monoatomic ion transmembrane transport","regulation of monoatomic ion transmembrane transporter activity","regulation of monoatomic ion transport","regulation of mononuclear cell migration","regulation of mononuclear cell proliferation","regulation of multicellular organismal development","regulation of multicellular organismal process","regulation of muscle cell differentiation","regulation of muscle contraction","regulation of muscle system process","regulation of myeloid cell differentiation","regulation of myeloid leukocyte differentiation","regulation of myoblast fusion","regulation of nervous system development","regulation of nervous system process","regulation of neural crest formation","regulation of neural precursor cell proliferation","regulation of neuroblast proliferation","regulation of neurogenesis","regulation of neuroinflammatory response","regulation of neuromuscular junction development","regulation of neuron differentiation","regulation of neuron projection development","regulation of neuron projection regeneration","regulation of neuronal action potential","regulation of neuronal synaptic plasticity","regulation of neurotransmitter receptor activity","regulation of neutrophil migration","regulation of nitrogen compound metabolic process","regulation of nodal signaling pathway","regulation of nuclear division","regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay","regulation of nuclear-transcribed mRNA poly(A) tail shortening","regulation of nucleobase-containing compound metabolic process","regulation of organ growth","regulation of organelle assembly","regulation of organelle organization","regulation of ossification","regulation of pancreatic A cell differentiation","regulation of pattern recognition receptor signaling pathway","regulation of peptidase activity","regulation of peptide hormone secretion","regulation of peptide secretion","regulation of peptide transport","regulation of peptidyl-tyrosine phosphorylation","regulation of phosphatase activity","regulation of phosphate metabolic process","regulation of phosphatidylinositol 3-kinase/protein kinase B signal transduction","regulation of phosphoprotein phosphatase activity","regulation of phosphorus metabolic process","regulation of phosphorylation","regulation of plasma lipoprotein particle levels","regulation of plasma membrane bounded cell projection assembly","regulation of plasma membrane bounded cell projection organization","regulation of polyamine transmembrane transport","regulation of polysaccharide biosynthetic process","regulation of polysaccharide metabolic process","regulation of postsynapse organization","regulation of postsynaptic membrane neurotransmitter receptor levels","regulation of postsynaptic membrane potential","regulation of potassium ion transmembrane transport","regulation of potassium ion transmembrane transporter activity","regulation of potassium ion transport","regulation of primary metabolic process","regulation of programmed cell death","regulation of proteasomal protein catabolic process","regulation of proteasomal ubiquitin-dependent protein catabolic process","regulation of protein ADP-ribosylation","regulation of protein catabolic process","regulation of protein dephosphorylation","regulation of protein depolymerization","regulation of protein exit from endoplasmic reticulum","regulation of protein glycosylation","regulation of protein kinase A signaling","regulation of protein kinase C signaling","regulation of protein kinase activity","regulation of protein localization","regulation of protein localization to cell periphery","regulation of protein localization to early endosome","regulation of protein localization to endosome","regulation of protein localization to membrane","regulation of protein localization to plasma membrane","regulation of protein metabolic process","regulation of protein modification process","regulation of protein phosphorylation","regulation of protein polymerization","regulation of protein secretion","regulation of protein serine/threonine kinase activity","regulation of protein stability","regulation of protein targeting","regulation of protein targeting to mitochondrion","regulation of protein transport","regulation of protein tyrosine kinase activity","regulation of protein-containing complex assembly","regulation of protein-containing complex disassembly","regulation of proteolysis","regulation of proteolysis involved in protein catabolic process","regulation of pseudopodium assembly","regulation of regulated secretory pathway","regulation of regulatory ncRNA processing","regulation of response to biotic stimulus","regulation of response to external stimulus","regulation of response to extracellular stimulus","regulation of response to food","regulation of response to nutrient levels","regulation of response to stimulus","regulation of response to stress","regulation of response to wounding","regulation of sarcomere organization","regulation of secretion","regulation of secretion by cell","regulation of sequestering of calcium ion","regulation of sequestering of triglyceride","regulation of siRNA processing","regulation of signal transduction","regulation of signal transduction by p53 class mediator","regulation of signaling","regulation of signaling receptor activity","regulation of sister chromatid segregation","regulation of small GTPase mediated signal transduction","regulation of small molecule metabolic process","regulation of smooth muscle cell proliferation","regulation of smoothened signaling pathway","regulation of sodium ion transmembrane transport","regulation of sodium ion transmembrane transporter activity","regulation of sodium ion transport","regulation of spindle assembly","regulation of spindle organization","regulation of sprouting angiogenesis","regulation of stem cell differentiation","regulation of stem cell population maintenance","regulation of steroid biosynthetic process","regulation of steroid metabolic process","regulation of sterol biosynthetic process","regulation of store-operated calcium channel activity","regulation of striated muscle cell differentiation","regulation of striated muscle contraction","regulation of substrate adhesion-dependent cell spreading","regulation of supramolecular fiber organization","regulation of synapse assembly","regulation of synapse organization","regulation of synapse structure or activity","regulation of synaptic assembly at neuromuscular junction","regulation of synaptic plasticity","regulation of synaptic transmission, GABAergic","regulation of synaptic vesicle cycle","regulation of syncytium formation by plasma membrane fusion","regulation of system process","regulation of telomere maintenance","regulation of telomere maintenance via telomerase","regulation of telomere maintenance via telomere lengthening","regulation of tissue remodeling","regulation of toll-like receptor 3 signaling pathway","regulation of toll-like receptor 4 signaling pathway","regulation of trans-synaptic signaling","regulation of transcription by RNA polymerase II","regulation of transcription elongation by RNA polymerase II","regulation of transcription initiation by RNA polymerase II","regulation of transferase activity","regulation of transforming growth factor beta receptor signaling pathway","regulation of translation","regulation of translational elongation","regulation of translational fidelity","regulation of translational initiation","regulation of translational termination","regulation of transmembrane receptor protein serine/threonine kinase signaling pathway","regulation of transmembrane transport","regulation of transmembrane transporter activity","regulation of transmission of nerve impulse","regulation of transport","regulation of transporter activity","regulation of tube diameter","regulation of tube size","regulation of type B pancreatic cell development","regulation of ubiquitin-dependent protein catabolic process","regulation of vacuolar transport","regulation of vascular associated smooth muscle cell proliferation","regulation of vascular permeability","regulation of vasculature development","regulation of ventricular cardiac muscle cell membrane repolarization","regulation of vesicle-mediated transport","regulation of viral process","regulation of viral transcription","regulation of wound healing","regulatory ncRNA processing","regulatory ncRNA-mediated gene silencing","regulatory ncRNA-mediated post-transcriptional gene silencing","relaxation of muscle","relaxation of skeletal muscle","release of sequestered calcium ion into cytosol","renal system development","renal tubule development","replication fork processing","reproduction","reproductive process","resolution of meiotic recombination intermediates","respiratory chain complex III assembly","respiratory chain complex IV assembly","respiratory electron transport chain","response to ATP","response to BMP","response to abiotic stimulus","response to acetylcholine","response to activity","response to alcohol","response to amino acid starvation","response to auditory stimulus","response to axon injury","response to bacterium","response to biotic stimulus","response to cadmium ion","response to carbohydrate","response to catecholamine","response to chemical","response to chemokine","response to copper ion","response to corticotropin-releasing hormone","response to cytokine","response to decreased oxygen levels","response to dopamine","response to endogenous stimulus","response to endoplasmic reticulum stress","response to estrogen","response to ethanol","response to external biotic stimulus","response to external stimulus","response to extracellular stimulus","response to fatty acid","response to fibroblast growth factor","response to food","response to fungus","response to glucagon","response to glucose","response to growth factor","response to growth hormone","response to heat","response to hexose","response to hormone","response to hypoxia","response to inorganic substance","response to insulin","response to interleukin-1","response to iron ion","response to iron(II) ion","response to light stimulus","response to lipid","response to lipopolysaccharide","response to mechanical stimulus","response to metal ion","response to misfolded protein","response to molecule of bacterial origin","response to monoamine","response to monosaccharide","response to muscle activity","response to nicotine","response to nitrogen compound","response to nutrient","response to nutrient levels","response to organic cyclic compound","response to organic substance","response to organonitrogen compound","response to organophosphorus","response to osmotic stress","response to other organism","response to oxidative stress","response to oxygen levels","response to oxygen-containing compound","response to peptide","response to peptide hormone","response to prostaglandin","response to purine-containing compound","response to radiation","response to salt stress","response to starvation","response to steroid hormone","response to sterol","response to sterol depletion","response to stimulus","response to stress","response to temperature stimulus","response to topologically incorrect protein","response to toxic substance","response to transforming growth factor beta","response to tumor necrosis factor","response to type II interferon","response to unfolded protein","response to virus","response to vitamin","response to vitamin B3","response to wounding","response to xenobiotic stimulus","response to yeast","retina development in camera-type eye","retina homeostasis","retina layer formation","retina morphogenesis in camera-type eye","retinal cone cell development","retinal cone cell differentiation","retinal ganglion cell axon guidance","retinal pigment epithelium development","retinal rod cell development","retinal rod cell differentiation","retinoic acid receptor signaling pathway","retinoid metabolic process","retinol metabolic process","retrograde transport, endosome to Golgi","retrograde vesicle-mediated transport, Golgi to endoplasmic reticulum","rhythmic process","ribonucleoprotein complex biogenesis","ribonucleoside biosynthetic process","ribonucleoside bisphosphate biosynthetic process","ribonucleoside bisphosphate metabolic process","ribonucleoside diphosphate biosynthetic process","ribonucleoside diphosphate catabolic process","ribonucleoside diphosphate metabolic process","ribonucleoside metabolic process","ribonucleoside monophosphate biosynthetic process","ribonucleoside monophosphate metabolic process","ribonucleoside triphosphate biosynthetic process","ribonucleoside triphosphate metabolic process","ribonucleotide biosynthetic process","ribonucleotide catabolic process","ribonucleotide metabolic process","ribose phosphate biosynthetic process","ribose phosphate metabolic process","ribosomal large subunit assembly","ribosomal large subunit biogenesis","ribosomal small subunit assembly","ribosomal small subunit biogenesis","ribosome assembly","ribosome biogenesis","ribosome disassembly","ruffle assembly","ruffle organization","sarcomere organization","sarcomerogenesis","secondary alcohol biosynthetic process","secondary alcohol metabolic process","secretion","secretion by cell","secretion of lysosomal enzymes","segment specification","segmentation","semaphorin-plexin signaling pathway","semaphorin-plexin signaling pathway involved in axon guidance","semaphorin-plexin signaling pathway involved in neuron projection guidance","semicircular canal development","semicircular canal formation","semicircular canal morphogenesis","sensory epithelium regeneration","sensory organ development","sensory organ morphogenesis","sensory perception","sensory perception of chemical stimulus","sensory perception of light stimulus","sensory perception of mechanical stimulus","sensory perception of sound","sensory system development","septum transversum development","sequestering of actin monomers","sequestering of calcium ion","sequestering of iron ion","sequestering of metal ion","sequestering of triglyceride","serine family amino acid biosynthetic process","serine family amino acid catabolic process","serine family amino acid metabolic process","serine transport","seryl-tRNA aminoacylation","sex differentiation","sexual reproduction","siRNA processing","signal peptide processing","signal release","signal release from synapse","signal transduction","signal transduction by p53 class mediator","signal transduction in absence of ligand","signal transduction in response to DNA damage","signaling","signaling receptor ligand precursor processing","sister chromatid cohesion","sister chromatid segregation","skeletal muscle contraction","skeletal muscle fiber development","skeletal muscle myosin thick filament assembly","skeletal muscle organ development","skeletal muscle thin filament assembly","skeletal muscle tissue development","skeletal myofibril assembly","skeletal system development","skeletal system morphogenesis","slow-twitch skeletal muscle fiber contraction","small GTPase-mediated signal transduction","small molecule biosynthetic process","small molecule catabolic process","small molecule metabolic process","smooth muscle cell proliferation","smooth muscle contraction","smoothened signaling pathway","snRNA 3'-end processing","snRNA metabolic process","snRNA processing","snRNA transcription","snRNA transcription by RNA polymerase II","snRNA transcription by RNA polymerase III","sodium ion export across plasma membrane","sodium ion homeostasis","sodium ion transmembrane transport","sodium ion transport","sodium-dependent phosphate transport","somite development","somite specification","somitogenesis","specification of symmetry","spermatogenesis","spermidine acetylation","spermidine metabolic process","sphingolipid biosynthetic process","sphingolipid catabolic process","sphingolipid metabolic process","sphingomyelin metabolic process","spinal cord development","spinal cord motor neuron cell fate specification","spinal cord motor neuron differentiation","spinal cord oligodendrocyte cell differentiation","spinal cord patterning","spindle assembly","spindle localization","spindle organization","spliceosomal complex assembly","spliceosomal snRNP assembly","sprouting angiogenesis","stem cell development","stem cell differentiation","stem cell fate commitment","stem cell fate specification","stem cell population maintenance","steroid biosynthetic process","steroid hormone mediated signaling pathway","steroid metabolic process","sterol biosynthetic process","sterol homeostasis","sterol metabolic process","sterol transport","stress granule assembly","stress response to cadmium ion","stress response to metal ion","striated muscle cell development","striated muscle cell differentiation","striated muscle contraction","striated muscle myosin thick filament assembly","striated muscle tissue development","substrate adhesion-dependent cell spreading","sucrose transport","sulfate assimilation","sulfate transport","sulfur amino acid biosynthetic process","sulfur amino acid metabolic process","sulfur compound biosynthetic process","sulfur compound metabolic process","sulfur compound transport","supramolecular fiber organization","swim bladder development","swim bladder inflation","swim bladder maturation","swimming behavior","synapse assembly","synapse organization","synaptic assembly at neuromuscular junction","synaptic signaling","synaptic transmission, GABAergic","synaptic transmission, cholinergic","synaptic vesicle cycle","synaptic vesicle endocytosis","synaptic vesicle endosomal processing","synaptic vesicle exocytosis","synaptic vesicle priming","synaptic vesicle recycling","synaptic vesicle recycling via endosome","syncytium formation","syncytium formation by plasma membrane fusion","system development","system process","tRNA 5'-end processing","tRNA 5'-leader removal","tRNA aminoacylation","tRNA aminoacylation for protein translation","tRNA metabolic process","tRNA methylation","tRNA modification","tRNA processing","tRNA splicing, via endonucleolytic cleavage and ligation","tRNA threonylcarbamoyladenosine metabolic process","tRNA transcription","tRNA transcription by RNA polymerase III","tRNA wobble base modification","tRNA wobble uridine modification","tail-anchored membrane protein insertion into ER membrane","taurine transport","taxis","telencephalon development","telomere maintenance","telomere maintenance via telomerase","telomere maintenance via telomere lengthening","telomere organization","termination of RNA polymerase III transcription","terpenoid biosynthetic process","terpenoid metabolic process","tetrahydrofolate interconversion","tetrahydrofolate metabolic process","tetrapyrrole metabolic process","thigmotaxis","thioester biosynthetic process","thioester metabolic process","threonine transport","threonyl-tRNA aminoacylation","thymus development","thyroid gland development","thyroid hormone mediated signaling pathway","tight junction assembly","tight junction organization","tissue development","tissue homeostasis","tissue migration","tissue morphogenesis","tissue regeneration","tissue remodeling","toll-like receptor 3 signaling pathway","toll-like receptor 4 signaling pathway","tooth mineralization","trans-synaptic signaling","transcription by RNA polymerase I","transcription by RNA polymerase II","transcription by RNA polymerase III","transcription elongation by RNA polymerase II","transcription initiation at RNA polymerase I promoter","transcription initiation at RNA polymerase II promoter","transcription preinitiation complex assembly","transcription-coupled nucleotide-excision repair","transforming growth factor beta receptor signaling pathway","transforming growth factor beta receptor superfamily signaling pathway","transition metal ion transport","translation","translational elongation","translational initiation","translational termination","translesion synthesis","transmembrane receptor protein serine/threonine kinase signaling pathway","transmembrane receptor protein tyrosine kinase signaling pathway","transmembrane transport","transmission of nerve impulse","transport","transport along microtubule","transsulfuration","tricarboxylic acid cycle","tricarboxylic acid metabolic process","triglyceride biosynthetic process","triglyceride metabolic process","tripartite regional subdivision","tripeptide transmembrane transport","tripeptide transport","triterpenoid biosynthetic process","triterpenoid metabolic process","tryptophanyl-tRNA aminoacylation","tube closure","tube development","tube formation","tube morphogenesis","twitch skeletal muscle contraction","type B pancreatic cell development","type B pancreatic cell differentiation","tyrosyl-tRNA aminoacylation","ubiquinone biosynthetic process","ubiquinone metabolic process","ubiquitin-dependent protein catabolic process","ubiquitin-dependent protein catabolic process via the multivesicular body sorting pathway","unsaturated fatty acid biosynthetic process","unsaturated fatty acid metabolic process","urea cycle","urea metabolic process","uridine transport","vacuolar proton-transporting V-type ATPase complex assembly","vacuolar transport","vacuole organization","vagus nerve development","valyl-tRNA aminoacylation","vascular associated smooth muscle cell proliferation","vascular process in circulatory system","vasculature development","vasculogenesis","vasoconstriction","venous blood vessel development","venous endothelial cell differentiation","venous endothelial cell fate commitment","ventral aorta development","ventral midline development","ventral spinal cord development","ventricular cardiac muscle cell membrane repolarization","ventricular system development","very long-chain fatty acid metabolic process","vesicle budding from membrane","vesicle cargo loading","vesicle coating","vesicle cytoskeletal trafficking","vesicle fusion","vesicle localization","vesicle organization","vesicle targeting","vesicle targeting, rough ER to cis-Golgi","vesicle targeting, to, from or within Golgi","vesicle targeting, trans-Golgi to endosome","vesicle transport along actin filament","vesicle transport along microtubule","vesicle-mediated transport","vesicle-mediated transport between endosomal compartments","vesicle-mediated transport in synapse","vesicle-mediated transport to the plasma membrane","viral gene expression","viral life cycle","viral process","viral transcription","virion attachment to host cell","visual behavior","visual perception","visual system development","vitamin A metabolic process","vitamin D biosynthetic process","vitamin D metabolic process","vitamin D3 metabolic process","vitamin biosynthetic process","vitamin metabolic process","voluntary skeletal muscle contraction","water transport","wound healing","xenobiotic metabolic process","xenobiotic transport","zinc ion transport"],[1,1,1,3,3,2,1,1,2,2,5,1,1,1,3,1,4,5,1,3,1,2,3,8,10,12,1,21,12,13,20,4,4,14,23,25,6,13,3,4,5,13,3,1,1,14,3,8,1,12,13,1,6,3,3,11,2,1,1,1,2,4,1,1,1,1,1,23,15,15,13,30,18,5,5,13,8,31,6,5,2,1,6,3,37,1,1,3,3,3,58,5,3,1,1,6,10,32,55,1,4,1,2,1,1,1,3,10,2,198,3,5,1,5,8,5,3,2,1,5,4,3,28,1,19,39,10,12,3,9,8,9,1,7,6,11,11,22,1,14,2,2,1,1,23,1,1,15,6,2,1,1,1,1,1,1,1,1,1,1,2,7,7,8,11,7,4,6,4,2,3,1,2,1,5,7,3,1,1,3,1,13,5,18,4,9,1,5,3,3,2,6,16,20,7,2,1,707,2,1,1,1,1,2,6,9,4,14,6,1,1,2,3,3,1,1,1,1,1,1,1,2,2,2,1,3,1,9,20,14,38,2,13,176,4,10,1,13,2,21,4,10,1,1,1,1,1,4,1,2,11,13,4,8,3,1,8,15,1,18,2,14,15,11,2,3,2,14,2,6,34,1,1,24,3,5,9,14,2,1,7,2,4,13,4,2,18,3,10,8,36,10,9,7,5,7,1,200,12,9,4,1,6,1,309,4,9,5,1,1,1,1,83,41,2,24,24,2,3,1,1,1,3,20,8,12,1,43,13,2,2,1,2,1,4,3,28,2,2,3,2,1,12,23,2,5,9,1,4,2,1,3,1,2,1,5,14,5,2,1,3,2,23,4,1,5,3,6,1,1,1,1,1,20,7,1,5,7,16,6,2,1,221,4,6,39,1,1,1,1,2,2,3,1,2,46,2,2,6,8,27,12,3,3,48,1,8,14,2,1,3,1,1,2,1,1,3,10,1,1,2,4,1,1,1,1,2,1,2,1,5,4,1,1,1,2,2,1,7,14,1,1,2,2,7,13,1,1,1,2,9,4,16,5,12,121,133,5,2,1,6,20,9,4,2,38,1,6,10,1,2,1,1,2,3,1,1,1,1,1,1,1,293,52,5,3,125,1,18,146,1,1,44,1,1,1,1,1,1,1,1,1,5,15,1,2,1,1,1,2,1,1,1,1,2,1,3,19,6,4,2,1,1,1,2,1,1,1,1,2,1,1,220,21,2,2,1,1,1,1,3,4,1,2,1,1,2,4,1,1,1,1,4,1,4,13,2,1,2,1,26,3,7,2,10,1,1,1,1,23,5,1,1,1,1,1,1,1,1,11,1,4,3,1,2,6,1,19,1,556,1509,3,539,1,13,3,21,4,1,1,1,1,18,1,1,3,1,1,1,20,1,1,1,1,1,2,1,1,1,1,2,9,1,1,1,2,13,14,3,1,3,49,17,3,7,13,1,4,5,39,8,51,2,4,32,2,11,8,36,6,19,3,1,1,1,1,3,4,1,4,1,4,1,1,4,1,1,2,1,2,1,1,1,13,2,98,1,1,1,9,2,43,1,5,238,1,62,6,2,4,10,39,19,93,115,2,2,1,14,8,1,6,7,16,19,2,28,5,1,1,38,23,29,24,16,28,50,1,1,1,1,1,92,1,20,3,13,2,8,9,46,22,5,8,1,1,2,36,396,1,521,52,128,9,141,17,1,276,289,4,1,115,1,20,1,3,8,45,116,87,5,771,5,1,8,339,21,491,3,1,1299,2,16,6,3,1,1,3,1,2,36,1,2,1,4,2,33,1,2,2,2,1,1,16,1,1,1,15,6,2,6,3,1,3,2,1,13,2,5,25,11,1,18,9,9,2,2,1,2,268,47,2,4,7,2,2,1,2,35,1,3,5,6,2,2,1,5,4,1,5,3,8,2,3,1,4,25,15,3,2,6,4,8,4,1,6,2,2,1,1,1,1,1,33,38,26,1,6,4,19,20,7,1,9,2,49,13,1,3,1,2,1,1,3,1,1,1,3,1,1,1,15,6,3,2,2,1,1,4,2,6,1,2,15,1,1,1,2,1,1,1,1,6,2,4,2,1,1,2,1,9,1,1,73,2,4,2,1,1,1,1,1,1,1,1,14,3,11,9,4,1,1,4,2,2,2,2,2,2,2,1,2,2,1,2,21,4,4,2,2,1,4,1,1,13,1,11,2,13,5,1,1,7,22,7,1,3,2,298,4,1,1,1,1,1,1,2,1,1,1,1,4,4,1,3,6,3,5,1,1,1,1,2,1,2,1,2,1,1,1,1,4,5,1,5,9,4,2,1,1,1,1,14,64,33,1,1,3,3,3,13,2,3,11,11,4,39,1,30,25,4,2,15,13,8,1,4,5,3,24,2,1,1,17,1,1,7,7,1,1,1,8,1,2,1,1,2,3,1,1,1,1,3,21,5,3,3,1,36,4,2,7,1,7,3,1,5,18,2,1,1,1,1,5,3,14,54,2,2,1,1,4,9,9,3,4,1,246,79,1,1,12,1,73,8,3,15,1,4,22,2,3,2,2,1,6,3,3,6,1,3,1,1,3,4,4,1,1,7,1,1,11,11,2,11,11,2,1,54,20,3,4,1,1,1,2,1,1,2,3,1,2,3,1,1,2,8,3,7,1,1,1,2,2,2,3,1,6,1,1,1,1,1,2,5,2,2,2,1,1,3,1,1,1,1,3,3,1,2,1,1,2,1,3,1,11,2,441,47,27,2,11,1,3,3,7,7,3,1,5,4,3,1,1,1,1,3,2,5,2,1,2,2,1,1,1,9,12,9,12,1,1,3,1,5,4,2,6,4,18,19,2,3,2,2,4,3,3,14,1,1,2,1,1,3,5,5,22,1,1,1,1,1,1,7,1,1,21,8,30,1,3,10,14,8,1,1,7,3,2,1,1,1,2,1,1,25,3,2,8,1,1,1,2,2,223,21,396,1,1,4,4,1,1,12,41,1,2,12,1,3,2,2,5,2,1,1,3,3,1,1,1,1,1,1,3,17,1,1,1,1,1,41,4,28,5,3,5,9,1,1,1,9,4,1,2,1,2,4,9,2,25,18,2,29,1,2,4,2,1,3,3,1,1,1,2,1,2,2,5,1,1,1,1,1,9,19,1,2,6,1,2,1,1,17,17,1,7,47,8,3,90,52,1,1,1,67,1,1,1,1,4,3,1,2,1,7,5,1,1,4,8,2,1,1,9,2,2,2,2,1,2,1,3,2,2,1,10,2,1,12,2,13,28,1,1,1,1,1,1,1,8,1,1,5,9,1,9,1,3,1,1,33,9,1,1,2,1,21,62,8,4,4,2,18,2,1,1,3,1,6,1,8,257,23,16,5,1,1,2,2,1,1,1,3,3,3,1,1,5,4,1,4,4,1,2,1,1,1,3,11,4,53,1,2,58,2,24,2,12,2,23,6,467,57,3,3,14,108,718,11,193,2,2,1,1,1,1,1,1,1,10,6,1,1,2,2,1,1,1,1,1,1,1,1,1,3,1,3,2,7,1,1,1,6,1,1,2,3,1,6,4,1,3,1,1,1,1,1,1,1,1,1,7,4,12,22,1,1,1,12,6,14,1,1,6,2,2,2,1,903,33,5,4,1,1,1,1,1,18,3,3,3,1,3,27,6,3,4,1,5,6,10,7,34,3,2,4,7,2,1,2,3,2,3,4,1,1,5,2,2,2,4,2,8,1,1,1,3,4,7,17,1,1,2,1,1,1,1,36,3,8,24,6,2,2,10,9,2,5,35,35,2,4,6,11,18,24,36,19,37,60,2,3,13,8,2,4,4,1,4,1,1,2,2,1,6,27,5,3,2,236,3,5,286,3,17,1,14,16,1,18,7,1,24,18,7,1,5,1,3,2,1,4,14,10,3,6,6,1,9,1,1,3,4,1,48,24,4,1,1,3,1,1,1,37,1,1,1,1,2,1,1,37,2,39,1,1,1,1,2,1,2,7,6,1,1,8,1,3,7,1,1,2,1,2,124,62,1,3,2,2,1,3,4,1,3,23,8,1,1,4,4,3,4,2,2,3,6,2,1,1,62,2,10,63,116,4,1,1,1,1,1,7,1,2,7,1,1,26,2,1,1,2,1,1,1,1,1,1,1,10,1,1,1,4,1,2,3,1,1,62,65,1,65,7,2,5,9,1,2,2,1,2,51,1,39,8,1,1,1,1,1,7,1,7,12,1,1,6,6,7,1,4,1,1,1,27,1,1,23,1,23,2,2,1,1,7,1,1,1,24,1,4,8,1,8,1,3,1,2,2,2,1,6,82,43,7,12,3,3,6,5,1,2,6,1,3,2,1,52,1,1,7,3,5,1,2,1,1,6,1,42,46,3,1,2,2,33,7,2,1,10,1,26,3,1,1,1,1,1,2,1,2,2,2,2,2,4,4,1,1,3,2,1,5,5,776,88,1,2,3,25,4,2,2,13,13,3,5,1,1,1,11,4,3,2,359,3,3,4,220,19,391,5,3,30,11,3,2,4,1,5,6,3,7,8,1,17,7,27,3,2,2,6,1,11,8,8,17,6,25,3,5,1,1,1,3,1,3,1,1,1,1,1,2,1,3,1,1,1,3,1,1,1,2,2,1,2,3,2,1,1,1,2,38,3,13,1,12,1,149,1,11,8,38,6,19,20,2,230,21,405,8,15,4,533,81,855,106,1,171,61,445,32,8,5,49,3,1,4,1,2,1,1,5,1,1,1,11,36,3,6,10,1,1,1,1,22,2,113,2,1,2,119,2,5,11,1,1,1,1,1,4,1,1,3,2,2,1,2,5,1,1,1,2,2,2,1,1,1,1,1,1,1,1,2,1,1,1,173,1,1,1,2,3,4,2,6,3,1,2,1,1,1,1,2,1,13,1,3,21,1,5,174,111,1,3,4,1,1,2,1,1,1,1,1,1,5,1,1,15,27,48,5,1,1,1,1,1,1,1,1,1,1,2,1,5,1,1,2,3,1,2,4,1,6,1,1,1,2,32,1,2,2,6,1,4,1,32,36,1,2,1,2,2,1,1,1,4,1,1,2,3,4,1,1,1,1,2,2,1,3,1,106,35,1,1,1,1,1,1,3,4,9,7,1,1,2,2,20,3,2,5,8,1,2,2,8,1,2,8,5,6,1,1,35,6,3,12,1,45,99,1,1,2,1,1,1,2,2,2,2,2,1,3,13,2,1,1,1,1,1,5,1,1,4,3,1,1,1,1,1,6,3,10,2,1,1,2,12,4,1,1,2,1,2,5,1,1,1,8,1,2,2,1,4,4,1,35,45,1,1,48,1,10,1,1,2,1,13,1,1,6,1,1,4,1,4,1,1,43,1,1,37,4,1,1,1,1,3,1,3,2,1,1,4,2,1,3,1,2,1,1,7,1,2,2,1,1,7,2,1,1,1,1,4,1,1,2,2,2,1,2,5,25,2,1,1,1,1,18,20,3,1,1,1,1,2,2,1,1,1,1,27,2,1,1,1,2,1,1,1,2,8,1,1,3,1,5,3,22,34,4,3,5,1,1,2,1,1,2,6,1,3,2,2,2,1,1,5,6,1,1,1,3,2,832,1,2,4,13,1,19,1,2,2,1,2,2,1,7,2,19,1,14,1,1,2,1,3,5,1,1,2,4,3,1,1,44,1,6,1,1,16,11,10,2,3,31,14,1,1,5,1,5,2,1,2,3,1,1,1,2,1,2,85,1,2,6,1,1,1,1,1,2,3,9,3,3,1,19,1,2,1,1,1,4,1,7,32,2,1,5,2,2,2,2,1,44,1,382,2,24,34,10,176,1,1,3,96,18,9,1,11,2,1,3,1,2,2,14,8,9,3,2,3,8,65,1,21,11,40,14,14,72,14,6,108,9,4,14,4,4,76,38,3,2,2,2,1,1,2,1,2,1,2,1,3,2,4,1,4,5,3,4,5,4,9,13,5,21,3,3,1,4,5,3,4,5,3,4,9,12,5,19,3,14,5,22,4,2,2,2,1,3,4,5,4,5,2,2,1,1,2,2,1,2,2,2,3,3,1,2,1,3,4,1,4,1,1,1,1,2,2,5,2,6,1,1,4,1,2,2,1,1,10,1,1,9,1,3,4,2,3,1,1,8,1,2,2,5,1,13,21,2,2,1,3,2,1,5,1,2,1,3,182,3,1,1,3,2,2,6,1,1,8,4,182,201,14,9,1,2,7,1,2,2,1,2,1,4,2,1,10,15,7,12,13,11,15,1,12,1,1,1,1,1,1,4,14,1,19,22,6,16,3,1,1,2,1,6,6,1,1,6,1,1,544,65,1,229,4,2,2,1,3,1,1,1,2,1,1,1,1,5,7,1,3,1,1,1,1,1,1,2,22,14,2,4,6,76,29,2,2,7,11,13,18,1,6,4,11,1,9,12,18,7,15,6,8,2,2,228,16,24,48,18,1,1,10,248,511,9,1,4,5,1,1,3,1,1,1,1,2,3,2,1,1,2,1,1,1,2,1,1,20,1,5,2,2,2,2,3,7,43,2,1,3,1,1,1,1,1,2,1,1,1,6,1,1,6,1,1,1,1,1,2,226,1,2,2,2,1,7,4,1,2,2,4,2,5,2,1,8,5,17,2,3,1,2,1,3,1,2,39,5,3,2,1,2,1,1,2,1,4,1,2,2,5,1,1,1,1,3,1,21,13,2,1,2,2,1,2,1,9,14,6,6,9,1,228,246,2,1,2,7,1,260,2,1,1,1,5,1,3,1,7,2,14,5,1,1,1,1,1,19,2,2,2,3,2,1,24,42,2,4,4,2,1,1,7,1,1,1,1,6,1,1,1,8,1,1,2,1,1,234,1,1,1,1,205,1,4,30,1,1,1,1,2,2,2,1,1,10,1,1,10,7,1,6,14,1,1,1,3,2,4,1,1,1,239,16,3,1,1,6,2,8,2,1,1,1,2,9,2,2,2,2,2,30,8,5,11,2,1,2,1,1,5,1,13,10,4,3,1,1,1,3,13,1,1,1,84,11,3,1,3,3,1,1,1,70,2,77,1,2,16,4,1,2,1,1,1,1,1,1,2,1,1,1,1,1,2,2,1,15,3,6,6,1,2,1,1,2,8,1,1,1,1,1,1,4,99,3,1,3,5,14,1,1,2,1,9,3,2,1,17,2,4,4,1,1,1,1,1,7,1,8,1,1,2,3,9,3,1,1,1,9,2,2,10,10,1,2,4,13,1,6,18,3,2,1,1,2,1,8,18,2,1,2,50,2,1,1,4,3,2,34,2,1,1,18,37,3,1,2,1,1,1,1,16,1,4,1,16,3,5,6,2,1,1,6,8,3,5,5,1,3,2,1,1,2,14,1,3,6,34,12,1,1,18,2,3,21,9,9,1,1,6,1,2,2,1,2,311,74,4,3,5,7,2,2,2,1,1,1,4,3,1,18,2,3,11,1,2,1,1,1,2,1,2,2,1,1,4,28,3,2,4,1,4,5,3,5,6,4,9,13,5,20,13,21,2,6,4,7,7,20,1,3,3,5,1,4,7,11,10,1,1,3,4,1,1,1,1,1,2,65,22,33,2,27,4,3,65,1,1,1,3,3,1,3,1,4,1,1,1,9,2,2,4,2,221,3,1,3,236,1,1,10,5,3,1,6,2,5,3,25,14,2,26,23,10,79,1,2,2,2,3,2,1,1,1,1,2,6,7,1,6,1,2,13,1,1,1,5,4,10,1,4,1,1,1,1,5,1,8,3,4,6,7,16,3,3,1,7,2,8,5,1,7,3,2,1,1,9,14,8,1,4,3,1,1,1,2,2,9,12,3,46,1,1,1,2,6,8,1,17,2,2,5,3,1,1,1,4,1,2,2,202,67,1,1,20,20,32,3,4,10,2,1,2,2,1,1,1,1,6,2,1,1,1,1,1,2,5,1,2,1,3,2,4,1,2,1,1,1,3,3,81,5,2,31,11,1,1,1,1,15,5,110,6,3,1,4,1,1,7,16,10,111,6,16,2,1,21,13,98,1,233,3,1,3,2,1,1,1,1,1,1,1,2,1,36,4,29,2,1,3,1,2,2,35,1,1,2,1,1,1,2,6,5,1,1,1,5,25,1,1,3,1,1,1,1,2,1,4,1,4,1,2,2,1,6,5,3,2,3,1,1,1,63,2,5,6,1,1,2,1,1,2,27,54,1,1,1,1,1,2,2,1,3,2,1,1]],"container":"<table class=\"compact\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>term_name<\/th>\n      <th>count<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":10,"columnDefs":[{"className":"dt-right","targets":2},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"term_name","targets":1},{"name":"count","targets":2}],"order":[],"autoWidth":false,"orderClasses":false,"orderCellsTop":true},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><p><br></p>
</div>
<div class="section level3">
<h3 id="step8">Step 8: Merge clusters<a class="anchor" aria-label="anchor" href="#step8"></a>
</h3>
<p>Following cluster characterisation, this step consists in identifying
and merging clusters that share the same enriched biological functions
from at least one specified data source (e.g. GO, KEGG etc.) in the ORA.
This step allows us to merge multiple clusters where the gene proteins
were not found to be sufficiently interactive in STRING to be considered
a single cluster. However, functional enrichment revealed that they
participate in the same biological processes. Therefore, clusters that
do merge form a larger and more comprehensive cluster, holding a unique
biological context.</p>
<p>To do so, the step uses the <span class="homemade-function"><code><a href="../reference/clustrfusion.html">clustrfusion()</a></code></span> function,
which requires the output of the <span class="homemade-function"><code><a href="../reference/clustrenrich.html">clustrenrich()</a></code></span> function as
its input data.</p>
<hr>
<p><span class="italic-grey">The merging process is conducted separately
for each data source, in descending order of the total number of unique
genes present in both the data source and the background list. An
optional parameter, <code>monoterm_fusion</code>, can be set to
<code>TRUE</code> to restrict merging to clusters that share a single,
identical enrichment term in at least one source. This ensures that
merged clusters are consistently linked to a specific biological
function, maintaining a coherent biological context. By default this is
set to <code>FALSE</code> as the merging process becomes highly
stringent, as it requires clusters to be associated with the same unique
biological function. This can significantly limit the likelihood of
merging. </span></p>
<hr>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clustr_fusionres</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustrfusion.html">clustrfusion</a></span><span class="op">(</span></span>
<span>  clustrenrich_data <span class="op">=</span> <span class="va">clustr_enrichres</span>,</span>
<span>  monoterm_fusion <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; 44 / 53 clusters left after the fusion process.</span></span></code></pre></div>
<p>The output is a named list with four components:</p>
<ul>
<li>
<strong><code>dr_g_a_fusion</code></strong>: A dataframe of type
<strong>g_a</strong> holding the cluster fusion results</li>
<li>
<strong><code>dr_g_a_fusion</code></strong>: A dataframe of type
<strong>c_a</strong> holding the cluster fusion results</li>
<li>
<strong><code>c_fusionlog</code></strong>: A dataframe tracing
cluster fusion events, indicating the sources from which they originated
(e.g. GO, KEGG or WP)</li>
<li>
<strong><code>params</code></strong>: A list of the main parameters
used; in this case the choice for <code>monoterm_fusion</code>
</li>
</ul>
<p>We can take a look at the structure of the output:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">clustr_fusionres</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 4</span></span>
<span><span class="co">#&gt;  $ dr_g_a_fusion:'data.frame':   754 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;   ..$ gene_id   : chr [1:754] "ENSDARG00000051783" "ENSDARG00000009285" "ENSDARG00000020197" "ENSDARG00000043509" ...</span></span>
<span><span class="co">#&gt;   ..$ old_clustr: num [1:754] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..$ new_clustr: num [1:754] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..$ term_name : chr [1:754] "Cytoplasmic ribosomal proteins" "Cytoplasmic ribosomal proteins" "Cytoplasmic ribosomal proteins" "Cytoplasmic ribosomal proteins" ...</span></span>
<span><span class="co">#&gt;   ..$ term_id   : chr [1:754] "WP:WP324" "WP:WP324" "WP:WP324" "WP:WP324" ...</span></span>
<span><span class="co">#&gt;   ..$ source    : chr [1:754] "WP" "WP" "WP" "WP" ...</span></span>
<span><span class="co">#&gt;  $ dr_c_a_fusion:'data.frame':   133 obs. of  7 variables:</span></span>
<span><span class="co">#&gt;   ..$ old_clustr : num [1:133] 1 1 1 1 1 1 2 2 2 2 ...</span></span>
<span><span class="co">#&gt;   ..$ new_clustr : num [1:133] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..$ term_name  : chr [1:133] "Cytoplasmic ribosomal proteins" "embryo development ending in birth or egg hatching" "myeloid cell differentiation" "ribonucleoprotein complex biogenesis" ...</span></span>
<span><span class="co">#&gt;   ..$ term_size  : int [1:133] 60 180 89 175 102 351 60 7 102 351 ...</span></span>
<span><span class="co">#&gt;   ..$ term_id    : chr [1:133] "WP:WP324" "GO:0009792" "GO:0030099" "GO:0022613" ...</span></span>
<span><span class="co">#&gt;   ..$ source     : chr [1:133] "WP" "GO:BP" "GO:BP" "GO:BP" ...</span></span>
<span><span class="co">#&gt;   ..$ highlighted: logi [1:133] FALSE TRUE TRUE TRUE FALSE TRUE ...</span></span>
<span><span class="co">#&gt;  $ c_fusionlog  :'data.frame':   754 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ old_clustr        : num [1:754] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..$ after_GO:BP_fusion: num [1:754] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..$ after_KEGG_fusion : num [1:754] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..$ after_WP_fusion   : num [1:754] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;  $ params       :List of 1</span></span>
<span><span class="co">#&gt;   ..$ monoterm_fusion: logi FALSE</span></span>
<span><span class="co">#&gt;  - attr(*, "class")= chr "clustrfusion"</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step9">Step 9: Fish lonely genes<a class="anchor" aria-label="anchor" href="#step9"></a>
</h3>
<p>Genes not assigned to any cluster are referred to as <em>lonely
genes</em>. These genes are not typically explored because they do not
contribute to functional enrichment, and thus, they do not help in
characterising the data. This step involves “fishing” these <em>lonely
genes</em> into existing clusters.</p>
<hr>
<p><span class="italic-grey">Genes become lonely in one of two cases
:</span></p>
<ul>
<li><span class="italic-grey">they are initially not sufficiently
interactive with another protein complex to participate in or form a
cluster</span></li>
<li><span class="italic-grey">they were initially part of a cluster but
failed to pass the cluster size filtering step, thus falling into the
lonely cluster.</span></li>
</ul>
<hr>
<p>To “fish” these genes meaningfully, the <span class="homemade-function"><code><a href="../reference/lonelyfishing.html">lonelyfishing()</a></code></span> function
matches lonely genes with clusters that share the same functional
annotations, and incorporates them into the cluster.</p>
<p>Each lonely gene has a <strong>friendliness</strong> metric,
representing the number of potentiel clusters it can be incorporated
into. The <strong>friendliness filter</strong> allows the user to set a
limit to this metric, solely fishing genes that fall below this limit,
while genes that exceed the limit remain in the lonely cluster.</p>
<hr>
<p><span class="italic-grey"> Low values of this limit, such as 1 to 3,
result in fewer lonely genes being incorporated but reduce redundancy
between clusters. Although this approach does not fish as much, it
ensures that lonely genes associated with a wide range of biological
functions are not distributed across many clusters, thereby minimising
noise in the results. By default, the limit is set to 0, meaning it is
not applied, and all lonely genes eligible for fishing are
incorporated.</span></p>
<hr>
<p>The function’s data inputs include:</p>
<ul>
<li>
<strong><code>dr_data</code></strong>: A dataframe of type t that
typically corresponds to the output of <span class="homemade-function"><code><a href="../reference/getids.html">getids()</a></code></span> or <span class="homemade-function"><code><a href="../reference/getregs.html">getregs()</a></code></span>. This input
holds at least <code>gene_id</code> and <code>term_name</code> columns,
respectively containing gene identifiers and biological function
annotations for the deregulated genes. Recommended to hold also
<code>transcript_id</code> for futur functions.</li>
<li>
<strong><code>clustrenrich_data</code></strong>: The named
<code>list</code> output of the <span class="homemade-function"><code><a href="../reference/clustrenrich.html">clustrenrich()</a></code></span>
function.</li>
<li>
<strong><code>clustrfusion_data</code></strong>: The named
<code>list</code> output of the <span class="homemade-function"><code><a href="../reference/clustrfusion.html">clustrfusion()</a></code></span>
function.</li>
</ul>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lonely_fishres</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lonelyfishing.html">lonelyfishing</a></span><span class="op">(</span></span>
<span>  dr_data <span class="op">=</span> <span class="va">dr_t_regs</span>,</span>
<span>  clustrenrich_data <span class="op">=</span> <span class="va">clustr_enrichres</span>,</span>
<span>  clustrfusion_data <span class="op">=</span> <span class="va">clustr_fusionres</span>,</span>
<span>  friendly_limit <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  path <span class="op">=</span> <span class="st">"your/chosen/path"</span>,</span>
<span>  output_filename <span class="op">=</span> <span class="st">"filename.rds"</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For more information on each argument of the function, run
<code>?cluefish::lonelyfishing()</code>.</p>
<p>The output of the <span class="homemade-function"><code><a href="../reference/lonelyfishing.html">lonelyfishing()</a></code></span> is a named
list holding 3 components, where:</p>
<ul>
<li>
<strong><code>dr_g_a_fishing</code></strong>: A dataframe of type
<strong>t_c_a</strong> holding the lonely fishing results</li>
<li>
<strong><code>dr_c_a_fishing</code></strong>: A dataframe of the
lonely fishing results similar to the
<code>clustrfusion_data$dr_c_a_fusion</code> dataframe with each row
being a combination of cluster ID and biological function
annotation.</li>
<li>
<strong><code>params</code></strong>: A list of the main parameters
used; in this case the <code>friendly_limit</code>
</li>
</ul>
<p>We can take a look a the structure of the output:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">lonely_fishres</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 3</span></span>
<span><span class="co">#&gt;  $ dr_t_c_a_fishing:'data.frame':    9841 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;   ..$ transcript_id: chr [1:9841] "ENSDART00000175435.2" "ENSDART00000175435.2" "ENSDART00000175435.2" "ENSDART00000175435.2" ...</span></span>
<span><span class="co">#&gt;   ..$ gene_id      : chr [1:9841] "ENSDARG00000005791" "ENSDARG00000005791" "ENSDARG00000005791" "ENSDARG00000005791" ...</span></span>
<span><span class="co">#&gt;   ..$ gene_name    : chr [1:9841] "rpl28_t2" "rpl28_t2" "rpl28_t2" "rpl28_t2" ...</span></span>
<span><span class="co">#&gt;   ..$ TF           : logi [1:9841] FALSE FALSE FALSE FALSE FALSE FALSE ...</span></span>
<span><span class="co">#&gt;   ..$ old_clustr   : chr [1:9841] "1" "1" "1" "1" ...</span></span>
<span><span class="co">#&gt;   ..$ new_clustr   : chr [1:9841] "1" "1" "1" "1" ...</span></span>
<span><span class="co">#&gt;   ..$ friendliness : int [1:9841] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..$ term_name    : chr [1:9841] "embryo development ending in birth or egg hatching" "embryo development ending in birth or egg hatching" "embryo development ending in birth or egg hatching" "embryo development ending in birth or egg hatching" ...</span></span>
<span><span class="co">#&gt;   ..$ term_id      : chr [1:9841] "GO:0009792" "GO:0009792" "GO:0009792" "GO:0009792" ...</span></span>
<span><span class="co">#&gt;   ..$ source       : chr [1:9841] "GO:BP" "GO:BP" "GO:BP" "GO:BP" ...</span></span>
<span><span class="co">#&gt;  $ dr_c_a_fishing  :'data.frame':    242 obs. of  7 variables:</span></span>
<span><span class="co">#&gt;   ..$ old_clustr : chr [1:242] "1" "1" "1" "1" ...</span></span>
<span><span class="co">#&gt;   ..$ new_clustr : chr [1:242] "1" "1" "1" "1" ...</span></span>
<span><span class="co">#&gt;   ..$ term_name  : chr [1:242] "Cytoplasmic ribosomal proteins" "embryo development ending in birth or egg hatching" "myeloid cell differentiation" "Protein export" ...</span></span>
<span><span class="co">#&gt;   ..$ term_size  : int [1:242] 60 180 89 26 175 102 351 60 7 102 ...</span></span>
<span><span class="co">#&gt;   ..$ term_id    : chr [1:242] "WP:WP324" "GO:0009792" "GO:0030099" "KEGG:03060" ...</span></span>
<span><span class="co">#&gt;   ..$ source     : chr [1:242] "WP" "GO:BP" "GO:BP" "KEGG" ...</span></span>
<span><span class="co">#&gt;   ..$ highlighted: logi [1:242] FALSE TRUE TRUE FALSE TRUE FALSE ...</span></span>
<span><span class="co">#&gt;  $ params          :List of 1</span></span>
<span><span class="co">#&gt;   ..$ friendly_limit: num 0</span></span>
<span><span class="co">#&gt;  - attr(*, "class")= chr "lonelyfishingres"</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step10">Step 10: Perform functional enrichment of the lonely cluster
genes<a class="anchor" aria-label="anchor" href="#step10"></a>
</h3>
<p>The lonely cluster, consisting of the remaining lonely genes, can
undergo a simple functional enrichment analysis to gain further insight
into the biological context it may contain. The customisations used in
the earlier cluster-wise enrichment analysis are fully applicable here
as well. Therefore, the parameters were set to match those chosen in the
previous cluster-wise functional enrichment analysis. The function <span class="homemade-function"><code><a href="../reference/simplenrich.html">simplenrich()</a></code></span> enables a
single ORA, with parameters mirroring the earlier <span class="homemade-function"><code><a href="../reference/clustrenrich.html">clustrenrich()</a></code></span> function.
The only difference is that a <code>input_genes</code> argument is
created to take the list of genes.</p>
<p>First, we need to extract the lonely genes:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lonelycluster_data</span> <span class="op">&lt;-</span> <span class="va">lonely_fishres</span><span class="op">$</span><span class="va">dr_t_c_a_fishing</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">new_clustr</span> <span class="op">==</span> <span class="st">"Lonely"</span><span class="op">)</span></span></code></pre></div>
<p>Then, we can conduct enrichment:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lonely_clustr_analysis_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simplenrich.html">simplenrich</a></span><span class="op">(</span></span>
<span>  input_genes <span class="op">=</span> <span class="va">lonelycluster_data</span><span class="op">$</span><span class="va">gene_id</span>,</span>
<span>  bg_genes <span class="op">=</span> <span class="va">bg_t_ids</span><span class="op">$</span><span class="va">gene_id</span>,</span>
<span>  bg_type <span class="op">=</span> <span class="st">"custom_annotated"</span>,</span>
<span>  sources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:BP"</span>, <span class="st">"KEGG"</span>, <span class="st">"WP"</span><span class="op">)</span>, </span>
<span>  organism <span class="op">=</span> <span class="st">"drerio"</span>,</span>
<span>  user_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  correction_method <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>  only_highlighted_GO <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_term_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  max_term_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  ngenes_enrich_filtr <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  path <span class="op">=</span> <span class="st">"your/chosen/path"</span>,</span>
<span>  output_filename <span class="op">=</span> <span class="st">"filename.rds"</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For more information on each argument of the function, run
<code>?cluefish::simplenrich()</code>.</p>
<p>The output is a named list with two components:</p>
<ul>
<li>
<code>unfiltered</code>: A named list of the unfiltered enrichment
results, holding two sub-components :
<ul>
<li>
<code>dr_g_a</code>: A dataframe of type <strong>g_a</strong>
holding the simple enrichment results</li>
<li>
<code>gostres</code>: A named list of the original result output of
g:GOst</li>
</ul>
</li>
<li>
<code>filtered</code>: A named list of the filtered enrichment
results, holding three sub-components:
<ul>
<li>
<code>dr_g_a</code>: A dataframe of type <strong>g_a</strong>
holding the simple enrichment results</li>
<li>
<code>dr_a</code>: A dataframe of type <strong>a</strong> holding
the simple enrichment</li>
<li>
<code>params</code>: A list of the main parameters used</li>
</ul>
</li>
</ul>
<p>We can take a look at the filtered results, with each row
corresponding to an enriched term :</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">lonely_clustr_analysis_res</span><span class="op">$</span><span class="va">filtered</span><span class="op">$</span><span class="va">dr_a</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 12</span></span>
<span><span class="co">#&gt; Columns: 17</span></span>
<span><span class="co">#&gt; $ query                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "query_1", "query_1", "query_1", "query_1", "que…</span></span>
<span><span class="co">#&gt; $ significant           <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, …</span></span>
<span><span class="co">#&gt; $ p_value               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 2.09e-13, 1.12e-06, 1.06e-02, 1.26e-02, 2.98e-02…</span></span>
<span><span class="co">#&gt; $ term_size             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 70, 139, 41, 106, 15, 23, 268, 143, 72, 46, 78, …</span></span>
<span><span class="co">#&gt; $ query_size            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 763, 763, 763, 763, 763, 763, 177, 177, 177, 177…</span></span>
<span><span class="co">#&gt; $ intersection_size     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 26, 26, 9, 15, 5, 6, 22, 14, 9, 6, 10, 6</span></span>
<span><span class="co">#&gt; $ precision             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.03408, 0.03408, 0.01180, 0.01966, 0.00655, 0.0…</span></span>
<span><span class="co">#&gt; $ recall                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.3714, 0.1871, 0.2195, 0.1415, 0.3333, 0.2609, …</span></span>
<span><span class="co">#&gt; $ term_id               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "GO:0002088", "GO:0007601", "GO:0000041", "GO:00…</span></span>
<span><span class="co">#&gt; $ source                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "GO:BP", "GO:BP", "GO:BP", "GO:BP", "GO:BP", "GO…</span></span>
<span><span class="co">#&gt; $ term_name             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "lens development in camera-type eye", "visual p…</span></span>
<span><span class="co">#&gt; $ effective_domain_size <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 15468, 15468, 15468, 15468, 15468, 15468, 5714, …</span></span>
<span><span class="co">#&gt; $ source_order          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 681, 3094, 17, 13905, 1212, 13702, 280, 279, 220…</span></span>
<span><span class="co">#&gt; $ parents               <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> &lt;"GO:0043010", "GO:0048856"&gt;, "GO:0050953", "GO:…</span></span>
<span><span class="co">#&gt; $ evidence_codes        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "IMP IGI,IBA,IBA,IBA,IBA,IBA,IBA,IBA,IBA,IBA,IB…</span></span>
<span><span class="co">#&gt; $ intersection          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSDARG00000013963,ENSDARG00000016793,ENSDARG00…</span></span>
<span><span class="co">#&gt; $ highlighted           <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, FALSE, FALSE…</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="outputs-to-explore-the-results">Outputs to explore the results<a class="anchor" aria-label="anchor" href="#outputs-to-explore-the-results"></a>
</h2>
<div class="section level3">
<h3 id="homemade-functions">Homemade functions<a class="anchor" aria-label="anchor" href="#homemade-functions"></a>
</h3>
<p>To fully explore the Cluefish results, including DR modelling metrics
from <code>DRomics</code> outputs is required.</p>
<ol style="list-style-type: decimal">
<li>
<p>The <span class="homemade-function"><code><a href="../reference/results_to_csv.html">results_to_csv()</a></code></span> function
generates a summary .csv file from the Cluefish workflow results,
supporting manual exploration.</p>
<p>It requires two main data inputs:</p>
<ul>
<li>
<strong><code>lonelyfishing_data</code></strong>: The named list
output of the <code><a href="../reference/lonelyfishing.html">lonelyfishing()</a></code> function.</li>
<li>
<strong><code>bmdboot_data</code></strong>: The DRomics bmdboot
dataframe results after <code><a href="https://rdrr.io/pkg/DRomics/man/bmdboot.html" class="external-link">DRomics::bmdboot()</a></code> or following the
additional step <code><a href="https://rdrr.io/pkg/DRomics/man/bmdfilter.html" class="external-link">DRomics::bmdfilter()</a></code>. This data should
include the deregulated transcript identifiers from <a href="#step2">Step 2</a>.</li>
</ul>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/results_to_csv.html">results_to_csv</a></span><span class="op">(</span></span>
<span> lonelyfishing_data <span class="op">=</span> <span class="va">lonely_fishres</span>,</span>
<span> bmdboot_data <span class="op">=</span> <span class="va">b_definedCI</span>,</span>
<span> path <span class="op">=</span> <span class="st">"your/chosen/path"</span>,</span>
<span> output_filename <span class="op">=</span> <span class="st">"filename.csv"</span>,</span>
<span> overwrite <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For more information on each argument of the function, run
<code>?cluefish::results_to_csv()</code>.</p>
</li>
</ol>
<p>Cluefish also leverages <code>DRomics</code> visualisation functions,
like <code><a href="https://rdrr.io/pkg/DRomics/man/curvesplot.html" class="external-link">DRomics::curvesplot()</a></code>, to plot fitted dose-response
curves.</p>
<ol start="2" style="list-style-type: decimal">
<li>
<p>The <span class="homemade-function"><code><a href="../reference/curves_to_pdf.html">curves_to_pdf()</a></code></span> function
generates a .pdf of the fitted curves for each cluster’s gene set,
offering visual insight into cluster composition.</p>
<p>It requires five main inputs:</p>
<ul>
<li>
<strong><code>lonelyfishing_data</code></strong>: The named list
output of the <span class="homemade-function"><code><a href="../reference/lonelyfishing.html">lonelyfishing()</a></code></span>
function.</li>
<li>
<strong><code>bmdboot_data</code></strong>: The <code>DRomics</code>
<code>bmdboot</code> dataframe results after
<code><a href="https://rdrr.io/pkg/DRomics/man/bmdboot.html" class="external-link">DRomics::bmdboot()</a></code> or following the additional step
<code><a href="https://rdrr.io/pkg/DRomics/man/bmdfilter.html" class="external-link">DRomics::bmdfilter()</a></code>. This must correspond to the data
loaded containing the deregulated transcript identifiers in the <a href="#step2">Step 2</a>.</li>
<li>
<strong><code>clustrfusion_data</code></strong>: The named list
output of the <span class="homemade-function"><code><a href="../reference/clustrfusion.html">clustrfusion()</a></code></span>
function.</li>
<li>
<strong><code>tested_doses</code></strong>: A vector of the tested
doses that can be found in the output of the
<code><a href="https://rdrr.io/pkg/DRomics/man/drcfit.html" class="external-link">DRomics::drcfit()</a></code> function as
<code>unique(f$omicdata$dose)</code>.</li>
<li>
<strong><code>annot_order</code></strong>: A vector specifying the
prioritised order of annotation sources used in the <span class="homemade-function"><code><a href="../reference/clustrenrich.html">clustrenrich()</a></code></span>
function.</li>
</ul>
<hr>
<p><span class="italic-grey">The
<strong><code>annot_order</code></strong> specifies the preferred order
of annotation sources for assigning a primary descriptor to each
cluster. If an enriched function is found in the first source listed,
that function will be used as the primary descriptor for the cluster.
This order allows you to prioritise annotation databases based on their
relevance and the quality of information they provide for your
study.</span></p>
<hr>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/curves_to_pdf.html">curves_to_pdf</a></span><span class="op">(</span></span>
<span> lonelyfishing_data <span class="op">=</span> <span class="va">lonely_fishres</span>,</span>
<span> bmdboot_data <span class="op">=</span> <span class="va">b_definedCI</span>, </span>
<span> clustrfusion_data <span class="op">=</span> <span class="va">clustr_fusionres</span>,</span>
<span> tested_doses <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">f</span><span class="op">$</span><span class="va">omicdata</span><span class="op">$</span><span class="va">dose</span><span class="op">)</span>, </span>
<span> annot_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:BP"</span>, <span class="st">"KEGG"</span>, <span class="st">"WP"</span><span class="op">)</span>,</span>
<span> colorby <span class="op">=</span> <span class="st">"trend"</span>,</span>
<span> addBMD <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span> scaling <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span> npoints<span class="op">=</span> <span class="fl">100</span>,</span>
<span> free.y.scales <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span> xmin <span class="op">=</span> <span class="fl">0.1</span>, </span>
<span> xmax <span class="op">=</span> <span class="fl">100</span>, </span>
<span> dose_log_transfo <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span> line.size <span class="op">=</span> <span class="fl">0.7</span>, </span>
<span> line.alpha <span class="op">=</span> <span class="fl">0.4</span>, </span>
<span> point.size <span class="op">=</span> <span class="fl">2</span>, </span>
<span> point.alpha <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span> xunit <span class="op">=</span> <span class="st">"µg/L"</span>,</span>
<span> xtitle <span class="op">=</span> <span class="st">"Dose (µg/L)"</span>,</span>
<span> ytitle <span class="op">=</span> <span class="st">"Signal"</span>,</span>
<span> colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inc"</span> <span class="op">=</span> <span class="st">"#1B9E77"</span>, <span class="st">"dec"</span> <span class="op">=</span> <span class="st">"#D95F02"</span>, <span class="st">"U"</span> <span class="op">=</span> <span class="st">"#7570B3"</span>, <span class="st">"bell"</span> <span class="op">=</span> <span class="st">"#E7298A"</span><span class="op">)</span>,</span>
<span> path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"outputs/"</span>, <span class="va">file_date</span>, <span class="st">"/"</span><span class="op">)</span>,</span>
<span> output_filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"workflow_curvesplots_"</span>, <span class="va">file_date</span>, <span class="st">".pdf"</span><span class="op">)</span>,</span>
<span> overwrite <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For more information on each argument of this function, run
<code>?cluefish::curves_to_pdf()</code>. For more information on each
argument of the <code><a href="https://rdrr.io/pkg/DRomics/man/curvesplot.html" class="external-link">DRomics::curvesplot()</a></code>, run
<code>?DRomics::curvesplot()</code>.</p>
</li>
</ol>
<p><br></p>
</div>
<div class="section level3">
<h3 id="quarto-report-templates">Quarto report templates<a class="anchor" aria-label="anchor" href="#quarto-report-templates"></a>
</h3>
<p>The Cluefish compendium includes Quarto report templates that can
streamline the summarisation of Cluefish results, enhancing
interpretability and ease of use. These reports are auto-generated but
can be customised as needed.</p>
<p>The available templates include:</p>
<ol style="list-style-type: decimal">
<li><p><strong><code>report_workflow_results.qmd</code></strong> This
report summarises key findings from the DRomics analysis and workflow
results. It includes visual summaries like BMD plots per cluster and
interactive visualisations using the <code>plotly</code> R package,
helping to prioritise results and explore each cluster.</p></li>
<li><p><strong><code>report_lonely_results.qmd</code></strong> This
report focuses on the Lonely cluster, summarising the results of genes
that did not associate with other clusters. It includes
<code>plotly</code> interactive <code>DRomics</code> plots like BMD
distributions, trend-colored fitted curves, and ECDF plots for detailed
exploration.</p></li>
<li><p><strong><code>report_comparison_results</code></strong> This
report compares Cluefish with a standard workflow
(<code><a href="../reference/simplenrich.html">simplenrich()</a></code>). A script is available to perform the
standard approach script can be found in the <code>analyses</code>
folder, under the name: <code>standard-pipeline</code>.</p></li>
</ol>
<p>To generate these reports, use the custom <code><a href="../reference/render_qmd.html">render_qmd()</a></code>
function, which builds on <code><a href="https://quarto-dev.github.io/quarto-r/reference/quarto_render.html" class="external-link">quarto::quarto_render()</a></code> and
allows output to a designated directory. It supports all
<code><a href="https://quarto-dev.github.io/quarto-r/reference/quarto_render.html" class="external-link">quarto::quarto_render()</a></code> arguments.</p>
<p><span class="italic-red"><u>Important</u>: Some adaptations may be
needed for directory paths, file naming, or other settings.</span></p>
</div>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>The <strong>Cluefish</strong> project code is distributed under the
<a href="https://github.com/ellfran-7/cluefish/blob/main/LICENSE" class="external-link">CECILL
v2.1 license</a>. Illustrations and figures are licensed under <a href="https://choosealicense.com/licenses/cc-by-sa-4.0/" class="external-link">CC-BY-SA-4.0
license</a></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.4.1 (2024-06-14 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span><span class="co">#&gt; Running under: Windows 11 x64 (build 26100)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=French_France.utf8  LC_CTYPE=French_France.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=French_France.utf8 LC_NUMERIC=C                  </span></span>
<span><span class="co">#&gt; [5] LC_TIME=French_France.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Paris</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices datasets  utils     methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] cluefish_1.0.1 ggplot2_3.5.1 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] rstudioapi_0.16.0           jsonlite_1.8.8             </span></span>
<span><span class="co">#&gt;   [3] magrittr_2.0.3              rmarkdown_2.27             </span></span>
<span><span class="co">#&gt;   [5] fs_1.6.4                    zlibbioc_1.50.0            </span></span>
<span><span class="co">#&gt;   [7] ragg_1.3.2                  vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;   [9] memoise_2.0.1               htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt;  [11] S4Arrays_1.4.1              usethis_3.0.0              </span></span>
<span><span class="co">#&gt;  [13] progress_1.2.3              curl_5.2.1                 </span></span>
<span><span class="co">#&gt;  [15] SparseArray_1.4.8           sass_0.4.9                 </span></span>
<span><span class="co">#&gt;  [17] bslib_0.7.0                 quarto_1.4.4               </span></span>
<span><span class="co">#&gt;  [19] htmlwidgets_1.6.4           desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [21] httr2_1.0.2                 plotly_4.10.4              </span></span>
<span><span class="co">#&gt;  [23] cachem_1.1.0                mime_0.12                  </span></span>
<span><span class="co">#&gt;  [25] lifecycle_1.0.4             DRomics_2.5-2              </span></span>
<span><span class="co">#&gt;  [27] pkgconfig_2.0.3             Matrix_1.7-0               </span></span>
<span><span class="co">#&gt;  [29] R6_2.5.1                    fastmap_1.2.0              </span></span>
<span><span class="co">#&gt;  [31] GenomeInfoDbData_1.2.12     MatrixGenerics_1.16.0      </span></span>
<span><span class="co">#&gt;  [33] shiny_1.9.1                 digest_0.6.36              </span></span>
<span><span class="co">#&gt;  [35] colorspace_2.1-0            ps_1.7.7                   </span></span>
<span><span class="co">#&gt;  [37] AnnotationDbi_1.66.0        S4Vectors_0.42.1           </span></span>
<span><span class="co">#&gt;  [39] DESeq2_1.44.0               rprojroot_2.0.4            </span></span>
<span><span class="co">#&gt;  [41] pkgload_1.4.0               crosstalk_1.2.1            </span></span>
<span><span class="co">#&gt;  [43] textshaping_0.4.0           GenomicRanges_1.56.1       </span></span>
<span><span class="co">#&gt;  [45] RSQLite_2.3.7               filelock_1.0.3             </span></span>
<span><span class="co">#&gt;  [47] fansi_1.0.6                 httr_1.4.7                 </span></span>
<span><span class="co">#&gt;  [49] abind_1.4-5                 compiler_4.4.1             </span></span>
<span><span class="co">#&gt;  [51] here_1.0.1                  remotes_2.5.0              </span></span>
<span><span class="co">#&gt;  [53] bit64_4.0.5                 withr_3.0.0                </span></span>
<span><span class="co">#&gt;  [55] BiocParallel_1.38.0         DBI_1.2.3                  </span></span>
<span><span class="co">#&gt;  [57] pkgbuild_1.4.4              biomaRt_2.60.1             </span></span>
<span><span class="co">#&gt;  [59] rappdirs_0.3.3              DelayedArray_0.30.1        </span></span>
<span><span class="co">#&gt;  [61] sessioninfo_1.2.2           tools_4.4.1                </span></span>
<span><span class="co">#&gt;  [63] httpuv_1.6.15               glue_1.7.0                 </span></span>
<span><span class="co">#&gt;  [65] promises_1.3.0              grid_4.4.1                 </span></span>
<span><span class="co">#&gt;  [67] generics_0.1.3              gtable_0.3.5               </span></span>
<span><span class="co">#&gt;  [69] tidyr_1.3.1                 data.table_1.15.4          </span></span>
<span><span class="co">#&gt;  [71] hms_1.1.3                   xml2_1.3.6                 </span></span>
<span><span class="co">#&gt;  [73] utf8_1.2.4                  XVector_0.44.0             </span></span>
<span><span class="co">#&gt;  [75] BiocGenerics_0.50.0         pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [77] stringr_1.5.1               limma_3.60.3               </span></span>
<span><span class="co">#&gt;  [79] later_1.3.2                 dplyr_1.1.4                </span></span>
<span><span class="co">#&gt;  [81] BiocFileCache_2.12.0        lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [83] renv_1.0.7                  bit_4.0.5                  </span></span>
<span><span class="co">#&gt;  [85] tidyselect_1.2.1            locfit_1.5-9.10            </span></span>
<span><span class="co">#&gt;  [87] Biostrings_2.72.1           miniUI_0.1.1.1             </span></span>
<span><span class="co">#&gt;  [89] knitr_1.48                  gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [91] IRanges_2.38.1              SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt;  [93] stats4_4.4.1                xfun_0.45                  </span></span>
<span><span class="co">#&gt;  [95] Biobase_2.64.0              statmod_1.5.0              </span></span>
<span><span class="co">#&gt;  [97] devtools_2.4.5              matrixStats_1.3.0          </span></span>
<span><span class="co">#&gt;  [99] DT_0.33                     stringi_1.8.4              </span></span>
<span><span class="co">#&gt; [101] UCSC.utils_1.0.0            lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt; [103] yaml_2.3.9                  evaluate_0.24.0            </span></span>
<span><span class="co">#&gt; [105] codetools_0.2-20            tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [107] BiocManager_1.30.23         cli_3.6.3                  </span></span>
<span><span class="co">#&gt; [109] xtable_1.8-4                systemfonts_1.1.0          </span></span>
<span><span class="co">#&gt; [111] processx_3.8.4              munsell_0.5.1              </span></span>
<span><span class="co">#&gt; [113] jquerylib_0.1.4             Rcpp_1.0.12                </span></span>
<span><span class="co">#&gt; [115] GenomeInfoDb_1.40.1         dbplyr_2.5.0               </span></span>
<span><span class="co">#&gt; [117] gprofiler2_0.2.3            png_0.1-8                  </span></span>
<span><span class="co">#&gt; [119] parallel_4.4.1              ggfortify_0.4.17           </span></span>
<span><span class="co">#&gt; [121] ellipsis_0.3.2              pkgdown_2.1.1              </span></span>
<span><span class="co">#&gt; [123] blob_1.2.4                  prettyunits_1.2.0          </span></span>
<span><span class="co">#&gt; [125] profvis_0.3.8               urlchecker_1.0.1           </span></span>
<span><span class="co">#&gt; [127] viridisLite_0.4.2           scales_1.3.0               </span></span>
<span><span class="co">#&gt; [129] purrr_1.0.2                 crayon_1.5.3               </span></span>
<span><span class="co">#&gt; [131] rlang_1.1.5                 KEGGREST_1.40.1</span></span></code></pre>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-ashburner2000" class="csl-entry">
Ashburner, Michael, Catherine A. Ball, Judith A. Blake, David Botstein,
Heather Butler, J. Michael Cherry, Allan P. Davis, et al. 2000.
<span>“Gene Ontology: Tool for the Unification of Biology.”</span>
<em>Nature Genetics</em> 25 (1): 25–29. <a href="https://doi.org/10.1038/75556" class="external-link">https://doi.org/10.1038/75556</a>.
</div>
<div id="ref-bayerlová2015" class="csl-entry">
Bayerlová, Michaela, Klaus Jung, Frank Kramer, Florian Klemm, Annalen
Bleckmann, and Tim Beißbarth. 2015. <span>“Comparative Study on Gene Set
and Pathway Topology-Based Enrichment Methods.”</span> <em>BMC
Bioinformatics</em> 16 (1): 334. <a href="https://doi.org/10.1186/s12859-015-0751-5" class="external-link">https://doi.org/10.1186/s12859-015-0751-5</a>.
</div>
<div id="ref-committee2017" class="csl-entry">
Committee, EFSA Scientific, Anthony Hardy, Diane Benford, Thorhallur
Halldorsson, Michael John Jeger, Katrine Helle Knutsen, Simon More, et
al. 2017. <span>“Update: Use of the Benchmark Dose Approach in Risk
Assessment.”</span> <em>EFSA Journal</em> 15 (1): e04658. <a href="https://doi.org/10.2903/j.efsa.2017.4658" class="external-link">https://doi.org/10.2903/j.efsa.2017.4658</a>.
</div>
<div id="ref-das2020" class="csl-entry">
Das, Samarendra, Craig J. McClain, and Shesh N. Rai. 2020.
<span>“Fifteen Years of Gene Set Analysis for High-Throughput Genomic
Data: A Review of Statistical Approaches and Future Challenges.”</span>
<em>Entropy</em> 22 (4): 427. <a href="https://doi.org/10.3390/e22040427" class="external-link">https://doi.org/10.3390/e22040427</a>.
</div>
<div id="ref-garcía-campos2015" class="csl-entry">
García-Campos, Miguel A., Jesús Espinal-Enríquez, and Enrique
Hernández-Lemus. 2015. <span>“Pathway Analysis: State of the
Art.”</span> <em>Frontiers in Physiology</em> 6. <a href="https://www.frontiersin.org/articles/10.3389/fphys.2015.00383" class="external-link">https://www.frontiersin.org/articles/10.3389/fphys.2015.00383</a>.
</div>
<div id="ref-huang2009" class="csl-entry">
Huang, Da Wei, Brad T. Sherman, and Richard A. Lempicki. 2009.
<span>“Bioinformatics Enrichment Tools: Paths Toward the Comprehensive
Functional Analysis of Large Gene Lists.”</span> <em>Nucleic Acids
Research</em> 37 (1): 1–13. <a href="https://doi.org/10.1093/nar/gkn923" class="external-link">https://doi.org/10.1093/nar/gkn923</a>.
</div>
<div id="ref-ihnatova2018" class="csl-entry">
Ihnatova, Ivana, Vlad Popovici, and Eva Budinska. 2018. <span>“A
Critical Comparison of Topology-Based Pathway Analysis Methods.”</span>
<em>PLOS ONE</em> 13 (1): e0191154. <a href="https://doi.org/10.1371/journal.pone.0191154" class="external-link">https://doi.org/10.1371/journal.pone.0191154</a>.
</div>
<div id="ref-kanehisa2000" class="csl-entry">
Kanehisa, Minoru, and Susumu Goto. 2000. <span>“KEGG: Kyoto Encyclopedia
of Genes and Genomes.”</span> <em>Nucleic Acids Research</em> 28 (1):
27–30. <a href="https://doi.org/10.1093/nar/28.1.27" class="external-link">https://doi.org/10.1093/nar/28.1.27</a>.
</div>
<div id="ref-karp2021" class="csl-entry">
Karp, Peter D., Peter E. Midford, Ron Caspi, and Arkady Khodursky. 2021.
<span>“Pathway Size Matters: The Influence of Pathway Granularity on
over-Representation (Enrichment Analysis) Statistics.”</span> <em>BMC
Genomics</em> 22 (1): 191. <a href="https://doi.org/10.1186/s12864-021-07502-8" class="external-link">https://doi.org/10.1186/s12864-021-07502-8</a>.
</div>
<div id="ref-khatri2012" class="csl-entry">
Khatri, Purvesh, Marina Sirota, and Atul J. Butte. 2012. <span>“Ten
Years of Pathway Analysis: Current Approaches and Outstanding
Challenges.”</span> Edited by Christos A. Ouzounis. <em>PLoS
Computational Biology</em> 8 (2): e1002375. <a href="https://doi.org/10.1371/journal.pcbi.1002375" class="external-link">https://doi.org/10.1371/journal.pcbi.1002375</a>.
</div>
<div id="ref-kolberg2023" class="csl-entry">
Kolberg, Liis, Uku Raudvere, Ivan Kuzmin, Priit Adler, Jaak Vilo, and
Hedi Peterson. 2023. <span>“G:profiler—Interoperable Web Service for
Functional Enrichment Analysis and Gene Identifier Mapping (2023
Update).”</span> <em>Nucleic Acids Research</em> 51 (W1): W207–12. <a href="https://doi.org/10.1093/nar/gkad347" class="external-link">https://doi.org/10.1093/nar/gkad347</a>.
</div>
<div id="ref-kolberg2020" class="csl-entry">
Kolberg, Liis, Uku Raudvere, Ivan Kuzmin, Jaak Vilo, and Hedi Peterson.
2020. <span>“Gprofiler2 – an <span>R</span> Package for Gene List
Functional Enrichment Analysis and Namespace Conversion Toolset
g:profiler.”</span> <em>F1000Research</em> 9 (November): ELIXIR–709. <a href="https://doi.org/10.12688/f1000research.24956.2" class="external-link">https://doi.org/10.12688/f1000research.24956.2</a>.
</div>
<div id="ref-larras2018" class="csl-entry">
Larras, Floriane, Elise Billoir, Vincent Baillard, Aurélie Siberchicot,
Stefan Scholz, Tesfaye Wubet, Mika Tarkka, Mechthild Schmitt-Jansen, and
Marie-Laure Delignette-Muller. 2018. <span>“DRomics: A Turnkey Tool to
Support the Use of the Dose<span></span>response Framework for Omics
Data in Ecological Risk Assessment.”</span> <em>Environmental Science
&amp; Technology</em> 52 (24): 14461–68. <a href="https://doi.org/10.1021/acs.est.8b04752" class="external-link">https://doi.org/10.1021/acs.est.8b04752</a>.
</div>
<div id="ref-liu2017" class="csl-entry">
Liu, Lu, Jinmao Wei, and Jianhua Ruan. 2017. <span>“Pathway Enrichment
Analysis with Networks.”</span> <em>Genes</em> 8 (10): 246. <a href="https://doi.org/10.3390/genes8100246" class="external-link">https://doi.org/10.3390/genes8100246</a>.
</div>
<div id="ref-lovell2016" class="csl-entry">
Lovell, David P. 2016. <span>“Chapter 9 - Experimental Design and
Statistical Analysis of Threshold Studies.”</span> In, edited by
Takehiko Nohmi and Shoji Fukushima, 129–54. Boston: Academic Press. <a href="https://doi.org/10.1016/B978-0-12-801663-3.00009-1" class="external-link">https://doi.org/10.1016/B978-0-12-801663-3.00009-1</a>.
</div>
<div id="ref-ma2019" class="csl-entry">
Ma, Jing, Ali Shojaie, and George Michailidis. 2019. <span>“A
Comparative Study of Topology-Based Pathway Enrichment Analysis
Methods.”</span> <em>BMC Bioinformatics</em> 20 (1): 546. <a href="https://doi.org/10.1186/s12859-019-3146-1" class="external-link">https://doi.org/10.1186/s12859-019-3146-1</a>.
</div>
<div id="ref-martens2021" class="csl-entry">
Martens, Marvin, Ammar Ammar, Anders Riutta, Andra Waagmeester, Denise N
Slenter, Kristina Hanspers, Ryan A. Miller, et al. 2021.
<span>“WikiPathways: Connecting Communities.”</span> <em>Nucleic Acids
Research</em> 49 (D1): D613–21. <a href="https://doi.org/10.1093/nar/gkaa1024" class="external-link">https://doi.org/10.1093/nar/gkaa1024</a>.
</div>
<div id="ref-mubeen2022" class="csl-entry">
Mubeen, Sarah, Alpha Tom Kodamullil, Martin Hofmann-Apitius, and Daniel
Domingo-Fernández. 2022. <span>“On the Influence of Several Factors on
Pathway Enrichment Analysis.”</span> <em>Briefings in
Bioinformatics</em> 23 (3): bbac143. <a href="https://doi.org/10.1093/bib/bbac143" class="external-link">https://doi.org/10.1093/bib/bbac143</a>.
</div>
<div id="ref-szklarczyk2019" class="csl-entry">
Szklarczyk, Damian, Annika L Gable, David Lyon, Alexander Junge, Stefan
Wyder, Jaime Huerta-Cepas, Milan Simonovic, et al. 2019. <span>“STRING
V11: Protein<span></span>protein Association Networks with Increased
Coverage, Supporting Functional Discovery in Genome-Wide Experimental
Datasets.”</span> <em>Nucleic Acids Research</em> 47 (D1): D607–13. <a href="https://doi.org/10.1093/nar/gky1131" class="external-link">https://doi.org/10.1093/nar/gky1131</a>.
</div>
<div id="ref-ushey2024" class="csl-entry">
Ushey, Kevin, and Hadley Wickham. 2024. <em>Renv: Project
Environments</em>. <a href="https://rstudio.github.io/renv/" class="external-link">https://rstudio.github.io/renv/</a>.
</div>
<div id="ref-vandongen2008" class="csl-entry">
Van Dongen, Stijn. 2008. <span>“Graph Clustering via a Discrete
Uncoupling Process.”</span> <em>SIAM Journal on Matrix Analysis and
Applications</em> 30 (1): 121–41. <a href="https://doi.org/10.1137/040608635" class="external-link">https://doi.org/10.1137/040608635</a>.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ellis Franklin, Elise Billoir, Marie Laure Delignette-Muller, Sophie M. Prud’homme.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
