<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Generate curvesplots per-cluster to PDF — curves_to_pdf • cluefish</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate curvesplots per-cluster to PDF — curves_to_pdf"><meta name="description" content="This function generates a PDF file featuring dose-response curves for each cluster of deregulated genes. Each plot is labeled with the cluster ID and the number of transcripts in the cluster. By default, the curves are color-coded to indicate whether the trend is increasing, decreasing, U-shaped, or bell-shaped. Users can also apply any modifications available in the original DRomics::curvesplot() function."><meta property="og:description" content="This function generates a PDF file featuring dose-response curves for each cluster of deregulated genes. Each plot is labeled with the cluster ID and the number of transcripts in the cluster. By default, the curves are color-coded to indicate whether the trend is increasing, decreasing, U-shaped, or bell-shaped. Users can also apply any modifications available in the original DRomics::curvesplot() function."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cluefish</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/cluefish.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ellfran-7/cluefish/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Generate curvesplots per-cluster to PDF</h1>
      <small class="dont-index">Source: <a href="https://github.com/ellfran-7/cluefish/blob/HEAD/R/curves_to_pdf.R" class="external-link"><code>R/curves_to_pdf.R</code></a></small>
      <div class="d-none name"><code>curves_to_pdf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function generates a PDF file featuring dose-response curves for each cluster of deregulated genes. Each plot is labeled with the cluster ID and the number of transcripts in the cluster. By default, the curves are color-coded to indicate whether the trend is increasing, decreasing, U-shaped, or bell-shaped. Users can also apply any modifications available in the original DRomics::curvesplot() function.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">curves_to_pdf</span><span class="op">(</span></span>
<span>  <span class="va">lonelyfishing_data</span>,</span>
<span>  <span class="va">bmdboot_data</span>,</span>
<span>  <span class="va">clustrfusion_data</span>,</span>
<span>  id_col_for_curves <span class="op">=</span> <span class="st">"id"</span>,</span>
<span>  <span class="va">tested_doses</span>,</span>
<span>  <span class="va">annot_order</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  <span class="va">xunit</span>,</span>
<span>  <span class="va">xtitle</span>,</span>
<span>  <span class="va">ytitle</span>,</span>
<span>  colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>inc <span class="op">=</span> <span class="st">"#1B9E77"</span>, dec <span class="op">=</span> <span class="st">"#D95F02"</span>, U <span class="op">=</span> <span class="st">"#7570B3"</span>, bell <span class="op">=</span> <span class="st">"#E7298A"</span><span class="op">)</span>,</span>
<span>  <span class="va">path</span>,</span>
<span>  <span class="va">output_filename</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-lonelyfishing-data">lonelyfishing_data<a class="anchor" aria-label="anchor" href="#arg-lonelyfishing-data"></a></dt>
<dd><p>The named <code>list</code> output of the <code><a href="lonelyfishing.html">lonelyfishing()</a></code> function.</p></dd>


<dt id="arg-bmdboot-data">bmdboot_data<a class="anchor" aria-label="anchor" href="#arg-bmdboot-data"></a></dt>
<dd><p>The DRomics bmdboot dataframe results after DRomics::bmdfilter()</p></dd>


<dt id="arg-clustrfusion-data">clustrfusion_data<a class="anchor" aria-label="anchor" href="#arg-clustrfusion-data"></a></dt>
<dd><p>The named <code>list</code> output of the <code><a href="clustrfusion.html">clustrfusion()</a></code> function.</p></dd>


<dt id="arg-id-col-for-curves">id_col_for_curves<a class="anchor" aria-label="anchor" href="#arg-id-col-for-curves"></a></dt>
<dd><p>The column giving the identification of each curve (default is "id")</p></dd>


<dt id="arg-tested-doses">tested_doses<a class="anchor" aria-label="anchor" href="#arg-tested-doses"></a></dt>
<dd><p>A vector of the tested doses that can be found in the output of the <code><a href="https://rdrr.io/pkg/DRomics/man/drcfit.html" class="external-link">DRomics::drcfit()</a></code>function (unique(f$omicdata$dose))</p></dd>


<dt id="arg-annot-order">annot_order<a class="anchor" aria-label="anchor" href="#arg-annot-order"></a></dt>
<dd><p>A vector specifying the prioritized order of annotation sources used in the <code><a href="clustrenrich.html">clustrenrich()</a></code> function. This order determines which term or pathway will be used as primary descriptor for each cluster. The relevance of each source can be based on the quantity or quality of information they provide, tailored to the specific case study. For example, for Danio rerio, you might prioritize sources like c("GO:BP", "KEGG", "WP") based on the abundance of information they offer.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to the <code><a href="https://rdrr.io/pkg/DRomics/man/curvesplot.html" class="external-link">DRomics::curvesplot()</a></code> function</p></dd>


<dt id="arg-xunit">xunit<a class="anchor" aria-label="anchor" href="#arg-xunit"></a></dt>
<dd><p>Unit for the x scale</p></dd>


<dt id="arg-xtitle">xtitle<a class="anchor" aria-label="anchor" href="#arg-xtitle"></a></dt>
<dd><p>X-axis title</p></dd>


<dt id="arg-ytitle">ytitle<a class="anchor" aria-label="anchor" href="#arg-ytitle"></a></dt>
<dd><p>Y-axis title</p></dd>


<dt id="arg-colors">colors<a class="anchor" aria-label="anchor" href="#arg-colors"></a></dt>
<dd><p>A vector of colors for different trends (default is set at c("inc" = "#1B9E77", "dec" = "#D95F02", "U" = "#7570B3", "bell" = "#E7298A"))</p></dd>


<dt id="arg-path">path<a class="anchor" aria-label="anchor" href="#arg-path"></a></dt>
<dd><p>Destination folder for the output data results.</p></dd>


<dt id="arg-output-filename">output_filename<a class="anchor" aria-label="anchor" href="#arg-output-filename"></a></dt>
<dd><p>Output PDF filename.</p></dd>


<dt id="arg-overwrite">overwrite<a class="anchor" aria-label="anchor" href="#arg-overwrite"></a></dt>
<dd><p>If <code>TRUE</code>, the function overwrites existing output files; otherwise, it reads the existing file. (default is set to <code>FALSE</code>).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>No return value; the <code>.pdf</code> file is downloaded and stored in the specified director. It holds a curvesplot per slide for each cluster, and those for the Friendly (if created), Lonely and All deregulated gene transcripts.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ellis Franklin, Elise Billoir, Marie Laure Delignette-Muller, Sophie Prud’homme.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

