<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Retrieve gene identifiers from Ensembl/Ensembl Genomes — getids • cluefish</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Retrieve gene identifiers from Ensembl/Ensembl Genomes — getids"><meta name="description" content="This function connects to the Ensembl or Ensembl Metazoa database, queries a specified species dataset using the biomaRt package, and returns the results of the query. If the query includes the 'external_gene_name' attribute, representing readable gene symbols, the function examines whether any duplicates of this identifier exist across various Ensembl transcript and gene IDs rows. This situation occurs when a gene symbol corresponds to different Ensembl gene IDs and Ensembl transcript IDs. If duplicates are detected, the function modifies the external gene name to address this discrepancy."><meta property="og:description" content="This function connects to the Ensembl or Ensembl Metazoa database, queries a specified species dataset using the biomaRt package, and returns the results of the query. If the query includes the 'external_gene_name' attribute, representing readable gene symbols, the function examines whether any duplicates of this identifier exist across various Ensembl transcript and gene IDs rows. This situation occurs when a gene symbol corresponds to different Ensembl gene IDs and Ensembl transcript IDs. If duplicates are detected, the function modifies the external gene name to address this discrepancy."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cluefish</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/cluefish.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ellfran-7/cluefish/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Retrieve gene identifiers from Ensembl/Ensembl Genomes</h1>
      <small class="dont-index">Source: <a href="https://github.com/ellfran-7/cluefish/blob/HEAD/R/getids.R" class="external-link"><code>R/getids.R</code></a></small>
      <div class="d-none name"><code>getids.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function connects to the Ensembl or Ensembl Metazoa database, queries a specified species dataset using the biomaRt package, and returns the results of the query. If the query includes the 'external_gene_name' attribute, representing readable gene symbols, the function examines whether any duplicates of this identifier exist across various Ensembl transcript and gene IDs rows. This situation occurs when a gene symbol corresponds to different Ensembl gene IDs and Ensembl transcript IDs. If duplicates are detected, the function modifies the external gene name to address this discrepancy.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getids</span><span class="op">(</span></span>
<span>  <span class="va">id_query</span>,</span>
<span>  <span class="va">biomart_db</span>,</span>
<span>  <span class="va">species_dataset</span>,</span>
<span>  version <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">transcript_id</span>,</span>
<span>  <span class="va">gene_id</span>,</span>
<span>  gene_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  other_ids <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-id-query">id_query<a class="anchor" aria-label="anchor" href="#arg-id-query"></a></dt>
<dd><p>A vector of transcript IDs that typically corresponds to the background transcript list</p></dd>


<dt id="arg-biomart-db">biomart_db<a class="anchor" aria-label="anchor" href="#arg-biomart-db"></a></dt>
<dd><p>The name of the BioMart database hosted by Ensembl or Ensembl Metazoa. Use <code>listEnsembl()</code> to view available Ensembl databases, or <code>listEnsemblGenomes()</code> to view available Ensembl Metazoa databases.</p></dd>


<dt id="arg-species-dataset">species_dataset<a class="anchor" aria-label="anchor" href="#arg-species-dataset"></a></dt>
<dd><p>The name of the species dataset desired on <code>ensembl.org</code>.</p></dd>


<dt id="arg-version">version<a class="anchor" aria-label="anchor" href="#arg-version"></a></dt>
<dd><p>The Ensembl version to connect to when wanting to connect to an archived Ensembl version</p></dd>


<dt id="arg-transcript-id">transcript_id<a class="anchor" aria-label="anchor" href="#arg-transcript-id"></a></dt>
<dd><p>The transcript identifier for the deregulated transcripts used in the query (e.g., "ensembl_transcript_id_version")</p></dd>


<dt id="arg-gene-id">gene_id<a class="anchor" aria-label="anchor" href="#arg-gene-id"></a></dt>
<dd><p>The gene identifier to be retrieved from the BioMart dataset of the specified species (preferably "ensembl_gene_id"  as it the identifier used in g:profiler)</p></dd>


<dt id="arg-gene-name">gene_name<a class="anchor" aria-label="anchor" href="#arg-gene-name"></a></dt>
<dd><p>A human-readable gene name identifier to be retrieved from the BioMart dataset of the specified species.</p></dd>


<dt id="arg-other-ids">other_ids<a class="anchor" aria-label="anchor" href="#arg-other-ids"></a></dt>
<dd><p>One or more additional identifiers or attributes to retrieve from the BioMart dataset of the specified species (e.g., "external_gene_name", "uniprotsptrembl", "string"). Ensure that the retrieved identifier is supported for the organism in the STRING database for subsequent analysis.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>dataframe</code> of type <em>t</em> containing the biomaRt query results with a modified "external_gene_name" column if itself present and duplicates exist.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ellis Franklin, Elise Billoir, Marie Laure Delignette-Muller, Sophie M. Prud’homme.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

